---
title: "Preprocessing of the Hoffman et al. human single-cell RNA-seq data"
author: "Tobias Tekath"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteBuilder{knitr}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

This vignette examplifies the preprocessing of **single-cell RNA-seq data for analysis with DTUrtle**. The data used in this vignette is publicly available as *Bioproject PRJNA594939* and the used *FASTQ*-files can be downloaded from [here](https://www.ebi.ac.uk/ena/browser/view/PRJNA594939). The corresponding publication from Hoffman et al. to this dataset can be found [here](https://doi.org/10.1038/s42003-020-0837-0). 

For this vignette we focus on just two of the available single-cell RNA-seq samples:

- [SAMN13541133](https://www.ebi.ac.uk/ena/browser/view/SAMN13541133) which represents a control sample (18 hour treatment with ethanol (EtOH))
- [SAMN13541131](https://www.ebi.ac.uk/ena/browser/view/SAMN13541131) which represants a sample after 2 hourse of Dexamethasone treatment (Dex2hr)

The *FASTQ*-files can be directly obtained from ENA, alternatively they are also available as *SRA*-files from [GEO](https://www.ncbi.nlm.nih.gov/bioproject/PRJNA594939), which can be converted to *FASTQ*-format.

After downloading a **MD5-check** is strongly encouraged.

### Preparing FASTQ-files
For this vignette, it is assumed that the *FASTQ*-files mentioned above have been downloaded to a directory called *samples*.

After downloading the *FASTQ*-files, the first step is to rename the files and split them into two files each. This dataset was prodcued using a Illumina SureCell 3' kit, which produces paired end reads (as all current single-cell protocols). For unknown reasons the original two *FASTQ*-files per sample have been appended for the upload. More information about the SureCell format can be found [here](https://teichlab.github.io/scg_lib_structs/methods_html/SureCell.html).

The read files can be split with the following *bash* commands:

```{bash eval=FALSE}
cd 'YOUR_PATH'/samples

#rename
mv SRR10669459.fastq.gz sc_EtOH.fastq.gz
mv SRR10669461.fastq.gz sc_Dex2hr.fastq.gz

#unpack the files
gunzip sc_*.fastq.gz 

#split
# the line numbers can be computed grepping the first linker sequence 
# or counting the number of '1'  at the end of the read name.
split -l 1215291200 --numeric-suffixes=1 --additional-suffix=.fastq.gz sc_EtOH.fastq sc_EtOH_
split -l 1394785580 --numeric-suffixes=1 --additional-suffix=.fastq.gz sc_Dex2hr.fastq sc_Dex2hr_
```

Another drawback of the concatenation of the files is, that the read names do not match between the samples. We also can overcome this with the following *bash* commands:

```{bash eval=FALSE}
#create files with matching names
cat sc_EtOH_2.fastq | sed -e 's/^\(@SRR10669459.\)\(.*\) /echo \1$((\2-303822800)) /e' > sc_EtOH_2_matched.fastq
cat sc_Dex2hr_2.fastq | sed -e 's/^\(@SRR10669461.\)\(.*\) /echo \1$((\2-348696395)) /e' > sc_Dex2hr_2_matched.fastq

#pack again
gzip sc_*_1.fastq
gzip sc_*_matched.fastq

#optionally: remove temporary files
rm *.fastq

#optionally: remove concatenated files
rm sc_EtOH.fastq.gz 
rm sc_Dex2hr.fastq.gz
```

Now there should be four packed *FASTQ*-files present in the folder, two per sample.


### Determing valid cellular barcodes
As with most single-cell data, the first real anaylsis step is the determination of valid cellular barcodes. As the Illumina SureCell 3' protocol was used, the R1 reads only contain barcode and UMI information.





Vignettes are long form documentation commonly included in packages. Because they are part of the distribution of the package, they need to be as compact as possible. The `html_vignette` output type provides a custom style sheet (and tweaks some options) to ensure that the resulting html is as small as possible. The `html_vignette` format:

- Never uses retina figures
- Has a smaller default figure size
- Uses a custom CSS stylesheet instead of the default Twitter Bootstrap style

## Vignette Info

Note the various macros within the `vignette` section of the metadata block above. These are required in order to instruct R how to build the vignette. Note that you should change the `title` field and the `\VignetteIndexEntry` to match the title of your vignette.

## Styles

The `html_vignette` template includes a basic CSS theme. To override this theme you can specify your own CSS in the document metadata as follows:

    output: 
      rmarkdown::html_vignette:
        css: mystyles.css

## Figures

The figure sizes have been customised so that you can easily put two images side-by-side. 

```{r, fig.show='hold'}
plot(1:10)
plot(10:1)
```

You can enable figure captions by `fig_caption: yes` in YAML:

    output:
      rmarkdown::html_vignette:
        fig_caption: yes

Then you can use the chunk option `fig.cap = "Your figure caption."` in **knitr**.

## More Examples

You can write math expressions, e.g. $Y = X\beta + \epsilon$, footnotes^[A footnote here.], and tables, e.g. using `knitr::kable()`.

```{r, echo=FALSE, results='asis'}
knitr::kable(head(mtcars, 10))
```

Also a quote using `>`:

> "He who gives up [code] safety for [code] speed deserves neither."
([via](https://twitter.com/hadleywickham/status/504368538874703872))
