<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Changelog • DTUrtle</title>

<!-- favicons -->
<link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png" />
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png" />
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png" />
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png" />

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>


<!-- docsearch -->
<script src="../docsearch.js"></script>
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.3/docsearch.min.css" integrity="sha256-QOSRU/ra9ActyXkIBbiIB144aDBdtvXBcNc3OTNuX/Q=" crossorigin="anonymous" />
<link href="../docsearch.css" rel="stylesheet">
<script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script>



<meta property="og:title" content="Changelog" />
<meta property="og:image" content="https://tobitekath.github.io/DTUrtle/logo.svg" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-news">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">DTUrtle</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.8.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../reference/index.html">Functions and data</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/index.html">Vignette overview</a>
    </li>
    <li class="divider"></li>
    <li class="dropdown-header">Single-cell</li>
    <li>
      <a></a>
    </li>
    <li class="dropdown-header">Illumina SureCell</li>
    <li>
      <a href="../articles/Hoffman_human_single-cell_preprocess.html">Human single-cell RNA-seq pre-processing</a>
    </li>
    <li>
      <a href="../articles/Hoffman_human_single-cell_analysis.html">Human single-cell RNA-seq analysis</a>
    </li>
    <li>
      <a></a>
    </li>
    <li class="dropdown-header">10X Chromium</li>
    <li>
      <a href="../articles/Tabular_Muris_mouse_single-cell_preprocess.html">Mouse Tabular Muris single-cell RNA-seq pre-processing</a>
    </li>
    <li>
      <a href="../articles/Tabular_Muris_mouse_single-cell_analysis.html">Mouse Tabular Muris single-cell RNA-seq analysis</a>
    </li>
    <li>
      <a></a>
    </li>
    <li class="dropdown-header">Smart-seq2</li>
    <li>
      <a href="../articles/Wuidart_mouse_single-cell_preprocess.html">Mouse single-cell RNA-seq pre-processing</a>
    </li>
    <li>
      <a href="../articles/Wuidart_mouse_single-cell_analysis.html">Mouse single-cell RNA-seq analysis</a>
    </li>
    <li class="divider"></li>
    <li class="dropdown-header">Bulk</li>
    <li>
      <a href="../articles/Hoffman_human_bulk_preprocess.html">Human bulk RNA-seq pre-processing</a>
    </li>
    <li>
      <a href="../articles/Hoffman_human_bulk_analysis.html">Human bulk RNA-seq analysis</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/TobiTekath/DTUrtle/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
      <form class="navbar-form navbar-right hidden-xs hidden-sm" role="search">
        <div class="form-group">
          <input type="search" class="form-control" name="search-input" id="search-input" placeholder="Search..." aria-label="Search for..." autocomplete="off">
        </div>
      </form>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
      <h1 data-toc-skip>Changelog <small></small></h1>
      <small>Source: <a href='https://github.com/TobiTekath/DTUrtle/blob/master/NEWS.md'><code>NEWS.md</code></a></small>
    </div>

    <div id="dturtle-0-8-1" class="section level1">
<h1 class="page-header" data-toc-text="0.8.1">
<a href="#dturtle-0-8-1" class="anchor"></a>DTUrtle 0.8.1</h1>
<div id="changes" class="section level2">
<h2 class="hasAnchor">
<a href="#changes" class="anchor"></a>Changes</h2>
<ul>
<li>remodeled a lot of code - preserving functionality if not stated otherwise.</li>
<li>added basic DGE-calling methods:
<ul>
<li>
<code><a href="../reference/import_dge_counts.html">import_dge_counts()</a></code>: import quantification data and summarize to gene-level.</li>
<li>
<code><a href="../reference/run_deseq2.html">run_deseq2()</a></code>: perform DGE analysis with DESeq2.</li>
</ul>
</li>
<li>added <code><a href="../reference/priming_bias_detection_probability.html">priming_bias_detection_probability()</a></code> functionality, to estimate the potential influence of a prime-biased sequencing protocol on the DTU detection ability.</li>
<li>added Vignettes for a Smart-seq2 single-cell data set from Wuidart et al.</li>
<li>Vignettes: added sections showcasing DGE calling and priming bias detection probability calculation.</li>
<li>new Dependency: <code>DESeq2</code>
</li>
<li>new Suggests: <code>glmGamPoi</code>, <code>apeglm</code>
</li>
</ul>
</div>
</div>
    <div id="dturtle-0-7-4" class="section level1">
<h1 class="page-header" data-toc-text="0.7.4">
<a href="#dturtle-0-7-4" class="anchor"></a>DTUrtle 0.7.4</h1>
<div id="changes-1" class="section level2">
<h2 class="hasAnchor">
<a href="#changes-1" class="anchor"></a>Changes</h2>
<ul>
<li>
<code><a href="../reference/plot_transcripts_view.html">plot_transcripts_view()</a></code>: added <code>tested_transcripts_only</code> parameter, to allow better control which transcripts of a gene should be displayed. Also allows to create plot for genes which have been excluded from the analysis.</li>
<li>
<code><a href="../reference/plot_transcripts_view.html">plot_transcripts_view()</a></code>: changed default value of <code>reduce_introns_fill</code> to ‘white’ to get visually more appealing plots by default.</li>
</ul>
</div>
</div>
    <div id="dturtle-0-7-3" class="section level1">
<h1 class="page-header" data-toc-text="0.7.3">
<a href="#dturtle-0-7-3" class="anchor"></a>DTUrtle 0.7.3</h1>
<div id="changes-2" class="section level2">
<h2 class="hasAnchor">
<a href="#changes-2" class="anchor"></a>Changes</h2>
<ul>
<li>readded “GenomicFeatures” as dependency - eases installation.</li>
<li>measure and show computation time in vignettes.</li>
<li>extended vignettes.</li>
<li>extended README.</li>
<li>
<code><a href="../reference/plot_proportion_barplot.html">plot_proportion_barplot()</a></code>: define data and mapping already in general ggplot function, to ease editability.</li>
<li>
<code><a href="../reference/plot_proportion_barplot.html">plot_proportion_barplot()</a></code>: added global ‘text_size’ parameter.</li>
<li>
<code><a href="../reference/plot_dimensional_reduction.html">plot_dimensional_reduction()</a></code>: added ‘plot_scale’ parameter, allowing for free x- or y-axis of facets, or both.</li>
<li>
<code><a href="../reference/plot_dimensional_reduction.html">plot_dimensional_reduction()</a></code>: added global ‘text_size’ parameter.</li>
<li>
<code><a href="../reference/plot_transcripts_view.html">plot_transcripts_view()</a></code>: added ‘arrow_start’ parameter, to allow adjustment of the transcript annotation start position.</li>
<li>
<code><a href="../reference/plot_transcripts_view.html">plot_transcripts_view()</a></code>: changed default parameters and resolution, to get visually more appealing high quality plots.</li>
<li>
<code><a href="../reference/import_gtf.html">import_gtf()</a></code>: generalized function and added parameter for returned feature type and return format. Now can create fitting <code><a href="../reference/plot_transcripts_view.html">plot_transcripts_view()</a></code> GTF input.</li>
<li>
<code><a href="../reference/run_drimseq.html">run_drimseq()</a></code>: added <code>susbet_feature</code> and <code>subset_sample</code> parameters, replacing <code>subset_counts</code>. Now samples and features can be excluded from analysis on the fly.</li>
<li>
<code><a href="../reference/run_drimseq.html">run_drimseq()</a></code>: added <code>add_pseudocount</code> parameter, to enable comparisons of features with zero expression in one comparison group.</li>
<li>
<code><a href="../reference/run_drimseq.html">run_drimseq()</a></code>: added <code>filter_only</code> parameter, to return (sparse) count matrix after group-based filtering.</li>
<li>
<code><a href="../reference/combine_to_matrix.html">combine_to_matrix()</a></code>: added ability to infer Seurat cellname extension.</li>
<li>removed dependency of ‘stringi’.</li>
<li>added ‘sparseDRIMSeq’ to depending packages, to get rid of sometimes not-occuring package load.</li>
<li>fixed some smaller bugs.</li>
</ul>
</div>
</div>
    <div id="dturtle-0-7-1" class="section level1">
<h1 class="page-header" data-toc-text="0.7.1">
<a href="#dturtle-0-7-1" class="anchor"></a>DTUrtle 0.7.1</h1>
<div id="changes-3" class="section level2">
<h2 class="hasAnchor">
<a href="#changes-3" class="anchor"></a>Changes</h2>
<ul>
<li>add new vignette for Tabula Muris (Mouse) data set.</li>
<li>
<code><a href="../reference/combine_to_matrix.html">combine_to_matrix()</a></code>: added <code>cell_extension_side</code> parameter to choose label append or prepend.</li>
<li>
<code><a href="../reference/plot_dimensional_reduction.html">plot_dimensional_reduction()</a></code>: added functionality to plot in dimensional reduction space.</li>
<li>
<code><a href="../reference/plot_transcripts_view.html">plot_transcripts_view()</a></code>: safer and more specific IdeoGram track generation.</li>
<li>
<code><a href="../reference/plot_transcripts_view.html">plot_transcripts_view()</a></code>: improved handling of non-UCSC chromosome identifiers.</li>
<li>
<code><a href="../reference/run_drimseq.html">run_drimseq()</a></code>: added check for matching names of counts and tx2gene.</li>
<li>
<code><a href="../reference/run_drimseq.html">run_drimseq()</a></code>: added <code>subset_counts</code> parameter - ability to force subset the count matrix to matching tx2gene features only.</li>
<li>
<code><a href="../reference/plot_dtu_table.html">plot_dtu_table()</a></code>: disable warning for server side rendering.</li>
<li>
<code><a href="../reference/plot_proportion_pheatmap.html">plot_proportion_pheatmap()</a></code>: plots names are now run through <code><a href="https://rdrr.io/r/base/make.names.html">make.names()</a></code>, as all other plots.</li>
<li>
<code><a href="../reference/create_dtu_table.html">create_dtu_table()</a></code>: changed default column names and added description of column meaning.</li>
<li>extended checks for provided file paths.</li>
<li>safer BiocParallel starting and stopping.</li>
<li>made whole package documentation searchable via docsearch.</li>
<li>README: fixed typo and added link to vignettes.</li>
<li>set GenomeInfoDb as standard Imports.</li>
<li>minor improvements.</li>
</ul>
</div>
</div>
    <div id="dturtle-0-2-6" class="section level1">
<h1 class="page-header" data-toc-text="0.2.6">
<a href="#dturtle-0-2-6" class="anchor"></a>DTUrtle 0.2.6</h1>
<div id="changes-4" class="section level2">
<h2 class="hasAnchor">
<a href="#changes-4" class="anchor"></a>Changes</h2>
<ul>
<li>add <code>add_to_table</code> parameter to plotting functions, to add plot savepaths directly to <code>dtu_table</code>.</li>
<li>
<code><a href="../reference/run_drimseq.html">run_drimseq()</a></code>: added description of predefined filter strategies.</li>
<li>
<code><a href="../reference/run_drimseq.html">run_drimseq()</a></code>: added check if two sample groups are left after excluding unused samples.</li>
<li>minor update of function documentation.</li>
<li>
<code><a href="../reference/plot_dtu_table.html">plot_dtu_table()</a></code>: add <code>min_page_length</code> parameter.</li>
<li>
<code><a href="../reference/plot_dtu_table.html">plot_dtu_table()</a></code>: slightly change icon and heading alignment.</li>
<li>added conditional vignette evaluation.</li>
<li>added <code>knitr</code> in Suggests for vignette building.</li>
</ul>
</div>
</div>
    <div id="dturtle-0-2-5" class="section level1">
<h1 class="page-header" data-toc-text="0.2.5">
<a href="#dturtle-0-2-5" class="anchor"></a>DTUrtle 0.2.5</h1>
<div id="changes-5" class="section level2">
<h2 class="hasAnchor">
<a href="#changes-5" class="anchor"></a>Changes</h2>
<ul>
<li>initial first version.</li>
</ul>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>

</div>


      <footer>
      <div class="copyright">
  <p>Developed by Tobias Tekath.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  
<script src="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.js" integrity="sha256-GKvGqXDznoRYHCwKXGnuchvKSwmx9SRMrZOTh2g4Sb0=" crossorigin="anonymous"></script>
<script>
  docsearch({
    
    
    apiKey: 'a6c9f43b010fdd4b2e64792389fd67d2',
    indexName: 'dturtle',
    inputSelector: 'input#search-input.form-control',
    transformData: function(hits) {
      return hits.map(function (hit) {
        hit.url = updateHitURL(hit);
        return hit;
      });
    }
  });
</script>



  </body>
</html>


