<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Differential Transcript Usage analysis of bulk and single-cell RNA-seq data • DTUrtle</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><!-- docsearch --><script src="docsearch.js"></script><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.3/docsearch.min.css" integrity="sha256-QOSRU/ra9ActyXkIBbiIB144aDBdtvXBcNc3OTNuX/Q=" crossorigin="anonymous">
<link href="docsearch.css" rel="stylesheet">
<script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script><meta property="og:title" content="Differential Transcript Usage analysis of bulk and single-cell RNA-seq data">
<meta property="og:description" content="Easily perform and visualize Differential Transcript Usage (DTU) analysis for bulk or single-cell RNA-seq data, utilizing transcript level quantification counts.">
<meta property="og:image" content="https://tobitekath.github.io/DTUrtle/logo.svg">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">DTUrtle</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="reference/index.html">Functions and data</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/index.html">Vignette overview</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Single-cell</li>
    <li>
      <a></a>
    </li>
    <li class="dropdown-header">Illumina SureCell</li>
    <li>
      <a href="articles/Hoffman_human_single-cell_preprocess.html">Human single-cell RNA-seq pre-processing</a>
    </li>
    <li>
      <a href="articles/Hoffman_human_single-cell_analysis.html">Human single-cell RNA-seq analysis</a>
    </li>
    <li>
      <a></a>
    </li>
    <li class="dropdown-header">10X Chromium</li>
    <li>
      <a href="articles/Tabular_Muris_mouse_single-cell_preprocess.html">Mouse Tabular Muris single-cell RNA-seq pre-processing</a>
    </li>
    <li>
      <a href="articles/Tabular_Muris_mouse_single-cell_analysis.html">Mouse Tabular Muris single-cell RNA-seq analysis</a>
    </li>
    <li>
      <a></a>
    </li>
    <li class="dropdown-header">Smart-seq2</li>
    <li>
      <a href="articles/Wuidart_mouse_single-cell_preprocess.html">Mouse single-cell RNA-seq pre-processing</a>
    </li>
    <li>
      <a href="articles/Wuidart_mouse_single-cell_analysis.html">Mouse single-cell RNA-seq analysis</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Bulk</li>
    <li>
      <a href="articles/Hoffman_human_bulk_preprocess.html">Human bulk RNA-seq pre-processing</a>
    </li>
    <li>
      <a href="articles/Hoffman_human_bulk_analysis.html">Human bulk RNA-seq analysis</a>
    </li>
  </ul>
</li>
<li>
  <a href="news/index.html">News</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/TobiTekath/DTUrtle/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
<form class="navbar-form navbar-right hidden-xs hidden-sm" role="search">
        <div class="form-group">
          <input type="search" class="form-control" name="search-input" id="search-input" placeholder="Search..." aria-label="Search for..." autocomplete="off">
</div>
      </form>
      
    </div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div id="dturtle" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#dturtle" class="anchor"></a>DTUrtle <img src="reference/figures/logo.svg" align="right" alt="" width="250">
</h1></div>
<p><strong>Perform differential transcript usage (DTU) analysis of bulk or single-cell RNA-seq data and visualize the results.</strong></p>
<!-- badges: start -->
<p><a href="https://www.tidyverse.org/lifecycle/#stable"><img src="https://img.shields.io/badge/lifecycle-stable-brightgreen.svg" alt="Lifecycle: stable"></a> <a href="https://github.com/TobiTekath/DTUrtle/actions"><img src="https://github.com/TobiTekath/DTUrtle/workflows/R-CMD-check/badge.svg" alt="R build status"></a> <!-- badges: end --></p>
<div id="background" class="section level2">
<h2 class="hasAnchor">
<a href="#background" class="anchor"></a>Background</h2>
<p>While most RNA-seq data sets are mainly analyzed in regard to differential gene expression, the transcript-level nature of the RNA-seq reads is often overlooked. The <strong>R package DTUrtle</strong> offers an easy-to-use way to perform a differential transcript usage (DTU) analysis for bulk or single-cell RNA-seq data, comparing the expression proportions of a gene’s distinct transcript isoforms between conditions. The results, a list of significantly differential genes and transcript isoforms, can be aggregated in an overview table and visualized in multiple ways. As a DTU analysis is best accompanied by a differential gene expression (DGE) analysis, DTUrtle also offers a basic DGE calling workflow for bulk and single-cell RNA-seq data.</p>
</div>
<div id="target-audience" class="section level2">
<h2 class="hasAnchor">
<a href="#target-audience" class="anchor"></a>Target audience</h2>
<p>No in-depth knowledge about DTU or the statistical background of the analysis is needed. If you have data of an RNA-seq experiment at hand (or are able to download data from a public repository), you are good to go. Three exemplary vignettes for different species and sequencing techniques will guide you through your analysis.</p>
</div>
<div id="required-input-data" class="section level2">
<h2 class="hasAnchor">
<a href="#required-input-data" class="anchor"></a>Required input data</h2>
<p>DTUrtle relies on transcript-level expression counts, which can be generated with transcript-level quantifiers like Salmon, Alevin, kallisto, etc. from ordinary RNA-seq fastq files. As an alternative, even standard splice-aware genomic aligners like STAR (generating a transcript-level BAM with ‘–quantMode TranscriptomeSAM’ and then quantifying it with Salmon) can be used for data generation.</p>
<p>Besides the expression counts, a transcript-level annotation file for your species is needed (which is also required for the transcript-level quantification).</p>
</div>
<div id="more-information" class="section level2">
<h2 class="hasAnchor">
<a href="#more-information" class="anchor"></a>More information</h2>
<p>For more information about the used methods and package capabilities, please refer to the publication in Bioinformatics:</p>
<p><a href="https://doi.org/10.1093/bioinformatics/btab629">Tekath, T., &amp; Dugas, M. (2021). Differential transcript usage analysis of bulk and single-cell RNA-seq data with DTUrtle.</a></p>
<hr>
</div>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<p>Install from GitHub:</p>
<div class="sourceCode" id="cb1"><pre class="downlit">
<span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span><span class="op">(</span><span class="st">"remotes"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span><span class="op">{</span>
    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"remotes"</span><span class="op">)</span>
<span class="op">}</span>
<span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html">install_github</a></span><span class="op">(</span><span class="st">"TobiTekath/DTUrtle"</span><span class="op">)</span></pre></div>
<p>As DTUrtle depends on multiple packages, the installation might fail due to the complex dependency structure. Normally, trying the above command multiple times helps (installed packages of the command are kept and reduce the dependency complexity). You might also try to install the needed Bioconductor packages beforehand:</p>
<div class="sourceCode" id="cb2"><pre class="downlit">
<span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span><span class="op">{</span>
    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span>
<span class="op">}</span>
<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"BiocParallel"</span>, <span class="st">"GenomicRanges"</span>, <span class="st">"Gviz"</span>, <span class="st">"rtracklayer"</span>, <span class="st">"stageR"</span>, <span class="st">"tximport"</span>, <span class="st">"DESeq2"</span><span class="op">)</span><span class="op">)</span></pre></div>
<hr>
<p>If you are not on R &gt;= 4.0 and installation of the newest Gviz version fails, you can manually install an older Gviz version, for example with:</p>
<div class="sourceCode" id="cb3"><pre class="downlit">
<span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html">install_github</a></span><span class="op">(</span><span class="st">"https://github.com/ivanek/Gviz/tree/RELEASE_3_10"</span><span class="op">)</span></pre></div>
<p>DTUrtle does <strong>not</strong> require R &gt;= 4.0 and should work fine with older Gviz versions.</p>
<hr>
</div>
<div id="basic-workflow" class="section level2">
<h2 class="hasAnchor">
<a href="#basic-workflow" class="anchor"></a>Basic workflow</h2>
<ul>
<li><strong>See vignettes for a more detailed workflow with real data</strong></li>
</ul>
<p><img src="reference/figures/DTUrtle_workflow.svg" height="1000px" style="display: block; margin: auto;"></p>
</div>
<div id="vignettes" class="section level2">
<h2 class="hasAnchor">
<a href="#vignettes" class="anchor"></a>Vignettes</h2>
<p>Extensive documentation can be found <a href="https://tobitekath.github.io/DTUrtle/">here</a>, including multiple <a href="https://tobitekath.github.io/DTUrtle/articles/index.html">Vignettes</a> analyzing real world data sets.</p>
</div>
<div id="dturtle-minimal-workflow" class="section level2">
<h2 class="hasAnchor">
<a href="#dturtle-minimal-workflow" class="anchor"></a>DTUrtle minimal workflow</h2>
<p>A minimal DTUrtle workflow might look like this:</p>
<div id="setup-environment" class="section level3">
<h3 class="hasAnchor">
<a href="#setup-environment" class="anchor"></a>Setup environment</h3>
<div class="sourceCode" id="cb4"><pre class="downlit">
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/TobiTekath/DTUrtle">DTUrtle</a></span><span class="op">)</span>

<span class="co">#the BiocParallel framework is used to parallelize the computations.</span>
    <span class="co">#Using 4 cores:</span>
    <span class="va">biocpar</span> <span class="op">&lt;-</span> <span class="fu">BiocParallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/MulticoreParam-class.html">MulticoreParam</a></span><span class="op">(</span><span class="fl">4</span><span class="op">)</span>
    <span class="co">#or standard serial computation (only 1 core)</span>
    <span class="co">#biocpar &lt;- BiocParallel::SerialParam()</span>
<span class="co">#multiple other options available for computational clusters.</span></pre></div>
</div>
<div id="import-and-format-data" class="section level3">
<h3 class="hasAnchor">
<a href="#import-and-format-data" class="anchor"></a>Import and format data</h3>
<div class="sourceCode" id="cb5"><pre class="downlit">
<span class="co">#import gtf Annotation to get transcript to gene mapping</span>
<span class="va">tx2gene</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/import_gtf.html">import_gtf</a></span><span class="op">(</span>gtf_file <span class="op">=</span> <span class="st">"path_to_your_gtf_file.gtf"</span><span class="op">)</span>

<span class="co">##optional:</span>
    <span class="co">#move transcript and gene identifier columns to front</span>
    <span class="va">tx2gene</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/move_columns_to_front.html">move_columns_to_front</a></span><span class="op">(</span>df <span class="op">=</span> <span class="va">tx2gene</span>, 
                                     columns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"transcript_name"</span>, <span class="st">"gene_name"</span><span class="op">)</span><span class="op">)</span>
    <span class="co">#ensure that a one to one mapping between names and IDs exists</span>
    <span class="va">tx2gene</span><span class="op">$</span><span class="va">gene_name</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/one_to_one_mapping.html">one_to_one_mapping</a></span><span class="op">(</span>name <span class="op">=</span> <span class="va">tx2gene</span><span class="op">$</span><span class="va">gene_name</span>, 
                                            id <span class="op">=</span> <span class="va">tx2gene</span><span class="op">$</span><span class="va">gene_id</span><span class="op">)</span>
    <span class="va">tx2gene</span><span class="op">$</span><span class="va">transcript_name</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/one_to_one_mapping.html">one_to_one_mapping</a></span><span class="op">(</span>name <span class="op">=</span> <span class="va">tx2gene</span><span class="op">$</span><span class="va">transcript_name</span>, 
                                                  id <span class="op">=</span> <span class="va">tx2gene</span><span class="op">$</span><span class="va">transcript_id</span><span class="op">)</span>

<span class="co">#import transcript-level quantification data, for example from Salmon, and perform appropriate scaling</span>
<span class="va">files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Sys.glob.html">Sys.glob</a></span><span class="op">(</span><span class="st">"path_to_your_data/*/quant.sf"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">files</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span><span class="op">(</span><span class="st">".*/"</span>,<span class="st">""</span>,<span class="fu"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span><span class="op">(</span><span class="st">"/quant.sf"</span>,<span class="st">""</span>,<span class="va">files</span><span class="op">)</span><span class="op">)</span>
<span class="va">cts</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/import_counts.html">import_counts</a></span><span class="op">(</span>files <span class="op">=</span> <span class="va">files</span>, type <span class="op">=</span> <span class="st">"salmon"</span><span class="op">)</span>

<span class="co">##for single-cell data only:</span>
    <span class="co">#import_counts returned a list of matrices -&gt; combine them to one matrix</span>
    <span class="va">cts</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/combine_to_matrix.html">combine_to_matrix</a></span><span class="op">(</span>tx_list <span class="op">=</span> <span class="va">cts</span><span class="op">)</span>

<span class="co">#create a sample data sheet, specifying which sample / cell belongs to which group</span>
<span class="va">pd</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="st">"id"</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">cts</span><span class="op">)</span>, <span class="st">"group"</span><span class="op">=</span><span class="st">"your_grouping_variable"</span>, 
                 stringsAsFactors <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></pre></div>
</div>
<div id="dtu-analysis" class="section level3">
<h3 class="hasAnchor">
<a href="#dtu-analysis" class="anchor"></a>DTU analysis</h3>
<p>The <code>dturtle</code> object is an easy-to-access list, containing all necessary analysis information and results</p>
<div class="sourceCode" id="cb6"><pre class="downlit">
<span class="co">#use DRIMSeq for fitting a Dirichlet-multinomial model --- appropriate parameters are chosen based on the selected filtering_strategy ('bulk' or 'sc')</span>
<span class="va">dturtle</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/run_drimseq.html">run_drimseq</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">cts</span>, tx2gene <span class="op">=</span> <span class="va">tx2gene</span>, pd<span class="op">=</span><span class="va">pd</span>, id_col <span class="op">=</span> <span class="st">"id"</span>,
                    cond_col <span class="op">=</span> <span class="st">"group"</span>, filtering_strategy <span class="op">=</span> <span class="st">"bulk"</span>, 
                    BPPARAM <span class="op">=</span> <span class="va">biocpar</span><span class="op">)</span>

<span class="co">#run posthoc filtering and two-staged statistical correction with stageR</span>
<span class="va">dturtle</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/posthoc_and_stager.html">posthoc_and_stager</a></span><span class="op">(</span>dturtle <span class="op">=</span> <span class="va">dturtle</span>, ofdr <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span></pre></div>
</div>
<div id="optional-dge-analysis" class="section level3">
<h3 class="hasAnchor">
<a href="#optional-dge-analysis" class="anchor"></a>(optional) DGE analysis</h3>
<p>Alongside the DTU analysis, a DGE analysis might be of interest. The transcript-level count matrices have been scaled for the DTU analysis, but for a DGE analysis preferably un-normalized counts are used. Therefore the counts are re-imported and summarized to gene-level.</p>
<div class="sourceCode" id="cb7"><pre class="downlit">
<span class="co">#import counts and summarize to gene-level --- simply reuse the already selected files.</span>
<span class="va">cts_dge</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/import_dge_counts.html">import_dge_counts</a></span><span class="op">(</span><span class="va">files</span>, type<span class="op">=</span><span class="st">"salmon"</span>, tx2gene<span class="op">=</span><span class="va">tx2gene</span><span class="op">)</span>

<span class="co">##for single-cell data only:</span>
    <span class="co">#import_counts returned a list of matrices -&gt; combine them to one matrix</span>
    <span class="va">cts_dge</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/combine_to_matrix.html">combine_to_matrix</a></span><span class="op">(</span>tx_list <span class="op">=</span> <span class="va">cts_dge</span><span class="op">)</span>

<span class="co">#perform the DGE calling with DESeq2 --- appropriate parameters are chosen based on the selected dge_calling_strategy ('bulk' or 'sc')</span>
<span class="va">dturtle</span><span class="op">$</span><span class="va">dge_analysis</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/run_deseq2.html">run_deseq2</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">cts_dge</span>, pd <span class="op">=</span> <span class="va">pd</span>, id_col <span class="op">=</span> <span class="st">"id"</span>, cond_col <span class="op">=</span> <span class="st">"group"</span>, lfc_threshold <span class="op">=</span> <span class="fl">0.5</span>,
                                   sig_threshold <span class="op">=</span> <span class="fl">0.01</span>, dge_calling_strategy <span class="op">=</span> <span class="st">"bulk"</span>, BPPARAM <span class="op">=</span> <span class="va">biocpar</span><span class="op">)</span></pre></div>
</div>
<div id="optional-estimate-influence-of-a-potential-priming-bias" class="section level3">
<h3 class="hasAnchor">
<a href="#optional-estimate-influence-of-a-potential-priming-bias" class="anchor"></a>(optional) Estimate influence of a potential priming bias</h3>
<p>If a prime-biased (single-cell) RNA-seq protocol was used, the ability to detect DTU in certain transcripts might be impaired. DTUrtle offers the calculation of a novel detection probability score to infer, which transcripts might be hard to detect with the given data - and thus might not pop-up in the DTU analysis. Suppose we have 3’-biased data (as generated by many prominent single-cell protocols), we can infer the effect of this bias on the transcripts by:</p>
<div class="sourceCode" id="cb8"><pre class="downlit">
<span class="co">#Attention: calculation for all available genes might be time consuming.</span>
<span class="va">priming_bias_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/priming_bias_detection_probability.html">priming_bias_detection_probability</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">cts</span>, gtf <span class="op">=</span> <span class="st">"path_to_your_gtf_file.gtf"</span>, tx2gene <span class="op">=</span> <span class="va">tx2gene</span>, one_to_one <span class="op">=</span> <span class="cn">TRUE</span>,
                                                      priming_enrichment <span class="op">=</span> <span class="st">"3"</span>, BPPARAM <span class="op">=</span> <span class="va">biocpar</span><span class="op">)</span></pre></div>
</div>
<div id="result-aggregation-and-visualization" class="section level3">
<h3 class="hasAnchor">
<a href="#result-aggregation-and-visualization" class="anchor"></a>Result aggregation and visualization</h3>
<div class="sourceCode" id="cb9"><pre class="downlit">
<span class="co">#highly flexible function to create a results data frame</span>
<span class="va">dturtle</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/create_dtu_table.html">create_dtu_table</a></span><span class="op">(</span>dturtle <span class="op">=</span> <span class="va">dturtle</span><span class="op">)</span>

    <span class="co">## View results data frame</span>
    <span class="fu"><a href="https://rdrr.io/r/utils/View.html">View</a></span><span class="op">(</span><span class="va">dturtle</span><span class="op">$</span><span class="va">dtu_table</span><span class="op">)</span>

<span class="co">#change to results folder</span>
<span class="fu"><a href="https://rdrr.io/r/base/getwd.html">setwd</a></span><span class="op">(</span><span class="st">"my_results_folder"</span><span class="op">)</span>    

<span class="co">#create plots, save them to disk and link them in the `dtu_table`.</span>
<span class="va">dturtle</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/plot_proportion_barplot.html">plot_proportion_barplot</a></span><span class="op">(</span>dturtle <span class="op">=</span> <span class="va">dturtle</span>, 
                                   savepath <span class="op">=</span> <span class="st">"images"</span>, 
                                   add_to_table <span class="op">=</span> <span class="st">"barplot"</span>,
                                   BPPARAM <span class="op">=</span> <span class="va">biocpar</span><span class="op">)</span>

<span class="va">dturtle</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/plot_proportion_pheatmap.html">plot_proportion_pheatmap</a></span><span class="op">(</span>dturtle <span class="op">=</span> <span class="va">dturtle</span>, 
                                    savepath <span class="op">=</span> <span class="st">"images"</span>, 
                                    include_expression <span class="op">=</span> <span class="cn">TRUE</span>,
                                    add_to_table <span class="op">=</span> <span class="st">"pheatmap"</span>,
                                    BPPARAM <span class="op">=</span> <span class="va">biocpar</span><span class="op">)</span>

<span class="va">dturtle</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/plot_transcripts_view.html">plot_transcripts_view</a></span><span class="op">(</span>dturtle <span class="op">=</span> <span class="va">dturtle</span>, 
                                 gtf <span class="op">=</span> <span class="st">"path_to_your_gtf_file.gtf"</span>, 
                                 genome <span class="op">=</span> <span class="st">'hg38'</span>, 
                                 one_to_one <span class="op">=</span> <span class="cn">TRUE</span>,
                                 savepath <span class="op">=</span> <span class="st">"images"</span>, 
                                 add_to_table <span class="op">=</span> <span class="st">"transcript_view"</span>,
                                 BPPARAM <span class="op">=</span> <span class="va">biocpar</span><span class="op">)</span>

<span class="va">dturtle</span> <span class="op">&lt;-</span> <span class="fu">plot_dimensional_redcution</span><span class="op">(</span>dturtle <span class="op">=</span> <span class="va">dturtle</span>,
                                      reduction_df <span class="op">=</span> <span class="st">"some_kind_of_dimensional_reduction_coordinates"</span>,
                                      savepath <span class="op">=</span> <span class="st">"images"</span>,
                                      add_to_table <span class="op">=</span> <span class="st">"dimensional_reduction"</span>,
                                      BPPARAM <span class="op">=</span> <span class="va">biocpar</span><span class="op">)</span>

<span class="co">#create interactive HTML-table from results data frame</span>
    <span class="co">#optional: specify colorful column formatters</span>
    <span class="va">column_formatter_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>
      <span class="st">"gene_qvalue"</span> <span class="op">=</span> <span class="fu"><a href="reference/table_pval_tile.html">table_pval_tile</a></span><span class="op">(</span><span class="st">"white"</span>, <span class="st">"orange"</span>, digits <span class="op">=</span> <span class="fl">3</span><span class="op">)</span>,
      <span class="st">"minimal_tx_qvalue"</span> <span class="op">=</span> <span class="fu"><a href="reference/table_pval_tile.html">table_pval_tile</a></span><span class="op">(</span><span class="st">"white"</span>, <span class="st">"orange"</span>, digits <span class="op">=</span> <span class="fl">3</span><span class="op">)</span>,
      <span class="st">"number_tx"</span> <span class="op">=</span> <span class="fu">formattable</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/formattable/man/color_tile.html">color_tile</a></span><span class="op">(</span><span class="st">'white'</span>, <span class="st">"lightblue"</span><span class="op">)</span>,
      <span class="st">"number_significant_tx"</span> <span class="op">=</span> <span class="fu">formattable</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/formattable/man/color_tile.html">color_tile</a></span><span class="op">(</span><span class="st">'white'</span>, <span class="st">"lightblue"</span><span class="op">)</span>,
      <span class="st">"max(Condition1-Condition2)"</span> <span class="op">=</span> <span class="fu"><a href="reference/table_percentage_bar.html">table_percentage_bar</a></span><span class="op">(</span><span class="st">'lightgreen'</span>, <span class="st">"#FF9999"</span>, digits<span class="op">=</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span>

<span class="fu"><a href="reference/plot_dtu_table.html">plot_dtu_table</a></span><span class="op">(</span>dturtle <span class="op">=</span> <span class="va">dturtle</span>, savepath <span class="op">=</span> <span class="st">"my_results.html"</span>, 
               column_formatters <span class="op">=</span> <span class="va">column_formatter_list</span><span class="op">)</span></pre></div>
<p><strong>See <a href="https://tobitekath.github.io/DTUrtle/articles/index.html">Vignettes</a> for a more detailed workflow with real data</strong></p>
</div>
<div id="help-feedback" class="section level3">
<h3 class="hasAnchor">
<a href="#help-feedback" class="anchor"></a>Help &amp; Feedback</h3>
<p>If you have questions, need help with your analysis or found a bug: Please do not hesitate to <a href="mailto:tobias.tekath@wwu.de">get in touch</a> or open a GitHub issue.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Browse source code at <br><a href="https://github.com/TobiTekath/DTUrtle/">https://​github.com/​TobiTekath/​DTUrtle/​</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li>GPL (&gt;= 3) + file <a href="LICENSE-text.html">LICENSE</a>
</li>
</ul>
</div>
<div class="citation">
<h2>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html">Citing DTUrtle</a></li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Tobias Tekath <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0002-9315-5452" target="orcid.widget" aria-label="ORCID"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>

  </div>
</div>


      <footer><div class="copyright">
  <p>Developed by Tobias Tekath.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  
<script src="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.js" integrity="sha256-GKvGqXDznoRYHCwKXGnuchvKSwmx9SRMrZOTh2g4Sb0=" crossorigin="anonymous"></script><script>
  docsearch({
    
    
    apiKey: 'a6c9f43b010fdd4b2e64792389fd67d2',
    indexName: 'dturtle',
    inputSelector: 'input#search-input.form-control',
    transformData: function(hits) {
      return hits.map(function (hit) {
        hit.url = updateHitURL(hit);
        return hit;
      });
    }
  });
</script>
</body>
</html>
