<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Pre-processing of the Hoffman et al. human bulk RNA-seq data • DTUrtle</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Pre-processing of the Hoffman et al. human bulk RNA-seq data">
<meta property="og:description" content="DTUrtle">
<meta property="og:image" content="https://tobitekath.github.io/DTUrtle/logo.svg">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">DTUrtle</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.2.6</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Functions and data</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Hoffman_human_single-cell_preprocess.html">Human single-cell RNA-seq pre-processing</a>
    </li>
    <li>
      <a href="../articles/Hoffman_human_single-cell_analysis.html">Human single-cell RNA-seq analysis</a>
    </li>
    <li>
      <a href="../articles/Hoffman_human_bulk_preprocess.html">Human bulk RNA-seq pre-processing</a>
    </li>
    <li>
      <a href="../articles/Hoffman_human_bulk_analysis.html">Human bulk RNA-seq analysis</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/TobiTekath/DTUrtle/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Pre-processing of the Hoffman et al. human bulk RNA-seq data</h1>
                        <h4 class="author">Tobias Tekath</h4>
            
            <h4 class="date">2020-07-31</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/TobiTekath/DTUrtle/blob/master/vignettes/Hoffman_human_bulk_preprocess.Rmd"><code>vignettes/Hoffman_human_bulk_preprocess.Rmd</code></a></small>
      <div class="hidden name"><code>Hoffman_human_bulk_preprocess.Rmd</code></div>

    </div>

    
    
<p>This vignette exemplifies the pre-processing of <strong>bulk RNA-seq data for analysis with DTUrtle</strong>. The data used in this vignette is publicly available as <em>Bioproject PRJNA594939</em> and the used <em>FASTQ</em>-files can be downloaded from <a href="https://www.ebi.ac.uk/ena/browser/view/PRJNA594939">here</a>. The corresponding publication from Hoffman et al. based on this data set can be found <a href="https://doi.org/10.1038/s42003-020-0837-0">here</a>.</p>
<p>For this vignette we focus on two groups with 3 biological replicates each:</p>
<ul>
<li>Group 1 - Control samples (EtOH):
<ul>
<li><a href="https://www.ebi.ac.uk/ena/browser/view/SAMN13541125">SAMN13541125</a></li>
<li><a href="https://www.ebi.ac.uk/ena/browser/view/SAMN13541124">SAMN13541124</a></li>
<li><a href="https://www.ebi.ac.uk/ena/browser/view/SAMN13541123">SAMN13541123</a></li>
</ul>
</li>
<li>Group 2 - Samples after 2 hours of Dexamethasone treatment (Dex2hr):
<ul>
<li><a href="https://www.ebi.ac.uk/ena/browser/view/SAMN13541127">SAMN13541127</a></li>
<li><a href="https://www.ebi.ac.uk/ena/browser/view/SAMN13541119">SAMN13541119</a></li>
<li><a href="https://www.ebi.ac.uk/ena/browser/view/SAMN13541118">SAMN13541118</a></li>
</ul>
</li>
</ul>
<p>The <em>FASTQ</em>-files can be directly obtained from ENA, alternatively they are also available as <em>SRA</em>-files from <a href="https://www.ncbi.nlm.nih.gov/bioproject/PRJNA594939">GEO</a>, which can be converted to <em>FASTQ</em>-format.</p>
<p>After downloading a <strong>MD5-check</strong> is strongly encouraged.</p>
<div id="preparing-fastq-files" class="section level2">
<h2 class="hasAnchor">
<a href="#preparing-fastq-files" class="anchor"></a>Preparing FASTQ-files</h2>
<p>For this vignette, it is assumed that the <em>FASTQ</em>-files mentioned above have been downloaded to a directory called <code>samples</code>.</p>
<p>After downloading the <em>FASTQ</em>-files, we first run a basic quality control step to assess if there might be some quality issues or left over adapter sequences. We will use the tools <code>FastQC</code> and <code>MultiQC</code> for this task with the following <em>bash</em> commands:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash"><code class="sourceCode bash"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="bu">cd</span> <span class="st">'YOUR_PATH'</span>/samples</a>
<a class="sourceLine" id="cb1-2" data-line-number="2"></a>
<a class="sourceLine" id="cb1-3" data-line-number="3"><span class="co">#rename files</span></a>
<a class="sourceLine" id="cb1-4" data-line-number="4"><span class="fu">mv</span> SRR10669441_1.fastq.gz bulk_EtOH_rep1_1.fastq.gz</a>
<a class="sourceLine" id="cb1-5" data-line-number="5"><span class="fu">mv</span> SRR10669441_2.fastq.gz bulk_EtOH_rep1_2.fastq.gz</a>
<a class="sourceLine" id="cb1-6" data-line-number="6"><span class="fu">mv</span> SRR10669442_1.fastq.gz bulk_EtOH_rep2_1.fastq.gz</a>
<a class="sourceLine" id="cb1-7" data-line-number="7"><span class="fu">mv</span> SRR10669442_2.fastq.gz bulk_EtOH_rep2_2.fastq.gz</a>
<a class="sourceLine" id="cb1-8" data-line-number="8"><span class="fu">mv</span> SRR10669443_1.fastq.gz bulk_EtOH_rep3_1.fastq.gz</a>
<a class="sourceLine" id="cb1-9" data-line-number="9"><span class="fu">mv</span> SRR10669443_2.fastq.gz bulk_EtOH_rep3_2.fastq.gz</a>
<a class="sourceLine" id="cb1-10" data-line-number="10"><span class="fu">mv</span> SRR10669447_1.fastq.gz bulk_D2hr_rep1_1.fastq.gz</a>
<a class="sourceLine" id="cb1-11" data-line-number="11"><span class="fu">mv</span> SRR10669447_2.fastq.gz bulk_D2hr_rep1_2.fastq.gz</a>
<a class="sourceLine" id="cb1-12" data-line-number="12"><span class="fu">mv</span> SRR10669448_1.fastq.gz bulk_D2hr_rep2_1.fastq.gz</a>
<a class="sourceLine" id="cb1-13" data-line-number="13"><span class="fu">mv</span> SRR10669448_2.fastq.gz bulk_D2hr_rep2_2.fastq.gz</a>
<a class="sourceLine" id="cb1-14" data-line-number="14"><span class="fu">mv</span> SRR10669449_1.fastq.gz bulk_D2hr_rep3_1.fastq.gz</a>
<a class="sourceLine" id="cb1-15" data-line-number="15"><span class="fu">mv</span> SRR10669449_2.fastq.gz bulk_D2hr_rep3_2.fastq.gz</a>
<a class="sourceLine" id="cb1-16" data-line-number="16">    </a>
<a class="sourceLine" id="cb1-17" data-line-number="17"><span class="co">#Quality control</span></a>
<a class="sourceLine" id="cb1-18" data-line-number="18"><span class="ex">fastqc</span> -t 4 -o fastqc/ *.fastq.gz</a>
<a class="sourceLine" id="cb1-19" data-line-number="19"></a>
<a class="sourceLine" id="cb1-20" data-line-number="20"><span class="co">#summarize reports</span></a>
<a class="sourceLine" id="cb1-21" data-line-number="21"><span class="ex">multiqc</span> -o multiqc ./fastqc/</a></code></pre></div>
<p>We calculated a <code>FastQC</code> report for every file and summarized the 12 reports with <code>MultiQC</code>. The reports show minor leftovers of an Illumina sequencing adapter in the samples. This makes sense, as the samples were sequenced by an Illumina NovaSeq 6000. Although the adapter content is not dramatic, we will trim the reads to remove those adapters and perform a very relaxed quality filtering. We will use the tool <code>trim-galore</code> (Version 0.6.4_dev) for this task.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash"><code class="sourceCode bash"><a class="sourceLine" id="cb2-1" data-line-number="1"><span class="ex">trim_galore</span> --quality 20  --illumina --length 20 --trim-n --cores 4 -o ../trim/ --paired bulk_*.fastq.gz</a></code></pre></div>
<p>The trimmed files are the newly created folder <code>/trim</code>. Less than 1% of the total base pairs have been removed by the trimming process. Optionally one can again perform a <code>FastQC</code> and <code>MultiQC</code> analysis to confirm that the adapter sequences have been removed.</p>
</div>
<div id="transcript-level-quantification" class="section level2">
<h2 class="hasAnchor">
<a href="#transcript-level-quantification" class="anchor"></a>Transcript level quantification</h2>
<p>The files are now ready for transcript level quantification. We will use the tool <code>Salmon</code> (Version 1.1.0) for this.</p>
<p><code>Salmon</code> does not perform a standard genomic alignment, but performs a quasi-mapping directly to the transcriptome. Reads, that could be originating from multiple transcript isoforms, are assigned to equivalence classes, where actual counts are derived by expectation maximization algorithms.</p>
<p>To run the <code>Salmon</code> quantification, we first need a transcriptomic index. For this example, we will use the <a href="ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_34/gencode.v34.transcripts.fa.gz">transcript sequences</a> and <a href="ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_34/gencode.v34.annotation.gtf.gz">comprehensive gene annotation</a> from <a href="https://www.gencodegenes.org/">Gencode</a> <a href="https://www.gencodegenes.org/human/release_34.html">Version 34</a>.</p>
<p>Assume you have downloaded and unpacked the above mentioned files in the analysis root directory. To build the actual index we run this <em>bash</em> code:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash"><code class="sourceCode bash"><a class="sourceLine" id="cb3-1" data-line-number="1"><span class="co">#installation in conda evironment</span></a>
<a class="sourceLine" id="cb3-2" data-line-number="2"><span class="ex">conda</span> install salmon</a>
<a class="sourceLine" id="cb3-3" data-line-number="3"></a>
<a class="sourceLine" id="cb3-4" data-line-number="4"><span class="co">#create folder and change directory</span></a>
<a class="sourceLine" id="cb3-5" data-line-number="5"><span class="fu">mkdir</span> ../salmon</a>
<a class="sourceLine" id="cb3-6" data-line-number="6"><span class="bu">cd</span> ../salmon</a>
<a class="sourceLine" id="cb3-7" data-line-number="7"></a>
<a class="sourceLine" id="cb3-8" data-line-number="8"><span class="co">#build index - `p` specifies the number of threads to use</span></a>
<a class="sourceLine" id="cb3-9" data-line-number="9"><span class="ex">salmon</span> index -t ../gencode.v34.transcripts.fa --gencode -p 4 -i index</a></code></pre></div>
<p>The actual quantification can be performed in <em>bash</em> with:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode bash"><code class="sourceCode bash"><a class="sourceLine" id="cb4-1" data-line-number="1"><span class="bu">cd</span> ../trim</a>
<a class="sourceLine" id="cb4-2" data-line-number="2"></a>
<a class="sourceLine" id="cb4-3" data-line-number="3"><span class="co">#loop over files</span></a>
<a class="sourceLine" id="cb4-4" data-line-number="4"><span class="kw">for</span> <span class="ex">i</span> in bulk_*_1.fq.gz<span class="kw">;</span> <span class="kw">do</span> <span class="ex">salmon</span> quant -l A -i ../salmon/index/ -1 <span class="va">$i</span> -2 <span class="va">${i/</span>_1_val_1<span class="va">/</span>_2_val_2<span class="va">}</span> --seqBias --gcBias --validateMappings --rangeFactorizationBins 4 --threads 4 -o ../salmon/<span class="va">${i/</span>_1_val_1.fq.gz<span class="va">/}</span><span class="kw">;</span> <span class="kw">done</span></a></code></pre></div>
<hr>
<p><strong>This concludes the preprocessing of the data. Please see the corresponding analysis vignette for an usage example of DTUrtle.</strong></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Tobias Tekath.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
