<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Analysis of the Hoffman et al. human single-cell RNA-seq data • DTUrtle</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Analysis of the Hoffman et al. human single-cell RNA-seq data">
<meta property="og:description" content="DTUrtle">
<meta property="og:image" content="https://tobitekath.github.io/DTUrtle/logo.svg">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">DTUrtle</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.2.6</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Functions and data</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Hoffman_human_single-cell_preprocess.html">Human single-cell RNA-seq pre-processing</a>
    </li>
    <li>
      <a href="../articles/Hoffman_human_single-cell_analysis.html">Human single-cell RNA-seq analysis</a>
    </li>
    <li>
      <a href="../articles/Hoffman_human_bulk_preprocess.html">Human bulk RNA-seq pre-processing</a>
    </li>
    <li>
      <a href="../articles/Hoffman_human_bulk_analysis.html">Human bulk RNA-seq analysis</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/TobiTekath/DTUrtle/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="Hoffman_human_single-cell_analysis_files/htmlwidgets-1.5.1/htmlwidgets.js"></script><script src="Hoffman_human_single-cell_analysis_files/jquery-1.12.4/jquery.min.js"></script><link href="Hoffman_human_single-cell_analysis_files/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet">
<script src="Hoffman_human_single-cell_analysis_files/datatables-binding-0.13/datatables.js"></script><link href="Hoffman_human_single-cell_analysis_files/dt-core-1.10.20/css/jquery.dataTables.min.css" rel="stylesheet">
<link href="Hoffman_human_single-cell_analysis_files/dt-core-1.10.20/css/jquery.dataTables.extra.css" rel="stylesheet">
<script src="Hoffman_human_single-cell_analysis_files/dt-core-1.10.20/js/jquery.dataTables.min.js"></script><script src="Hoffman_human_single-cell_analysis_files/jszip-1.10.20/jszip.min.js"></script><link href="Hoffman_human_single-cell_analysis_files/dt-ext-buttons-1.10.20/css/buttons.dataTables.min.css" rel="stylesheet">
<script src="Hoffman_human_single-cell_analysis_files/dt-ext-buttons-1.10.20/js/dataTables.buttons.min.js"></script><script src="Hoffman_human_single-cell_analysis_files/dt-ext-buttons-1.10.20/js/buttons.flash.min.js"></script><script src="Hoffman_human_single-cell_analysis_files/dt-ext-buttons-1.10.20/js/buttons.html5.min.js"></script><script src="Hoffman_human_single-cell_analysis_files/dt-ext-buttons-1.10.20/js/buttons.colVis.min.js"></script><script src="Hoffman_human_single-cell_analysis_files/dt-ext-buttons-1.10.20/js/buttons.print.min.js"></script><link href="Hoffman_human_single-cell_analysis_files/nouislider-7.0.10/jquery.nouislider.min.css" rel="stylesheet">
<script src="Hoffman_human_single-cell_analysis_files/nouislider-7.0.10/jquery.nouislider.min.js"></script><link href="Hoffman_human_single-cell_analysis_files/selectize-0.12.0/selectize.bootstrap3.css" rel="stylesheet">
<script src="Hoffman_human_single-cell_analysis_files/selectize-0.12.0/selectize.min.js"></script><link href="Hoffman_human_single-cell_analysis_files/crosstalk-1.1.0.1/css/crosstalk.css" rel="stylesheet">
<script src="Hoffman_human_single-cell_analysis_files/crosstalk-1.1.0.1/js/crosstalk.min.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Analysis of the Hoffman et al. human single-cell RNA-seq data</h1>
                        <h4 class="author">Tobias Tekath</h4>
            
            <h4 class="date">2020-08-03</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/TobiTekath/DTUrtle/blob/master/vignettes/Hoffman_human_single-cell_analysis.Rmd"><code>vignettes/Hoffman_human_single-cell_analysis.Rmd</code></a></small>
      <div class="hidden name"><code>Hoffman_human_single-cell_analysis.Rmd</code></div>

    </div>

    
    
<p>This vignette exemplifies the analysis of <strong>single-cell RNA-seq data with DTUrtle</strong>. The data used in this vignette is publicly available as <em>Bioproject PRJNA594939</em> and the used <em>FASTQ</em>-files can be downloaded from <a href="https://www.ebi.ac.uk/ena/browser/view/PRJNA594939">here</a>. The corresponding publication from Hoffman et al. can be found <a href="https://doi.org/10.1038/s42003-020-0837-0">here</a>.</p>
<hr>
<p>The following code shows an example of an DTUrtle workflow. Assume we have performed the preprocessing as described <a href="https://tobitekath.github.io/DTUrtle/articles/Hoffman_human_single-cell_preprocess.html">here</a> and the R working directory is a newly created folder called <code>dtu_results</code>.</p>
<hr>
<div id="setup" class="section level2">
<h2 class="hasAnchor">
<a href="#setup" class="anchor"></a>Setup</h2>
<p>Load the DTUrtle package and set the BiocParallel parameter. It is recommended to perform computations in parallel, if possible.</p>
<div class="sourceCode" id="cb1"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">DTUrtle</span>)
<span class="co">#use up to 10 cores for computation</span>
<span class="no">biocpar</span> <span class="kw">&lt;-</span> <span class="kw pkg">BiocParallel</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/MulticoreParam-class.html">MulticoreParam</a></span>(<span class="fl">10</span>)</pre></body></html></div>
<hr>
</div>
<div id="import-and-format-data" class="section level2">
<h2 class="hasAnchor">
<a href="#import-and-format-data" class="anchor"></a>Import and format data</h2>
<p>We want to start by reading in our quantification counts, as well as a file specifying which transcript ID or name belongs to which gene ID or name.</p>
<div id="importing-and-processing-gtf-annotation-tx2gene" class="section level3">
<h3 class="hasAnchor">
<a href="#importing-and-processing-gtf-annotation-tx2gene" class="anchor"></a>Importing and processing GTF annotation (tx2gene)</h3>
<p>To get this transcript to gene (<code>tx2gene</code>) mapping, we will utilize the already present Gencode annotation file <code>gencode.v34.annotation.gtf</code>. The <code><a href="../reference/import_gtf.html">import_gtf()</a></code> function utilizes the a <code>rtracklayer</code> package and returns a transcript-level filtered version of the available data.</p>
<div class="sourceCode" id="cb2"><html><body><pre class="r"><span class="no">tx2gene</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/import_gtf.html">import_gtf</a></span>(<span class="kw">gtf_file</span> <span class="kw">=</span> <span class="st">"../gencode.v34.annotation.gtf"</span>)</pre></body></html></div>
<div class="sourceCode" id="cb3"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="no">tx2gene</span>, <span class="kw">n</span><span class="kw">=</span><span class="fl">3</span>)
<span class="co">#&gt;   seqnames start   end width strand source       type score phase</span>
<span class="co">#&gt; 1     chr1 11869 14409  2541      + HAVANA transcript    NA    NA</span>
<span class="co">#&gt; 2     chr1 12010 13670  1661      + HAVANA transcript    NA    NA</span>
<span class="co">#&gt; 3     chr1 14404 29570 15167      - HAVANA transcript    NA    NA</span>
<span class="co">#&gt;             gene_id                          gene_type gene_name level</span>
<span class="co">#&gt; 1 ENSG00000223972.5 transcribed_unprocessed_pseudogene   DDX11L1     2</span>
<span class="co">#&gt; 2 ENSG00000223972.5 transcribed_unprocessed_pseudogene   DDX11L1     2</span>
<span class="co">#&gt; 3 ENSG00000227232.5             unprocessed_pseudogene    WASH7P     2</span>
<span class="co">#&gt;      hgnc_id          havana_gene     transcript_id</span>
<span class="co">#&gt; 1 HGNC:37102 OTTHUMG00000000961.1 ENST00000456328.2</span>
<span class="co">#&gt; 2 HGNC:37102 OTTHUMG00000000961.1 ENST00000450305.2</span>
<span class="co">#&gt; 3 HGNC:38034 OTTHUMG00000000958.1 ENST00000488147.1</span>
<span class="co">#&gt;                      transcript_type transcript_name transcript_support_level</span>
<span class="co">#&gt; 1               processed_transcript     DDX11L1-202                        1</span>
<span class="co">#&gt; 2 transcribed_unprocessed_pseudogene     DDX11L1-201                       NA</span>
<span class="co">#&gt; 3             unprocessed_pseudogene      WASH7P-201                       NA</span>
<span class="co">#&gt;     tag    havana_transcript exon_number exon_id         ont protein_id ccdsid</span>
<span class="co">#&gt; 1 basic OTTHUMT00000362751.1        &lt;NA&gt;    &lt;NA&gt;        &lt;NA&gt;       &lt;NA&gt;   &lt;NA&gt;</span>
<span class="co">#&gt; 2 basic OTTHUMT00000002844.1        &lt;NA&gt;    &lt;NA&gt; PGO:0000019       &lt;NA&gt;   &lt;NA&gt;</span>
<span class="co">#&gt; 3 basic OTTHUMT00000002839.1        &lt;NA&gt;    &lt;NA&gt; PGO:0000005       &lt;NA&gt;   &lt;NA&gt;</span></pre></body></html></div>
<p>There are a lot of columns present in the data frame, but at the moment we are mainly interested in the columns <code>gene_id</code>, <code>gene_name</code>, <code>transcript_id</code> and <code>transcript_name</code>.</p>
<hr>
<p>As we want to use gene and transcript names as the main identifiers in our analysis (so we can directly say: Gene x is differential), we should ensure that each gene / transcript name maps only to a single gene / transcript id.</p>
<p>For this we can use the DTUrtle function <code><a href="../reference/one_to_one_mapping.html">one_to_one_mapping()</a></code>, which checks if there are identifiers, which relate to the same name. If this is the case, the names (<strong>not</strong> the identifiers) are slightly altered by appending a number. If <em>id_x</em> and <em>id_y</em> both have the name <em>ABC</em>, the <em>id_y</em> name is altered to <em>ABC_2</em> by default.</p>
<div class="sourceCode" id="cb4"><html><body><pre class="r"><span class="no">tx2gene</span>$<span class="no">gene_name</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/one_to_one_mapping.html">one_to_one_mapping</a></span>(<span class="kw">name</span> <span class="kw">=</span> <span class="no">tx2gene</span>$<span class="no">gene_name</span>, <span class="kw">id</span> <span class="kw">=</span> <span class="no">tx2gene</span>$<span class="no">gene_id</span>)
<span class="co">#&gt; Changed 78 names.</span>
<span class="no">tx2gene</span>$<span class="no">transcript_name</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/one_to_one_mapping.html">one_to_one_mapping</a></span>(<span class="kw">name</span> <span class="kw">=</span> <span class="no">tx2gene</span>$<span class="no">transcript_name</span>, <span class="kw">id</span> <span class="kw">=</span> <span class="no">tx2gene</span>$<span class="no">transcript_id</span>)
<span class="co">#&gt; Changed 161 names.</span></pre></body></html></div>
<p>We see that it was a good idea to ensure the one to one mapping, as many doublets have been corrected.</p>
<hr>
<p>For the <code><a href="../reference/run_drimseq.html">run_drimseq()</a></code> <code>tx2gene</code> parameter, we need a data frame, where the first column specifies the transcript identifiers and the second column specifying the corresponding gene names. Rather than subsetting the data frame, a column reordering is proposed, so that additional data can still be used in further steps. DTUrtle makes sure to carry over additional data columns in the analysis steps. To reorder the columns of our tx2gene data frame, we can utilize the <code><a href="../reference/move_columns_to_front.html">move_columns_to_front()</a></code> functionality.</p>
<div class="sourceCode" id="cb5"><html><body><pre class="r"><span class="no">tx2gene</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/move_columns_to_front.html">move_columns_to_front</a></span>(<span class="kw">df</span> <span class="kw">=</span> <span class="no">tx2gene</span>, <span class="kw">columns</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"transcript_name"</span>, <span class="st">"gene_name"</span>))</pre></body></html></div>
<div class="sourceCode" id="cb6"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="no">tx2gene</span>, <span class="kw">n</span><span class="kw">=</span><span class="fl">5</span>)
<span class="co">#&gt;   transcript_name   gene_name seqnames start   end width strand  source</span>
<span class="co">#&gt; 1     DDX11L1-202     DDX11L1     chr1 11869 14409  2541      +  HAVANA</span>
<span class="co">#&gt; 2     DDX11L1-201     DDX11L1     chr1 12010 13670  1661      +  HAVANA</span>
<span class="co">#&gt; 3      WASH7P-201      WASH7P     chr1 14404 29570 15167      -  HAVANA</span>
<span class="co">#&gt; 4   MIR6859-1-201   MIR6859-1     chr1 17369 17436    68      - ENSEMBL</span>
<span class="co">#&gt; 5 MIR1302-2HG-202 MIR1302-2HG     chr1 29554 31097  1544      +  HAVANA</span>
<span class="co">#&gt;         type score phase           gene_id                          gene_type</span>
<span class="co">#&gt; 1 transcript    NA    NA ENSG00000223972.5 transcribed_unprocessed_pseudogene</span>
<span class="co">#&gt; 2 transcript    NA    NA ENSG00000223972.5 transcribed_unprocessed_pseudogene</span>
<span class="co">#&gt; 3 transcript    NA    NA ENSG00000227232.5             unprocessed_pseudogene</span>
<span class="co">#&gt; 4 transcript    NA    NA ENSG00000278267.1                              miRNA</span>
<span class="co">#&gt; 5 transcript    NA    NA ENSG00000243485.5                             lncRNA</span>
<span class="co">#&gt;   level    hgnc_id          havana_gene     transcript_id</span>
<span class="co">#&gt; 1     2 HGNC:37102 OTTHUMG00000000961.1 ENST00000456328.2</span>
<span class="co">#&gt; 2     2 HGNC:37102 OTTHUMG00000000961.1 ENST00000450305.2</span>
<span class="co">#&gt; 3     2 HGNC:38034 OTTHUMG00000000958.1 ENST00000488147.1</span>
<span class="co">#&gt; 4     3 HGNC:50039                 &lt;NA&gt; ENST00000619216.1</span>
<span class="co">#&gt; 5     2 HGNC:52482 OTTHUMG00000000959.1 ENST00000473358.1</span>
<span class="co">#&gt;                      transcript_type transcript_support_level   tag</span>
<span class="co">#&gt; 1               processed_transcript                        1 basic</span>
<span class="co">#&gt; 2 transcribed_unprocessed_pseudogene                       NA basic</span>
<span class="co">#&gt; 3             unprocessed_pseudogene                       NA basic</span>
<span class="co">#&gt; 4                              miRNA                       NA basic</span>
<span class="co">#&gt; 5                             lncRNA                        5 basic</span>
<span class="co">#&gt;      havana_transcript exon_number exon_id         ont protein_id ccdsid</span>
<span class="co">#&gt; 1 OTTHUMT00000362751.1        &lt;NA&gt;    &lt;NA&gt;        &lt;NA&gt;       &lt;NA&gt;   &lt;NA&gt;</span>
<span class="co">#&gt; 2 OTTHUMT00000002844.1        &lt;NA&gt;    &lt;NA&gt; PGO:0000019       &lt;NA&gt;   &lt;NA&gt;</span>
<span class="co">#&gt; 3 OTTHUMT00000002839.1        &lt;NA&gt;    &lt;NA&gt; PGO:0000005       &lt;NA&gt;   &lt;NA&gt;</span>
<span class="co">#&gt; 4                 &lt;NA&gt;        &lt;NA&gt;    &lt;NA&gt;        &lt;NA&gt;       &lt;NA&gt;   &lt;NA&gt;</span>
<span class="co">#&gt; 5 OTTHUMT00000002840.1        &lt;NA&gt;    &lt;NA&gt;        &lt;NA&gt;       &lt;NA&gt;   &lt;NA&gt;</span></pre></body></html></div>
<p>This concludes the tx2gene formatting.</p>
<hr>
</div>
<div id="reading-in-quantification-data" class="section level3">
<h3 class="hasAnchor">
<a href="#reading-in-quantification-data" class="anchor"></a>Reading in quantification data</h3>
<p>The read-in of the quantification counts can be achieved with <code><a href="../reference/import_counts.html">import_counts()</a></code>, that uses the <code>tximport</code> package in the background. This function is able to parse the output of many different quantification tools. Advanced users might be able to tune parameters to parse arbitrary output files from currently not supported tools.</p>
<p>In the pre-processing vignette we quantified the counts with <code>Alevin</code>. The folder structure of the quantification results folder looks like this:</p>
<div class="sourceCode" id="cb7"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/list.files.html">list.files</a></span>(<span class="st">"../alevin/"</span>)
<span class="co">#&gt; [1] "index"     "sc_Dex2hr" "sc_EtOH"   "txmap.tsv"</span></pre></body></html></div>
<p>We will create a named files vector, pointing to the <code>quants_mat.gz</code> file for each sample. The names help to differentiate the samples later on.</p>
<div class="sourceCode" id="cb8"><html><body><pre class="r"><span class="no">files</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Sys.glob.html">Sys.glob</a></span>(<span class="st">"../alevin/sc_*/alevin/quants_mat.gz"</span>)
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="no">files</span>) <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"sc_Dex2hr"</span>, <span class="st">"sc_EtOH"</span>)</pre></body></html></div>
<p>The files object looks like this:</p>
<pre><code>#&gt;                                  sc_Dex2hr 
#&gt; "../alevin/sc_Dex2hr/alevin/quants_mat.gz" 
#&gt;                                    sc_EtOH 
#&gt;   "../alevin/sc_EtOH/alevin/quants_mat.gz"</code></pre>
<p>The actual import will be performed with <code><a href="../reference/import_counts.html">import_counts()</a></code>.</p>
<div class="sourceCode" id="cb10"><html><body><pre class="r"><span class="no">cts_list</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/import_counts.html">import_counts</a></span>(<span class="kw">files</span> <span class="kw">=</span> <span class="no">files</span>, <span class="kw">type</span> <span class="kw">=</span> <span class="st">"alevin"</span>)
<span class="co">#&gt; Reading in 2 alevin runs.</span>
<span class="co">#&gt; reading in alevin gene-level counts across cells with fishpond</span>
<span class="co">#&gt; reading in alevin gene-level counts across cells with fishpond</span></pre></body></html></div>
<p>The <code>cts_list</code> object is a named list, with a sparse Matrix per sample. In single-cell data, each sample normally consists of many different cells with an unique cell barcode. These cell barcodes might overlap between samples though. For this reason, many single-cell workflow use a cell barcode extension, uniquely assigning each cell to a sample. This can also be done in DTUrtle with <code><a href="../reference/combine_to_matrix.html">combine_to_matrix()</a></code>, which is only applicable if you are analyzing single-cell data.</p>
<p>This function will make sure that there are no duplicated barcodes between you samples, before merging the matrices together. If there are duplicated barcodes, a cell extension is added. Additionally, all not expressed features are removed to reduce the size of the data.</p>
<div class="sourceCode" id="cb11"><html><body><pre class="r"><span class="no">cts</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/combine_to_matrix.html">combine_to_matrix</a></span>(<span class="kw">tx_list</span> <span class="kw">=</span> <span class="no">cts_list</span>)
<span class="co">#&gt; Merging matrices</span>
<span class="co">#&gt; Excluding 149854 overall not expressed features.</span>
<span class="co">#&gt; 77347 features left.</span></pre></body></html></div>
<p>Apparently there were no duplicated cell barcodes between our samples, so a cell barcode extension is not necessary. Optionally you could still force the addition by specifying a vector of <code>cell_extensions</code>.</p>
<div class="sourceCode" id="cb12"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span>(<span class="no">cts</span>)
<span class="co">#&gt; [1] 77347   796</span></pre></body></html></div>
<p>There are ~77k features left for 796 cells.</p>
<hr>
</div>
<div id="sample-metadata" class="section level3">
<h3 class="hasAnchor">
<a href="#sample-metadata" class="anchor"></a>Sample metadata</h3>
<p>Finally, we need a sample metadata data frame, specifying which sample belongs to which comparison group. This table is also convenient to store and carry over additional metadata.</p>
<p>For single-cell data, this is not a sample metadata data frame, but a cell metadata data frame. We have to specify the information on cell level, with the barcodes as identifiers.</p>
<p>If such a table is not already present, it can be easily prepared:</p>
<div class="sourceCode" id="cb13"><html><body><pre class="r"><span class="no">pd</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(<span class="st">"id"</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="no">cts</span>),
                 <span class="st">"group"</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="no">cts</span>) <span class="kw">%in%</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="no">cts_list</span>$<span class="no">sc_Dex2hr</span>),<span class="st">"Dex2hr"</span>,<span class="st">"EtOH"</span>),
                 <span class="kw">stringsAsFactors</span> <span class="kw">=</span> <span class="no">F</span>)</pre></body></html></div>
<div class="sourceCode" id="cb14"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="no">pd</span>, <span class="kw">n</span><span class="kw">=</span><span class="fl">5</span>)
<span class="co">#&gt;                   id  group</span>
<span class="co">#&gt; 1 GCCAGAGCGAATGTTAAC Dex2hr</span>
<span class="co">#&gt; 2 TGAATTCAGACTAGTCTG Dex2hr</span>
<span class="co">#&gt; 3 CGGCGTCTTACGTTCTTG Dex2hr</span>
<span class="co">#&gt; 4 TCGGGAGGACGATACCGA Dex2hr</span>
<span class="co">#&gt; 5 TCGGGACCTCTAGAGTGA Dex2hr</span></pre></body></html></div>
<hr>
</div>
</div>
<div id="dtu-analysis" class="section level2">
<h2 class="hasAnchor">
<a href="#dtu-analysis" class="anchor"></a>DTU analysis</h2>
<p>We have prepared all necessary data to perform the differentially transcript usage (DTU) analysis. DTUrtle only needs two simple commands to perform it. Please be aware that these steps are the most compute intensive and, depending on your data, might take some time to complete. It is recommended to parallelize the computations with the <code>BBPARAM</code> parameter, if applicable.</p>
<p>First, we want to set-up and perform the statistical analysis with DRIMSeq, a DTU specialized statistical framework utilizing a Dirichlet-multinomial model. This can be done with the <code><a href="../reference/run_drimseq.html">run_drimseq()</a></code> command. We use the previously imported data as parameters, specifying which column in the cell metadata data frame contains ids and which the group information we want. We should also specify which of the groups should be compared (if there are more than two) and in which order. The order given in the <code>cond_levels</code> parameter also specifies the comparison formula.</p>
<blockquote>
<div class="alert alert-info">
<p><strong>Note:</strong> By default <code><a href="../reference/run_drimseq.html">run_drimseq()</a></code> converts sparse count matrix to a dense format for statistical computations (<code>force_dense=TRUE</code>). While this increases memory usage, it currently also reduces the run time. The computations can be performed keeping the sparse counts by setting <code>force_dense=FALSE</code>.</p>
</div>
</blockquote>
<div class="sourceCode" id="cb15"><html><body><pre class="r">dturtle &lt;- run_drimseq(counts = cts, tx2gene = tx2gene, pd=pd, id_col = "id",
                    cond_col = "group", cond_levels = c("Dex2hr", "EtOH"), filtering_strategy = "sc", 
                    BPPARAM = biocpar)
#&gt; Using tx2gene columns:
#&gt;  transcript_name ---&gt; 'feature_id'
#&gt;  gene_name ---&gt; 'gene_id'
#&gt; 
#&gt; Comparing Dex2hr vs EtOH
#&gt; 
#&gt; Proceed with cells/samples: 
#&gt; Dex2hr   EtOH 
#&gt;    399    397
#&gt; 
#&gt; Filtering...
#&gt; 
  |                                                                            
  |                                                                      |   0%
  |                                                                            
  |=======                                                               |  10%
  |                                                                            
  |==============                                                        |  20%
  |                                                                            
  |=====================                                                 |  30%
  |                                                                            
  |============================                                          |  40%
  |                                                                            
  |===================================                                   |  50%
  |                                                                            
  |==========================================                            |  60%
  |                                                                            
  |=================================================                     |  70%
  |                                                                            
  |========================================================              |  80%
  |                                                                            
  |===============================================================       |  90%
  |                                                                            
  |======================================================================| 100%
#&gt; Retain 21452 of 77347 features.
#&gt; Removed 55895 features.
#&gt; 
#&gt; Performing statistical tests...
#&gt; * Calculating mean gene expression..
#&gt; Took 0.7417 seconds.
#&gt; * Estimating common precision..
#&gt; Took 2.012 seconds.
#&gt; ! Using common_precision = 993.3516 as prec_init !
#&gt; * Estimating genewise precision..
#&gt; ! Using loess fit as a shrinkage factor !
#&gt; Took 8.0867 seconds.
#&gt; * Fitting the DM model..
#&gt;    Using the one way approach.
#&gt; Took 10.4332 seconds.
#&gt; * Fitting the BB model..
#&gt;    Using the one way approach.
#&gt; Took 2.9904 seconds.
#&gt; * Fitting the DM model..
#&gt;    Using the one way approach.
#&gt; Took 8.3371 seconds.
#&gt; * Calculating likelihood ratio statistics..
#&gt; Took 0.0026 seconds.
#&gt; * Fitting the BB model..
#&gt;    Using the one way approach.
#&gt; Took 3.4831 seconds.
#&gt; * Calculating likelihood ratio statistics..
#&gt; Took 0.0099 seconds.</pre></body></html></div>
<p>As in all statistical procedures, it is of favor to perform as few tests as possible but as much tests as necessary, to maintain a high statistical power. This is achieved by filtering the data to remove inherently uninteresting items, for example very lowly expressed genes or features. DTUrtle includes a powerful and customizable filtering functionality for this task, which is an optimized version of the <code>dmFilter()</code> function of the DRIMSeq package.</p>
<p>Above we used a predefined filtering strategy for single-cell data, requiring that features contribute at least 5% of the total expression in at least 5% of the cells of the smallest group. Additionally, all genes are filtered, which only have a single transcript left, as they can not be analyzed in DTU analysis. The filtering options can be extended or altered by the user.</p>
<div class="sourceCode" id="cb16"><html><body><pre class="r"><span class="no">dturtle</span>$<span class="no">used_filtering_options</span>
<span class="co">#&gt; $DRIM</span>
<span class="co">#&gt; $DRIM$min_samps_gene_expr</span>
<span class="co">#&gt; [1] 0</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $DRIM$min_samps_feature_expr</span>
<span class="co">#&gt; [1] 0</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $DRIM$min_samps_feature_prop</span>
<span class="co">#&gt; [1] 19.85</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $DRIM$min_gene_expr</span>
<span class="co">#&gt; [1] 0</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $DRIM$min_feature_expr</span>
<span class="co">#&gt; [1] 0</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $DRIM$min_feature_prop</span>
<span class="co">#&gt; [1] 0.05</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $DRIM$run_gene_twice</span>
<span class="co">#&gt; [1] TRUE</span></pre></body></html></div>
<p>The resulting <code>dturtle</code> object will be used as our main results object, storing all necessary and interesting data of the analysis. It is a simple and easy-accessible list, which can be easily extended / altered by the user. By default three different meta data tables are generated: - <code>meta_table_gene</code>: Contains gene level meta data. - <code>meta_table_tx</code>: Contains transcript level meta data. - <code>meta_table_sample</code>: Contains sample level meta data (as the created <code>pd</code> data frame)</p>
<p>These meta data tables are used in for visualization purposes and can be extended by the user.</p>
<div class="sourceCode" id="cb17"><html><body><pre class="r"><span class="no">dturtle</span>$<span class="no">meta_table_gene</span>[<span class="fl">1</span>:<span class="fl">5</span>,<span class="fl">1</span>:<span class="fl">5</span>]
<span class="co">#&gt;        gene     exp_in exp_in_Dex2hr exp_in_EtOH seqnames</span>
<span class="co">#&gt; A1BG   A1BG 0.03768844    0.04511278  0.03022670    chr19</span>
<span class="co">#&gt; AAAS   AAAS 0.04020101    0.04511278  0.03526448    chr12</span>
<span class="co">#&gt; AAGAB AAGAB 0.07286432    0.09273183  0.05289673    chr15</span>
<span class="co">#&gt; AAMDC AAMDC 0.03894472    0.03508772  0.04282116    chr11</span>
<span class="co">#&gt; AAMP   AAMP 0.09422111    0.09774436  0.09068010     chr2</span></pre></body></html></div>
<hr>
<p>As proposed in <a href="https://doi.org/10.12688/f1000research.15398.3">Love et al. (2018)</a>, we will use a two-stage statistical testing procedure together with a post-hoc filtering on the standard deviations in proportions (<code><a href="../reference/posthoc_and_stager.html">posthoc_and_stager()</a></code>). We will use <em>stageR</em> to determine genes, that show a overall significant change in transcript proportions. For these significant genes, we will try to pinpoint specific transcripts, which significantly drive this overall change. As a result, we will have a list of <code>significant genes</code> (genes showing the overall change) and a list of <code>significant transcripts</code> (one or more transcripts of the <code>significant genes</code>). Please note, that not every <code>significant gene</code> does have one or more <code>significant trasncripts</code>. It is not always possible to attribute the overall change in proportions to single transcripts. These two list of significant items are computed and corrected against a <strong>overall false discovery rate (OFDR)</strong>.</p>
<p>Additionally, we will apply a post-hoc filtering scheme to improve the targeted OFDR control level. The filtering strategy will discard transcripts, which standard deviation of the proportion per cell/sample is below the specified threshold. For example by setting <code>posthoc=0.1</code>, we will exclude all transcripts, which proportional expression (in regard to the total gene expression) deviates by less than 0.1 between cells. This filtering step should mostly exclude ‘uninteresting’ transcripts, which would not have been called as significant either way.</p>
<div class="sourceCode" id="cb18"><html><body><pre class="r"><span class="no">dturtle</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/posthoc_and_stager.html">posthoc_and_stager</a></span>(<span class="kw">dturtle</span> <span class="kw">=</span> <span class="no">dturtle</span>, <span class="kw">ofdr</span> <span class="kw">=</span> <span class="fl">0.05</span>, <span class="kw">posthoc</span> <span class="kw">=</span> <span class="fl">0.1</span>)
<span class="co">#&gt; Posthoc filtered 20608 features.</span>
<span class="co">#&gt; The returned adjusted p-values are based on a stage-wise testing approach and are only valid for the provided target OFDR level of 5%. If a different target OFDR level is of interest,the entire adjustment should be re-run.</span>
<span class="co">#&gt; Found 32 significant genes with 37 significant transcripts (OFDR: 0.05)</span></pre></body></html></div>
<p>The <code>dturtle</code> object now contains additional elements, including the lists of <code>significant genes</code> and <code>significant transcripts</code>.</p>
<div class="sourceCode" id="cb19"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="no">dturtle</span>$<span class="no">sig_gene</span>)
<span class="co">#&gt; [1] "AGR2"    "ATP5F1B" "COX6A1"  "EEF1A1"  "HSPB1"   "NPM1"</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="no">dturtle</span>$<span class="no">sig_tx</span>)
<span class="co">#&gt;          AGR2          AGR2       ATP5F1B       ATP5F1B        COX6A1 </span>
<span class="co">#&gt;    "AGR2-203"    "AGR2-204" "ATP5F1B-204" "ATP5F1B-201"  "COX6A1-201" </span>
<span class="co">#&gt;        COX6A1 </span>
<span class="co">#&gt;  "COX6A1-202"</span></pre></body></html></div>
<hr>
</div>
<div id="result-aggregation-and-visualization" class="section level2">
<h2 class="hasAnchor">
<a href="#result-aggregation-and-visualization" class="anchor"></a>Result aggregation and visualization</h2>
<p>The <code>DTUrtle</code> package contains multiple visualization options, enabling a in-depth inspection.</p>
<div id="dtu-table-creation" class="section level3">
<h3 class="hasAnchor">
<a href="#dtu-table-creation" class="anchor"></a>DTU table creation</h3>
<p>We will start by aggregating the analysis results to a data frame with <code><a href="../reference/create_dtu_table.html">create_dtu_table()</a></code>. This function is highly flexible and allows aggregation of gene or transcript level metadata in various ways. By default some useful information are included in the <em>dtu table</em>, in this example we further specify to include the <code>seqnames</code> column of the gene level metadata (which contains chromosome information) as well as the maximal <em>expressed in</em> ratio from the transcript level metadata.</p>
<div class="sourceCode" id="cb20"><html><body><pre class="r"><span class="no">dturtle</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/create_dtu_table.html">create_dtu_table</a></span>(<span class="kw">dturtle</span> <span class="kw">=</span> <span class="no">dturtle</span>, <span class="kw">add_gene_metadata</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="st">"chromosome"</span><span class="kw">=</span><span class="st">"seqnames"</span>),
                               <span class="kw">add_tx_metadata</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="st">"tx_expr_in_max"</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"exp_in"</span>, <span class="no">max</span>)))</pre></body></html></div>
<div class="sourceCode" id="cb21"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="no">dturtle</span>$<span class="no">dtu_table</span>, <span class="kw">n</span><span class="kw">=</span><span class="fl">5</span>)
<span class="co">#&gt;          geneID    gene_qval  min_tx_qval n_tx n_sig_tx max(Dex2hr-EtOH)</span>
<span class="co">#&gt; PSMB4     PSMB4 2.384975e-03 7.737662e-04    7        1        0.1549467</span>
<span class="co">#&gt; RPL13A   RPL13A 1.562860e-12 4.164981e-14   12        2        0.1533086</span>
<span class="co">#&gt; RPLP1     RPLP1 8.761113e-15 1.821671e-16    4        3        0.1389187</span>
<span class="co">#&gt; COX6A1   COX6A1 2.352015e-02 4.328509e-03    3        2       -0.1365610</span>
<span class="co">#&gt; ATP5F1B ATP5F1B 4.299094e-05 1.178645e-06    5        2        0.1355355</span>
<span class="co">#&gt;         chromosome tx_expr_in_max</span>
<span class="co">#&gt; PSMB4         chr1      0.6306533</span>
<span class="co">#&gt; RPL13A       chr19      0.9183417</span>
<span class="co">#&gt; RPLP1        chr15      0.9623116</span>
<span class="co">#&gt; COX6A1       chr12      0.5778894</span>
<span class="co">#&gt; ATP5F1B      chr12      0.8128141</span></pre></body></html></div>
<p>This table is our basis for creating an <strong>interactive HTML-table</strong> of the results.</p>
<hr>
</div>
<div id="proportion-barplot" class="section level3">
<h3 class="hasAnchor">
<a href="#proportion-barplot" class="anchor"></a>Proportion barplot</h3>
<p>As a first visualization option we will create a barplot of the proportions of each transcript per sample. We can use the function <code><a href="../reference/plot_proportion_barplot.html">plot_proportion_barplot()</a></code> for this task, which also adds the mean proportion fit per subgroup to the plot (by default as a red line).</p>
<p>As an example, we will create the plot for the gene <em>PPIA</em>, which is one of the significant genes found in the analysis. We will optionally provide the gene_id for <em>PPIA</em>, which is stored in the column <code>gene_id.1</code> of <code>dturtle$meta_table_gene</code>.</p>
<div class="sourceCode" id="cb22"><html><body><pre class="r"><span class="no">temp</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/plot_proportion_barplot.html">plot_proportion_barplot</a></span>(<span class="kw">dturtle</span> <span class="kw">=</span> <span class="no">dturtle</span>, <span class="kw">genes</span> <span class="kw">=</span> <span class="st">"PPIA"</span>, <span class="kw">meta_gene_id</span> <span class="kw">=</span> <span class="st">"gene_id.1"</span>)
<span class="no">temp</span>$<span class="no">PPIA</span></pre></body></html></div>
<p><img src="Hoffman_human_single-cell_analysis_files/figure-html/unnamed-chunk-22-1.png" width="80%" style="display: block; margin: auto;"></p>
<p>We see, that most of the proportional differences for <em>PPIA</em> are driven by 2 of the 7 transcripts (<em>PPIA-204</em> and <em>PPIA-209</em>). These transcripts are also the significant transcripts found in the analysis (as they are marked in red). We additionally see, that a DGE analysis would most likely not report <em>PPIA</em> as differential, because the mean gene expression is quite alike between the groups (together with a common high coefficient of variation).</p>
<p>For the interactive HTML-table we would need to save the images to disk (in the to-be-created sub folder “images” of the current working directory). There is also a convenience option, to directly add the file paths to the <code>dtu_table</code>. As multiple plots are created, we can provide a <code>BiocParallel</code> object to speed up the creation. If no specific genes are provided, all significant genes will be plotted.</p>
<div class="sourceCode" id="cb23"><html><body><pre class="r">dturtle &lt;- plot_proportion_barplot(dturtle = dturtle, 
                                   meta_gene_id = "gene_id.1", 
                                   savepath = "images", 
                                   add_to_table = "barplot",
                                   BPPARAM = biocpar)
#&gt; 
  |                                                                            
  |                                                                      |   0%
  |                                                                            
  |=======                                                               |  10%
#&gt; 
#&gt; 
  |                                                                            
  |==============                                                        |  20%
#&gt; 
#&gt; 
  |                                                                            
  |=====================                                                 |  30%
#&gt; 
#&gt; 
  |                                                                            
  |============================                                          |  40%
#&gt; 
#&gt; 
  |                                                                            
  |===================================                                   |  50%
#&gt; 
#&gt; 
  |                                                                            
  |==========================================                            |  60%
#&gt; 
#&gt; 
  |                                                                            
  |=================================================                     |  70%
#&gt; 
#&gt; 
  |                                                                            
  |========================================================              |  80%
#&gt; 
#&gt; 
  |                                                                            
  |===============================================================       |  90%
#&gt; 
#&gt; 
  |                                                                            
  |======================================================================| 100%
#&gt; </pre></body></html></div>
<div class="sourceCode" id="cb24"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="no">dturtle</span>$<span class="no">dtu_table</span>$<span class="no">barplot</span>)
<span class="co">#&gt; [1] "images/PSMB4_barplot.png"   "images/RPL13A_barplot.png" </span>
<span class="co">#&gt; [3] "images/RPLP1_barplot.png"   "images/COX6A1_barplot.png" </span>
<span class="co">#&gt; [5] "images/ATP5F1B_barplot.png" "images/RPL7A_barplot.png"</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/list.files.html">list.files</a></span>(<span class="st">"./images/"</span>))
<span class="co">#&gt; [1] "AGR2_barplot.png"        "AGR2_pheatmap.png"      </span>
<span class="co">#&gt; [3] "AGR2_transcripts.png"    "ATP5F1B_barplot.png"    </span>
<span class="co">#&gt; [5] "ATP5F1B_pheatmap.png"    "ATP5F1B_transcripts.png"</span></pre></body></html></div>
<hr>
</div>
<div id="proportion-heatmap" class="section level3">
<h3 class="hasAnchor">
<a href="#proportion-heatmap" class="anchor"></a>Proportion heatmap</h3>
<p>A different visualization option is a heatmap, where additional meta data can be displayed alongside the transcript proportions (<code><a href="../reference/plot_proportion_pheatmap.html">plot_proportion_pheatmap()</a></code>). This visualization uses the <code>pheatmap</code> package, the user can specify any of the available parameters to customize the results.</p>
<div class="sourceCode" id="cb25"><html><body><pre class="r"><span class="no">temp</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/plot_proportion_pheatmap.html">plot_proportion_pheatmap</a></span>(<span class="kw">dturtle</span> <span class="kw">=</span> <span class="no">dturtle</span>, <span class="kw">genes</span> <span class="kw">=</span> <span class="st">"PPIA"</span>,
                                 <span class="kw">include_expression</span> <span class="kw">=</span> <span class="no">T</span>, <span class="kw">treeheight_col</span><span class="kw">=</span><span class="fl">20</span>)
<span class="no">temp</span>$<span class="no">PPIA</span></pre></body></html></div>
<p><img src="Hoffman_human_single-cell_analysis_files/figure-html/unnamed-chunk-25-1.png" width="80%" style="display: block; margin: auto;"></p>
<p>By default, row and column annotations are added. This plot helps to examine the transcript composition of groups of cells. We see, there is a subgroup of cells, that seem to almost only express both <em>PPIA-204</em> and <em>PPIA-209</em> at equal proportions. The same goes for a even smaller subgroup of cells, which express <em>PPIA-204</em> alongside with <em>PPIA-202</em>.</p>
<p>Again, we can save the plots to disk and add them to the <code>dtu_table</code>:</p>
<div class="sourceCode" id="cb26"><html><body><pre class="r">dturtle &lt;- plot_proportion_pheatmap(dturtle = dturtle,
                                    include_expression = T,
                                    treeheight_col=20,
                                    savepath = "images", 
                                    add_to_table = "pheatmap",
                                    BPPARAM = biocpar)
#&gt; 
  |                                                                            
  |                                                                      |   0%
  |                                                                            
  |=======                                                               |  10%
  |                                                                            
  |==============                                                        |  20%
  |                                                                            
  |=====================                                                 |  30%
  |                                                                            
  |============================                                          |  40%
  |                                                                            
  |===================================                                   |  50%
  |                                                                            
  |==========================================                            |  60%
  |                                                                            
  |=================================================                     |  70%
  |                                                                            
  |========================================================              |  80%
  |                                                                            
  |===============================================================       |  90%
  |                                                                            
  |======================================================================| 100%</pre></body></html></div>
<div class="sourceCode" id="cb27"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="no">dturtle</span>$<span class="no">dtu_table</span>$<span class="no">pheatmap</span>)
<span class="co">#&gt; [1] "images/PSMB4_pheatmap.png"   "images/RPL13A_pheatmap.png" </span>
<span class="co">#&gt; [3] "images/RPLP1_pheatmap.png"   "images/COX6A1_pheatmap.png" </span>
<span class="co">#&gt; [5] "images/ATP5F1B_pheatmap.png" "images/RPL7A_pheatmap.png"</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/list.files.html">list.files</a></span>(<span class="st">"./images/"</span>))
<span class="co">#&gt; [1] "AGR2_barplot.png"        "AGR2_pheatmap.png"      </span>
<span class="co">#&gt; [3] "AGR2_transcripts.png"    "ATP5F1B_barplot.png"    </span>
<span class="co">#&gt; [5] "ATP5F1B_pheatmap.png"    "ATP5F1B_transcripts.png"</span></pre></body></html></div>
<hr>
</div>
<div id="transcript-overview" class="section level3">
<h3 class="hasAnchor">
<a href="#transcript-overview" class="anchor"></a>Transcript overview</h3>
<p>Until now, we looked at the different transcripts as abstract entities. Alongside proportional differences, the actual difference in the exon-intron structure of transcripts is of great importance for many research questions. This structure can be visualized with the <code><a href="../reference/plot_transcripts_view.html">plot_transcripts_view()</a></code> functionality of <em>DTUrtle</em>.</p>
<p>This visualization is based on the <em>Gviz</em> package and needs a read-in or a path of a GTF file. In <a href="#import-and-format-data">Import and format data</a> we already imported a GTF file. This was subset to transcript-level (by the <code><a href="../reference/import_gtf.html">import_gtf()</a></code> function), thus this is <strong>not</strong> sufficient for the visualization. But of course we can reuse the actual GTF file, which should in general match with the one used for the <code>tx2gene</code> data frame.</p>
<p>As we have ensured the one_to_one mapping in <a href="#import-and-format-data">Import and format data</a> and potentially renamed some genes, we should specify the <code>one_to_one</code> parameter in this call.</p>
<div class="sourceCode" id="cb28"><html><body><pre class="r"><span class="fu"><a href="../reference/plot_transcripts_view.html">plot_transcripts_view</a></span>(<span class="kw">dturtle</span> <span class="kw">=</span> <span class="no">dturtle</span>,
                              <span class="kw">genes</span> <span class="kw">=</span> <span class="st">"PPIA"</span>,
                              <span class="kw">gtf</span> <span class="kw">=</span> <span class="st">"../gencode.v34.annotation.gtf"</span>,
                              <span class="kw">genome</span> <span class="kw">=</span> <span class="st">'hg38'</span>,
                              <span class="kw">one_to_one</span> <span class="kw">=</span> <span class="no">T</span>)
<span class="co">#&gt; </span>
<span class="co">#&gt; Importing gtf file from disk.</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Performing one to one mapping in gtf</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Found gtf GRanges for 1 of 1 provided genes.</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Fetching ideogram tracks ...</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Plotting ...</span></pre></body></html></div>
<p><img src="Hoffman_human_single-cell_analysis_files/figure-html/unnamed-chunk-28-1.png" width="80%" style="display: block; margin: auto;"></p>
<pre><code>#&gt; $PPIA
#&gt; $PPIA$chr7
#&gt; Ideogram track 'chr7' for chromosome 7 of the hg38 genome 
#&gt; 
#&gt; $PPIA$OverlayTrack
#&gt; OverlayTrack 'OverlayTrack' containing 2 subtracks
#&gt; 
#&gt; $PPIA$OverlayTrack
#&gt; OverlayTrack 'OverlayTrack' containing 2 subtracks
#&gt; 
#&gt; $PPIA$OverlayTrack
#&gt; OverlayTrack 'OverlayTrack' containing 2 subtracks
#&gt; 
#&gt; $PPIA$OverlayTrack
#&gt; OverlayTrack 'OverlayTrack' containing 2 subtracks
#&gt; 
#&gt; $PPIA$OverlayTrack
#&gt; OverlayTrack 'OverlayTrack' containing 2 subtracks
#&gt; 
#&gt; $PPIA$OverlayTrack
#&gt; OverlayTrack 'OverlayTrack' containing 2 subtracks
#&gt; 
#&gt; $PPIA$OverlayTrack
#&gt; OverlayTrack 'OverlayTrack' containing 2 subtracks
#&gt; 
#&gt; $PPIA$titles
#&gt; An object of class "ImageMap"
#&gt; Slot "coords":
#&gt;              x1        y1   x2        y2
#&gt; chr7          6  120.0000 58.2  189.1341
#&gt; OverlayTrack  6  189.1341 58.2  325.8292
#&gt; OverlayTrack  6  325.8292 58.2  462.5243
#&gt; OverlayTrack  6  462.5243 58.2  599.2195
#&gt; OverlayTrack  6  599.2195 58.2  735.9146
#&gt; OverlayTrack  6  735.9146 58.2  872.6097
#&gt; OverlayTrack  6  872.6097 58.2 1009.3049
#&gt; OverlayTrack  6 1009.3049 58.2 1146.0000
#&gt; 
#&gt; Slot "tags":
#&gt; $title
#&gt;           chr7   OverlayTrack   OverlayTrack   OverlayTrack   OverlayTrack 
#&gt;         "chr7" "OverlayTrack" "OverlayTrack" "OverlayTrack" "OverlayTrack" 
#&gt;   OverlayTrack   OverlayTrack   OverlayTrack 
#&gt; "OverlayTrack" "OverlayTrack" "OverlayTrack"</code></pre>
<p>This visualization shows the structure of the transcripts of <em>PPIA</em>. Our two significant transcripts (<em>PPIA-204</em> and <em>PPIA-209</em>) are quite different, with alternative start and end points as well as some retained intron sequences. The arrows on the right side indicate the mean fitted proportional change in the comparison groups, thus showing a over-expression of <em>PPIA-204</em> in <code>Dex2hr</code> compared to <code>EtOH</code>.</p>
<p>The grey areas indicate intron sequences, which have been compressed in this representation to highlight the exon structure. Only consensus introns are compressed to a defined minimal size. This can also be turned off with <code>reduce_introns=FALSE</code></p>
<p>Analogous as before, we can save plots to disk and add them to the <code>dtu_table</code>:</p>
<div class="sourceCode" id="cb30"><html><body><pre class="r">dturtle &lt;- plot_transcripts_view(dturtle = dturtle, 
                                 gtf = "../gencode.v34.annotation.gtf", 
                                 genome = 'hg38', 
                                 one_to_one = T,
                                 savepath = "images", 
                                 add_to_table = "transcript_view",
                                 BPPARAM = biocpar)
#&gt; 
#&gt; Importing gtf file from disk.
#&gt; 
#&gt; Performing one to one mapping in gtf
#&gt; 
#&gt; Found gtf GRanges for 32 of 32 provided genes.
#&gt; 
#&gt; Fetching ideogram tracks ...
#&gt; 
#&gt; Plotting ...
#&gt; 
  |                                                                            
  |                                                                      |   0%
  |                                                                            
  |=======                                                               |  10%
#&gt; 
#&gt; 
  |                                                                            
  |==============                                                        |  20%
#&gt; 
#&gt; 
  |                                                                            
  |=====================                                                 |  30%
#&gt; 
#&gt; 
  |                                                                            
  |============================                                          |  40%
#&gt; 
#&gt; 
  |                                                                            
  |===================================                                   |  50%
#&gt; 
#&gt; 
  |                                                                            
  |==========================================                            |  60%
#&gt; 
#&gt; 
  |                                                                            
  |=================================================                     |  70%
#&gt; 
#&gt; 
  |                                                                            
  |========================================================              |  80%
#&gt; 
#&gt; 
  |                                                                            
  |===============================================================       |  90%
#&gt; 
#&gt; 
  |                                                                            
  |======================================================================| 100%
#&gt; </pre></body></html></div>
<div class="sourceCode" id="cb31"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="no">dturtle</span>$<span class="no">dtu_table</span>$<span class="no">transcript_view</span>)
<span class="co">#&gt; [1] "images/PSMB4_transcripts.png"   "images/RPL13A_transcripts.png" </span>
<span class="co">#&gt; [3] "images/RPLP1_transcripts.png"   "images/COX6A1_transcripts.png" </span>
<span class="co">#&gt; [5] "images/ATP5F1B_transcripts.png" "images/RPL7A_transcripts.png"</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/list.files.html">list.files</a></span>(<span class="st">"./images/"</span>))
<span class="co">#&gt; [1] "AGR2_barplot.png"        "AGR2_pheatmap.png"      </span>
<span class="co">#&gt; [3] "AGR2_transcripts.png"    "ATP5F1B_barplot.png"    </span>
<span class="co">#&gt; [5] "ATP5F1B_pheatmap.png"    "ATP5F1B_transcripts.png"</span></pre></body></html></div>
<hr>
</div>
<div id="visualize-dtu-table" class="section level3">
<h3 class="hasAnchor">
<a href="#visualize-dtu-table" class="anchor"></a>Visualize DTU table</h3>
<p>The <code>dturtle$dtu_table</code> is now ready to be visualized as an interactive HTML-table. Please note, that it is optional to add any plots or additional columns to the table. Thus the visualization will work directly after calling <code><a href="../reference/create_dtu_table.html">create_dtu_table()</a></code>.</p>
<p>The <code>dtu_table</code> object looks like this:</p>
<div class="sourceCode" id="cb32"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="no">dturtle</span>$<span class="no">dtu_table</span>)
<span class="co">#&gt;          geneID    gene_qval  min_tx_qval n_tx n_sig_tx max(Dex2hr-EtOH)</span>
<span class="co">#&gt; PSMB4     PSMB4 2.384975e-03 7.737662e-04    7        1        0.1549467</span>
<span class="co">#&gt; RPL13A   RPL13A 1.562860e-12 4.164981e-14   12        2        0.1533086</span>
<span class="co">#&gt; RPLP1     RPLP1 8.761113e-15 1.821671e-16    4        3        0.1389187</span>
<span class="co">#&gt; COX6A1   COX6A1 2.352015e-02 4.328509e-03    3        2       -0.1365610</span>
<span class="co">#&gt; ATP5F1B ATP5F1B 4.299094e-05 1.178645e-06    5        2        0.1355355</span>
<span class="co">#&gt; RPL7A     RPL7A 2.041170e-28 6.935383e-18    6        3        0.1320572</span>
<span class="co">#&gt;         chromosome tx_expr_in_max                    barplot</span>
<span class="co">#&gt; PSMB4         chr1      0.6306533   images/PSMB4_barplot.png</span>
<span class="co">#&gt; RPL13A       chr19      0.9183417  images/RPL13A_barplot.png</span>
<span class="co">#&gt; RPLP1        chr15      0.9623116   images/RPLP1_barplot.png</span>
<span class="co">#&gt; COX6A1       chr12      0.5778894  images/COX6A1_barplot.png</span>
<span class="co">#&gt; ATP5F1B      chr12      0.8128141 images/ATP5F1B_barplot.png</span>
<span class="co">#&gt; RPL7A         chr9      0.9786432   images/RPL7A_barplot.png</span>
<span class="co">#&gt;                            pheatmap                transcript_view</span>
<span class="co">#&gt; PSMB4     images/PSMB4_pheatmap.png   images/PSMB4_transcripts.png</span>
<span class="co">#&gt; RPL13A   images/RPL13A_pheatmap.png  images/RPL13A_transcripts.png</span>
<span class="co">#&gt; RPLP1     images/RPLP1_pheatmap.png   images/RPLP1_transcripts.png</span>
<span class="co">#&gt; COX6A1   images/COX6A1_pheatmap.png  images/COX6A1_transcripts.png</span>
<span class="co">#&gt; ATP5F1B images/ATP5F1B_pheatmap.png images/ATP5F1B_transcripts.png</span>
<span class="co">#&gt; RPL7A     images/RPL7A_pheatmap.png   images/RPL7A_transcripts.png</span></pre></body></html></div>
<p>Before creating the actual table, we can optionally define column formatter functions, which colour the specified columns. The colouring might help with to quickly dissect the results.</p>
<p><em>DTUrtle</em> come with some pre-defined column formatter functions (for p-values and percentages), other formatter functions from the <code>formattable</code> package can also be used. Advanced users might also define their own functions.</p>
<p>We create a named list, linking column names to formatter functions:</p>
<div class="sourceCode" id="cb33"><html><body><pre class="r"><span class="no">column_formatter_list</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(
      <span class="st">"gene_qval"</span> <span class="kw">=</span> <span class="fu"><a href="../reference/table_pval_tile.html">table_pval_tile</a></span>(<span class="st">"white"</span>, <span class="st">"orange"</span>, <span class="kw">digits</span> <span class="kw">=</span> <span class="fl">3</span>),
      <span class="st">"min_tx_qval"</span> <span class="kw">=</span> <span class="fu"><a href="../reference/table_pval_tile.html">table_pval_tile</a></span>(<span class="st">"white"</span>, <span class="st">"orange"</span>, <span class="kw">digits</span> <span class="kw">=</span> <span class="fl">3</span>),
      <span class="st">"n_tx"</span> <span class="kw">=</span> <span class="kw pkg">formattable</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/formattable/man/color_tile.html">color_tile</a></span>(<span class="st">'white'</span>, <span class="st">"lightblue"</span>),
      <span class="st">"n_sig_tx"</span> <span class="kw">=</span> <span class="kw pkg">formattable</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/formattable/man/color_tile.html">color_tile</a></span>(<span class="st">'white'</span>, <span class="st">"lightblue"</span>),
      <span class="st">"max(Dex2hr-EtOH)"</span> <span class="kw">=</span> <span class="fu"><a href="../reference/table_percentage_bar.html">table_percentage_bar</a></span>(<span class="st">'lightgreen'</span>, <span class="st">"#FF9999"</span>, <span class="kw">digits</span><span class="kw">=</span><span class="fl">2</span>),
      <span class="st">"tx_expr_in_max"</span> <span class="kw">=</span> <span class="fu"><a href="../reference/table_percentage_bar.html">table_percentage_bar</a></span>(<span class="st">'white'</span>, <span class="st">"lightblue"</span>, <span class="kw">color_break</span> <span class="kw">=</span> <span class="fl">0</span>, <span class="kw">digits</span><span class="kw">=</span><span class="fl">2</span>))</pre></body></html></div>
<p>This <code>column_formatter_list</code> is subsequently provided to <code><a href="../reference/plot_dtu_table.html">plot_dtu_table()</a></code>:</p>
<div class="sourceCode" id="cb34"><html><body><pre class="r"><span class="fu"><a href="../reference/plot_dtu_table.html">plot_dtu_table</a></span>(<span class="kw">dturtle</span> <span class="kw">=</span> <span class="no">dturtle</span>, <span class="kw">savepath</span> <span class="kw">=</span> <span class="st">"my_results.html"</span>,
               <span class="kw">column_formatters</span> <span class="kw">=</span> <span class="no">column_formatter_list</span>)</pre></body></html></div>
<blockquote>
<div class="alert alert-info">
<p><strong>Note:</strong> ️As seen above, the paths to the plots are relative. Please make sure that the saving directory in <code><a href="../reference/plot_dtu_table.html">plot_dtu_table()</a></code> is correctly set and the plots are reachable from that directory with the given path.</p>
<p><strong>The links in the following example are just for demonstration purposes and do not work!</strong></p>
</div>
</blockquote>
<style>
.vscroll-plot {
    overflow: auto;
    resize: both;
}
</style>
<div class="vscroll-plot">
<div id="DTUrtle_table-ir5PnZVFFVLkrev" style="width:90%;height:auto;" class="datatables html-widget"></div>
<script type="application/json" data-for="DTUrtle_table-ir5PnZVFFVLkrev">{"x":{"filter":"top","filterHTML":"<tr>\n  <td data-type=\"character\" style=\"vertical-align: top;\">\n    <div class=\"form-group has-feedback\" style=\"margin-bottom: auto;\">\n      <input type=\"search\" placeholder=\"All\" class=\"form-control\" style=\"width: 100%;\"/>\n      <span class=\"glyphicon glyphicon-remove-circle form-control-feedback\"><\/span>\n    <\/div>\n  <\/td>\n  <td data-type=\"character\" style=\"vertical-align: top;\">\n    <div class=\"form-group has-feedback\" style=\"margin-bottom: auto;\">\n      <input type=\"search\" placeholder=\"All\" class=\"form-control\" style=\"width: 100%;\"/>\n      <span class=\"glyphicon glyphicon-remove-circle form-control-feedback\"><\/span>\n    <\/div>\n  <\/td>\n  <td data-type=\"character\" style=\"vertical-align: top;\">\n    <div class=\"form-group has-feedback\" style=\"margin-bottom: auto;\">\n      <input type=\"search\" placeholder=\"All\" class=\"form-control\" style=\"width: 100%;\"/>\n      <span class=\"glyphicon glyphicon-remove-circle form-control-feedback\"><\/span>\n    <\/div>\n  <\/td>\n  <td data-type=\"character\" style=\"vertical-align: top;\">\n    <div class=\"form-group has-feedback\" style=\"margin-bottom: auto;\">\n      <input type=\"search\" placeholder=\"All\" class=\"form-control\" style=\"width: 100%;\"/>\n      <span class=\"glyphicon glyphicon-remove-circle form-control-feedback\"><\/span>\n    <\/div>\n  <\/td>\n  <td data-type=\"character\" style=\"vertical-align: top;\">\n    <div class=\"form-group has-feedback\" style=\"margin-bottom: auto;\">\n      <input type=\"search\" placeholder=\"All\" class=\"form-control\" style=\"width: 100%;\"/>\n      <span class=\"glyphicon glyphicon-remove-circle form-control-feedback\"><\/span>\n    <\/div>\n  <\/td>\n  <td data-type=\"character\" style=\"vertical-align: top;\">\n    <div class=\"form-group has-feedback\" style=\"margin-bottom: auto;\">\n      <input type=\"search\" placeholder=\"All\" class=\"form-control\" style=\"width: 100%;\"/>\n      <span class=\"glyphicon glyphicon-remove-circle form-control-feedback\"><\/span>\n    <\/div>\n  <\/td>\n  <td data-type=\"character\" style=\"vertical-align: top;\">\n    <div class=\"form-group has-feedback\" style=\"margin-bottom: auto;\">\n      <input type=\"search\" placeholder=\"All\" class=\"form-control\" style=\"width: 100%;\"/>\n      <span class=\"glyphicon glyphicon-remove-circle form-control-feedback\"><\/span>\n    <\/div>\n  <\/td>\n  <td data-type=\"character\" style=\"vertical-align: top;\">\n    <div class=\"form-group has-feedback\" style=\"margin-bottom: auto;\">\n      <input type=\"search\" placeholder=\"All\" class=\"form-control\" style=\"width: 100%;\"/>\n      <span class=\"glyphicon glyphicon-remove-circle form-control-feedback\"><\/span>\n    <\/div>\n  <\/td>\n  <td data-type=\"character\" style=\"vertical-align: top;\">\n    <div class=\"form-group has-feedback\" style=\"margin-bottom: auto;\">\n      <input type=\"search\" placeholder=\"All\" class=\"form-control\" style=\"width: 100%;\"/>\n      <span class=\"glyphicon glyphicon-remove-circle form-control-feedback\"><\/span>\n    <\/div>\n  <\/td>\n  <td data-type=\"character\" style=\"vertical-align: top;\">\n    <div class=\"form-group has-feedback\" style=\"margin-bottom: auto;\">\n      <input type=\"search\" placeholder=\"All\" class=\"form-control\" style=\"width: 100%;\"/>\n      <span class=\"glyphicon glyphicon-remove-circle form-control-feedback\"><\/span>\n    <\/div>\n  <\/td>\n  <td data-type=\"character\" style=\"vertical-align: top;\">\n    <div class=\"form-group has-feedback\" style=\"margin-bottom: auto;\">\n      <input type=\"search\" placeholder=\"All\" class=\"form-control\" style=\"width: 100%;\"/>\n      <span class=\"glyphicon glyphicon-remove-circle form-control-feedback\"><\/span>\n    <\/div>\n  <\/td>\n<\/tr>","class":"display","extensions":["Buttons"],"data":[["PSMB4","RPL13A","RPLP1","COX6A1","ATP5F1B","RPL7A","RPL18A","RPS18","AGR2","RPL5","RPS12","PPIA","RPS20","RPL12","TPT1","RPS8","RPL6","RPS2","RPL30","EEF1A1","HSPB1","NPM1","RPL32","RPL8","RPS3","RPLP2","RPLP0","RPL27A","RPS6","RPL13","RPL23","RPS11"],["<span style=\"display: inline-block; padding: 0 8px; border-radius: 8px; background-color: #fffcf7\">2.385e-03<\/span>","<span style=\"display: inline-block; padding: 0 8px; border-radius: 8px; background-color: #ffe9c1\">1.563e-12<\/span>","<span style=\"display: inline-block; padding: 0 8px; border-radius: 8px; background-color: #ffe4b4\">8.761e-15<\/span>","<span style=\"display: inline-block; padding: 0 8px; border-radius: 8px; background-color: #fffefd\">2.352e-02<\/span>","<span style=\"display: inline-block; padding: 0 8px; border-radius: 8px; background-color: #fff8ed\">4.299e-05<\/span>","<span style=\"display: inline-block; padding: 0 8px; border-radius: 8px; background-color: #ffc865\">2.041e-28<\/span>","<span style=\"display: inline-block; padding: 0 8px; border-radius: 8px; background-color: #fff7e9\">9.141e-06<\/span>","<span style=\"display: inline-block; padding: 0 8px; border-radius: 8px; background-color: #ffc459\">1.579e-30<\/span>","<span style=\"display: inline-block; padding: 0 8px; border-radius: 8px; background-color: #fffefe\">3.629e-02<\/span>","<span style=\"display: inline-block; padding: 0 8px; border-radius: 8px; background-color: #fff7e9\">1.164e-05<\/span>","<span style=\"display: inline-block; padding: 0 8px; border-radius: 8px; background-color: #fffbf5\">1.249e-03<\/span>","<span style=\"display: inline-block; padding: 0 8px; border-radius: 8px; background-color: #ffe1ac\">3.298e-16<\/span>","<span style=\"display: inline-block; padding: 0 8px; border-radius: 8px; background-color: #fffcf7\">2.979e-03<\/span>","<span style=\"display: inline-block; padding: 0 8px; border-radius: 8px; background-color: #fff2db\">3.249e-08<\/span>","<span style=\"display: inline-block; padding: 0 8px; border-radius: 8px; background-color: #fff5e3\">9.112e-07<\/span>","<span style=\"display: inline-block; padding: 0 8px; border-radius: 8px; background-color: #ffefd1\">8.950e-10<\/span>","<span style=\"display: inline-block; padding: 0 8px; border-radius: 8px; background-color: #fffdfb\">1.072e-02<\/span>","<span style=\"display: inline-block; padding: 0 8px; border-radius: 8px; background-color: #ffa500\">7.862e-46<\/span>","<span style=\"display: inline-block; padding: 0 8px; border-radius: 8px; background-color: #ffebc6\">1.042e-11<\/span>","<span style=\"display: inline-block; padding: 0 8px; border-radius: 8px; background-color: #ffd17f\">5.448e-24<\/span>","<span style=\"display: inline-block; padding: 0 8px; border-radius: 8px; background-color: #fffefe\">3.643e-02<\/span>","<span style=\"display: inline-block; padding: 0 8px; border-radius: 8px; background-color: #fffefe\">4.116e-02<\/span>","<span style=\"display: inline-block; padding: 0 8px; border-radius: 8px; background-color: #fffaf2\">3.652e-04<\/span>","<span style=\"display: inline-block; padding: 0 8px; border-radius: 8px; background-color: #ffd384\">4.583e-23<\/span>","<span style=\"display: inline-block; padding: 0 8px; border-radius: 8px; background-color: #ffe7bc\">2.224e-13<\/span>","<span style=\"display: inline-block; padding: 0 8px; border-radius: 8px; background-color: #ffffff\">4.942e-02<\/span>","<span style=\"display: inline-block; padding: 0 8px; border-radius: 8px; background-color: #fff0d4\">2.483e-09<\/span>","<span style=\"display: inline-block; padding: 0 8px; border-radius: 8px; background-color: #fffdfb\">1.261e-02<\/span>","<span style=\"display: inline-block; padding: 0 8px; border-radius: 8px; background-color: #fff3df\">1.811e-07<\/span>","<span style=\"display: inline-block; padding: 0 8px; border-radius: 8px; background-color: #fff4e1\">4.507e-07<\/span>","<span style=\"display: inline-block; padding: 0 8px; border-radius: 8px; background-color: #fffdf9\">5.895e-03<\/span>","<span style=\"display: inline-block; padding: 0 8px; border-radius: 8px; background-color: #fffaf1\">2.494e-04<\/span>"],["<span style=\"display: inline-block; padding: 0 8px; border-radius: 8px; background-color: #ffffff\">7.738e-04<\/span>","<span style=\"display: inline-block; padding: 0 8px; border-radius: 8px; background-color: #ffffff\">4.165e-14<\/span>","<span style=\"display: inline-block; padding: 0 8px; border-radius: 8px; background-color: #ffffff\">1.822e-16<\/span>","<span style=\"display: inline-block; padding: 0 8px; border-radius: 8px; background-color: #ffffff\">4.329e-03<\/span>","<span style=\"display: inline-block; padding: 0 8px; border-radius: 8px; background-color: #ffffff\">1.179e-06<\/span>","<span style=\"display: inline-block; padding: 0 8px; border-radius: 8px; background-color: #ffffff\">6.935e-18<\/span>","<span style=\"display: inline-block; padding: 0 8px; border-radius: 8px; background-color: #ffffff\">1.367e-07<\/span>","<span style=\"display: inline-block; padding: 0 8px; border-radius: 8px; background-color: #ffffff\">2.064e-11<\/span>","<span style=\"display: inline-block; padding: 0 8px; border-radius: 8px; background-color: #ffffff\">7.126e-03<\/span>","<span style=\"display: inline-block; padding: 0 8px; border-radius: 8px; background-color: #ffffff\">3.894e-03<\/span>","<span style=\"display: inline-block; padding: 0 8px; border-radius: 8px\">0.000e+00<\/span>","<span style=\"display: inline-block; padding: 0 8px; border-radius: 8px; background-color: #ffffff\">5.441e-18<\/span>","<span style=\"display: inline-block; padding: 0 8px; border-radius: 8px; background-color: #ffffff\">3.135e-03<\/span>","<span style=\"display: inline-block; padding: 0 8px; border-radius: 8px; background-color: #ffffff\">2.128e-09<\/span>","<span style=\"display: inline-block; padding: 0 8px; border-radius: 8px; background-color: #ffffff\">1.914e-02<\/span>","<span style=\"display: inline-block; padding: 0 8px; border-radius: 8px; background-color: #ffffff\">7.144e-04<\/span>","<span style=\"display: inline-block; padding: 0 8px; border-radius: 8px; background-color: #ffffff\">1.517e-03<\/span>","<span style=\"display: inline-block; padding: 0 8px; border-radius: 8px; background-color: #ffffff\">6.345e-01<\/span>","<span style=\"display: inline-block; padding: 0 8px; border-radius: 8px; background-color: #ffffff\">1.045e-06<\/span>","<span style=\"display: inline-block; padding: 0 8px; border-radius: 8px; background-color: #ffffff\">1.000e+00<\/span>","<span style=\"display: inline-block; padding: 0 8px; border-radius: 8px; background-color: #ffffff\">5.686e-03<\/span>","<span style=\"display: inline-block; padding: 0 8px; border-radius: 8px; background-color: #ffffff\">1.000e+00<\/span>","<span style=\"display: inline-block; padding: 0 8px; border-radius: 8px; background-color: #ffffff\">3.931e-02<\/span>","<span style=\"display: inline-block; padding: 0 8px; border-radius: 8px; background-color: #ffffff\">1.000e+00<\/span>","<span style=\"display: inline-block; padding: 0 8px; border-radius: 8px; background-color: #ffffff\">8.292e-01<\/span>","<span style=\"display: inline-block; padding: 0 8px; border-radius: 8px; background-color: #ffffff\">1.000e+00<\/span>","<span style=\"display: inline-block; padding: 0 8px; border-radius: 8px; background-color: #ffffff\">1.000e+00<\/span>","<span style=\"display: inline-block; padding: 0 8px; border-radius: 8px; background-color: #ffffff\">1.000e+00<\/span>","<span style=\"display: inline-block; padding: 0 8px; border-radius: 8px; background-color: #ffffff\">6.476e-08<\/span>","<span style=\"display: inline-block; padding: 0 8px; border-radius: 8px; background-color: #ffffff\">7.435e-06<\/span>","<span style=\"display: inline-block; padding: 0 8px; border-radius: 8px; background-color: #ffffff\">4.327e-03<\/span>","<span style=\"display: inline-block; padding: 0 8px; border-radius: 8px; background-color: #ffffff\">1.000e+00<\/span>"],["<span style=\"display: block; padding: 0 4px; border-radius: 4px; background-color: #dff0f5\">7<\/span>","<span style=\"display: block; padding: 0 4px; border-radius: 4px; background-color: #bfe1eb\">12<\/span>","<span style=\"display: block; padding: 0 4px; border-radius: 4px; background-color: #f2f9fb\">4<\/span>","<span style=\"display: block; padding: 0 4px; border-radius: 4px; background-color: #f8fcfd\">3<\/span>","<span style=\"display: block; padding: 0 4px; border-radius: 4px; background-color: #ecf6f9\">5<\/span>","<span style=\"display: block; padding: 0 4px; border-radius: 4px; background-color: #e5f3f7\">6<\/span>","<span style=\"display: block; padding: 0 4px; border-radius: 4px; background-color: #e5f3f7\">6<\/span>","<span style=\"display: block; padding: 0 4px; border-radius: 4px; background-color: #d9edf3\">8<\/span>","<span style=\"display: block; padding: 0 4px; border-radius: 4px; background-color: #f8fcfd\">3<\/span>","<span style=\"display: block; padding: 0 4px; border-radius: 4px; background-color: #e5f3f7\">6<\/span>","<span style=\"display: block; padding: 0 4px; border-radius: 4px; background-color: #ffffff\">2<\/span>","<span style=\"display: block; padding: 0 4px; border-radius: 4px; background-color: #dff0f5\">7<\/span>","<span style=\"display: block; padding: 0 4px; border-radius: 4px; background-color: #c6e4ed\">11<\/span>","<span style=\"display: block; padding: 0 4px; border-radius: 4px; background-color: #f2f9fb\">4<\/span>","<span style=\"display: block; padding: 0 4px; border-radius: 4px; background-color: #b9dee9\">13<\/span>","<span style=\"display: block; padding: 0 4px; border-radius: 4px; background-color: #dff0f5\">7<\/span>","<span style=\"display: block; padding: 0 4px; border-radius: 4px; background-color: #f8fcfd\">3<\/span>","<span style=\"display: block; padding: 0 4px; border-radius: 4px; background-color: #bfe1eb\">12<\/span>","<span style=\"display: block; padding: 0 4px; border-radius: 4px; background-color: #d9edf3\">8<\/span>","<span style=\"display: block; padding: 0 4px; border-radius: 4px; background-color: #dff0f5\">7<\/span>","<span style=\"display: block; padding: 0 4px; border-radius: 4px; background-color: #f8fcfd\">3<\/span>","<span style=\"display: block; padding: 0 4px; border-radius: 4px; background-color: #ecf6f9\">5<\/span>","<span style=\"display: block; padding: 0 4px; border-radius: 4px; background-color: #dff0f5\">7<\/span>","<span style=\"display: block; padding: 0 4px; border-radius: 4px; background-color: #bfe1eb\">12<\/span>","<span style=\"display: block; padding: 0 4px; border-radius: 4px; background-color: #add8e6\">15<\/span>","<span style=\"display: block; padding: 0 4px; border-radius: 4px; background-color: #d9edf3\">8<\/span>","<span style=\"display: block; padding: 0 4px; border-radius: 4px; background-color: #b3dbe7\">14<\/span>","<span style=\"display: block; padding: 0 4px; border-radius: 4px; background-color: #d2eaf1\">9<\/span>","<span style=\"display: block; padding: 0 4px; border-radius: 4px; background-color: #f2f9fb\">4<\/span>","<span style=\"display: block; padding: 0 4px; border-radius: 4px; background-color: #bfe1eb\">12<\/span>","<span style=\"display: block; padding: 0 4px; border-radius: 4px; background-color: #d2eaf1\">9<\/span>","<span style=\"display: block; padding: 0 4px; border-radius: 4px; background-color: #d2eaf1\">9<\/span>"],["<span style=\"display: block; padding: 0 4px; border-radius: 4px; background-color: #e3f2f6\">1<\/span>","<span style=\"display: block; padding: 0 4px; border-radius: 4px; background-color: #c8e5ee\">2<\/span>","<span style=\"display: block; padding: 0 4px; border-radius: 4px; background-color: #add8e6\">3<\/span>","<span style=\"display: block; padding: 0 4px; border-radius: 4px; background-color: #c8e5ee\">2<\/span>","<span style=\"display: block; padding: 0 4px; border-radius: 4px; background-color: #c8e5ee\">2<\/span>","<span style=\"display: block; padding: 0 4px; border-radius: 4px; background-color: #add8e6\">3<\/span>","<span style=\"display: block; padding: 0 4px; border-radius: 4px; background-color: #e3f2f6\">1<\/span>","<span style=\"display: block; padding: 0 4px; border-radius: 4px; background-color: #c8e5ee\">2<\/span>","<span style=\"display: block; padding: 0 4px; border-radius: 4px; background-color: #c8e5ee\">2<\/span>","<span style=\"display: block; padding: 0 4px; border-radius: 4px; background-color: #e3f2f6\">1<\/span>","<span style=\"display: block; padding: 0 4px; border-radius: 4px; background-color: #c8e5ee\">2<\/span>","<span style=\"display: block; padding: 0 4px; border-radius: 4px; background-color: #c8e5ee\">2<\/span>","<span style=\"display: block; padding: 0 4px; border-radius: 4px; background-color: #e3f2f6\">1<\/span>","<span style=\"display: block; padding: 0 4px; border-radius: 4px; background-color: #c8e5ee\">2<\/span>","<span style=\"display: block; padding: 0 4px; border-radius: 4px; background-color: #e3f2f6\">1<\/span>","<span style=\"display: block; padding: 0 4px; border-radius: 4px; background-color: #e3f2f6\">1<\/span>","<span style=\"display: block; padding: 0 4px; border-radius: 4px; background-color: #e3f2f6\">1<\/span>","<span style=\"display: block; padding: 0 4px; border-radius: 4px; background-color: #ffffff\">0<\/span>","<span style=\"display: block; padding: 0 4px; border-radius: 4px; background-color: #add8e6\">3<\/span>","<span style=\"display: block; padding: 0 4px; border-radius: 4px; background-color: #ffffff\">0<\/span>","<span style=\"display: block; padding: 0 4px; border-radius: 4px; background-color: #e3f2f6\">1<\/span>","<span style=\"display: block; padding: 0 4px; border-radius: 4px; background-color: #ffffff\">0<\/span>","<span style=\"display: block; padding: 0 4px; border-radius: 4px; background-color: #e3f2f6\">1<\/span>","<span style=\"display: block; padding: 0 4px; border-radius: 4px; background-color: #ffffff\">0<\/span>","<span style=\"display: block; padding: 0 4px; border-radius: 4px; background-color: #ffffff\">0<\/span>","<span style=\"display: block; padding: 0 4px; border-radius: 4px; background-color: #ffffff\">0<\/span>","<span style=\"display: block; padding: 0 4px; border-radius: 4px; background-color: #ffffff\">0<\/span>","<span style=\"display: block; padding: 0 4px; border-radius: 4px; background-color: #ffffff\">0<\/span>","<span style=\"display: block; padding: 0 4px; border-radius: 4px; background-color: #e3f2f6\">1<\/span>","<span style=\"display: block; padding: 0 4px; border-radius: 4px; background-color: #e3f2f6\">1<\/span>","<span style=\"display: block; padding: 0 4px; border-radius: 4px; background-color: #e3f2f6\">1<\/span>","<span style=\"display: block; padding: 0 4px; border-radius: 4px; background-color: #ffffff\">0<\/span>"],["<span style=\"display: block; direction: ltr; border-radius: 3px; background-color: #FF9999; width: 100.00%\">15.49%<\/span>","<span style=\"display: block; direction: ltr; border-radius: 3px; background-color: #FF9999; width: 98.94%\">15.33%<\/span>","<span style=\"display: block; direction: ltr; border-radius: 3px; background-color: #FF9999; width: 89.66%\">13.89%<\/span>","<span style=\"display: block; direction: ltr; border-radius: 3px; background-color: lightgreen; width: 88.13%\">-13.66%<\/span>","<span style=\"display: block; direction: ltr; border-radius: 3px; background-color: #FF9999; width: 87.47%\">13.55%<\/span>","<span style=\"display: block; direction: ltr; border-radius: 3px; background-color: #FF9999; width: 85.23%\">13.21%<\/span>","<span style=\"display: block; direction: ltr; border-radius: 3px; background-color: lightgreen; width: 77.91%\">-12.07%<\/span>","<span style=\"display: block; direction: ltr; border-radius: 3px; background-color: #FF9999; width: 61.91%\">9.59%<\/span>","<span style=\"display: block; direction: ltr; border-radius: 3px; background-color: lightgreen; width: 57.47%\">-8.91%<\/span>","<span style=\"display: block; direction: ltr; border-radius: 3px; background-color: #FF9999; width: 57.10%\">8.85%<\/span>","<span style=\"display: block; direction: ltr; border-radius: 3px; background-color: #FF9999; width: 55.67%\">8.63%<\/span>","<span style=\"display: block; direction: ltr; border-radius: 3px; background-color: #FF9999; width: 54.56%\">8.45%<\/span>","<span style=\"display: block; direction: ltr; border-radius: 3px; background-color: #FF9999; width: 54.20%\">8.40%<\/span>","<span style=\"display: block; direction: ltr; border-radius: 3px; background-color: lightgreen; width: 52.99%\">-8.21%<\/span>","<span style=\"display: block; direction: ltr; border-radius: 3px; background-color: #FF9999; width: 49.15%\">7.62%<\/span>","<span style=\"display: block; direction: ltr; border-radius: 3px; background-color: #FF9999; width: 46.99%\">7.28%<\/span>","<span style=\"display: block; direction: ltr; border-radius: 3px; background-color: lightgreen; width: 45.05%\">-6.98%<\/span>","<span style=\"display: block; direction: ltr; border-radius: 3px; background-color: #FF9999; width: 43.43%\">6.73%<\/span>","<span style=\"display: block; direction: ltr; border-radius: 3px; background-color: #FF9999; width: 42.92%\">6.65%<\/span>","<span style=\"display: block; direction: ltr; border-radius: 3px; background-color: lightgreen; width: 39.60%\">-6.14%<\/span>","<span style=\"display: block; direction: ltr; border-radius: 3px; background-color: #FF9999; width: 38.83%\">6.02%<\/span>","<span style=\"display: block; direction: ltr; border-radius: 3px; background-color: lightgreen; width: 38.53%\">-5.97%<\/span>","<span style=\"display: block; direction: ltr; border-radius: 3px; background-color: #FF9999; width: 36.29%\">5.62%<\/span>","<span style=\"display: block; direction: ltr; border-radius: 3px; background-color: #FF9999; width: 35.46%\">5.49%<\/span>","<span style=\"display: block; direction: ltr; border-radius: 3px; background-color: #FF9999; width: 34.43%\">5.34%<\/span>","<span style=\"display: block; direction: ltr; border-radius: 3px; background-color: #FF9999; width: 30.28%\">4.69%<\/span>","<span style=\"display: block; direction: ltr; border-radius: 3px; background-color: lightgreen; width: 30.10%\">-4.66%<\/span>","<span style=\"display: block; direction: ltr; border-radius: 3px; background-color: #FF9999; width: 29.93%\">4.64%<\/span>","<span style=\"display: block; direction: ltr; border-radius: 3px; background-color: lightgreen; width: 28.73%\">-4.45%<\/span>","<span style=\"display: block; direction: ltr; border-radius: 3px; background-color: #FF9999; width: 28.37%\">4.40%<\/span>","<span style=\"display: block; direction: ltr; border-radius: 3px; background-color: lightgreen; width: 25.83%\">-4.00%<\/span>","<span style=\"display: block; direction: ltr; border-radius: 3px; background-color: #FF9999; width: 24.75%\">3.83%<\/span>"],["chr1","chr19","chr15","chr12","chr12","chr9","chr19","chr6","chr7","chr1","chr6","chr7","chr8","chr9","chr13","chr1","chr12","chr16","chr8","chr6","chr7","chr5","chr3","chr8","chr11","chr11","chr12","chr11","chr9","chr16","chr17","chr19"],["<span style=\"display: block; direction: ltr; border-radius: 3px; background-color: lightblue; width: 63.62%\">63.07%<\/span>","<span style=\"display: block; direction: ltr; border-radius: 3px; background-color: lightblue; width: 92.65%\">91.83%<\/span>","<span style=\"display: block; direction: ltr; border-radius: 3px; background-color: lightblue; width: 97.08%\">96.23%<\/span>","<span style=\"display: block; direction: ltr; border-radius: 3px; background-color: lightblue; width: 58.30%\">57.79%<\/span>","<span style=\"display: block; direction: ltr; border-radius: 3px; background-color: lightblue; width: 82.00%\">81.28%<\/span>","<span style=\"display: block; direction: ltr; border-radius: 3px; background-color: lightblue; width: 98.73%\">97.86%<\/span>","<span style=\"display: block; direction: ltr; border-radius: 3px; background-color: lightblue; width: 83.02%\">82.29%<\/span>","<span style=\"display: block; direction: ltr; border-radius: 3px; background-color: lightblue; width: 99.11%\">98.24%<\/span>","<span style=\"display: block; direction: ltr; border-radius: 3px; background-color: lightblue; width: 73.51%\">72.86%<\/span>","<span style=\"display: block; direction: ltr; border-radius: 3px; background-color: lightblue; width: 79.59%\">78.89%<\/span>","<span style=\"display: block; direction: ltr; border-radius: 3px; background-color: lightblue; width: 92.52%\">91.71%<\/span>","<span style=\"display: block; direction: ltr; border-radius: 3px; background-color: lightblue; width: 98.99%\">98.12%<\/span>","<span style=\"display: block; direction: ltr; border-radius: 3px; background-color: lightblue; width: 95.82%\">94.97%<\/span>","<span style=\"display: block; direction: ltr; border-radius: 3px; background-color: lightblue; width: 94.42%\">93.59%<\/span>","<span style=\"display: block; direction: ltr; border-radius: 3px; background-color: lightblue; width: 93.03%\">92.21%<\/span>","<span style=\"display: block; direction: ltr; border-radius: 3px; background-color: lightblue; width: 90.87%\">90.08%<\/span>","<span style=\"display: block; direction: ltr; border-radius: 3px; background-color: lightblue; width: 86.69%\">85.93%<\/span>","<span style=\"display: block; direction: ltr; border-radius: 3px; background-color: lightblue; width: 100.00%\">99.12%<\/span>","<span style=\"display: block; direction: ltr; border-radius: 3px; background-color: lightblue; width: 98.10%\">97.24%<\/span>","<span style=\"display: block; direction: ltr; border-radius: 3px; background-color: lightblue; width: 99.87%\">98.99%<\/span>","<span style=\"display: block; direction: ltr; border-radius: 3px; background-color: lightblue; width: 76.43%\">75.75%<\/span>","<span style=\"display: block; direction: ltr; border-radius: 3px; background-color: lightblue; width: 68.95%\">68.34%<\/span>","<span style=\"display: block; direction: ltr; border-radius: 3px; background-color: lightblue; width: 97.97%\">97.11%<\/span>","<span style=\"display: block; direction: ltr; border-radius: 3px; background-color: lightblue; width: 95.94%\">95.10%<\/span>","<span style=\"display: block; direction: ltr; border-radius: 3px; background-color: lightblue; width: 96.58%\">95.73%<\/span>","<span style=\"display: block; direction: ltr; border-radius: 3px; background-color: lightblue; width: 89.48%\">88.69%<\/span>","<span style=\"display: block; direction: ltr; border-radius: 3px; background-color: lightblue; width: 96.96%\">96.11%<\/span>","<span style=\"display: block; direction: ltr; border-radius: 3px; background-color: lightblue; width: 84.54%\">83.79%<\/span>","<span style=\"display: block; direction: ltr; border-radius: 3px; background-color: lightblue; width: 94.30%\">93.47%<\/span>","<span style=\"display: block; direction: ltr; border-radius: 3px; background-color: lightblue; width: 99.75%\">98.87%<\/span>","<span style=\"display: block; direction: ltr; border-radius: 3px; background-color: lightblue; width: 89.99%\">89.20%<\/span>","<span style=\"display: block; direction: ltr; border-radius: 3px; background-color: lightblue; width: 87.71%\">86.93%<\/span>"],["<a href='images/PSMB4_barplot.png' target='_blank'>Link<\/a>","<a href='images/RPL13A_barplot.png' target='_blank'>Link<\/a>","<a href='images/RPLP1_barplot.png' target='_blank'>Link<\/a>","<a href='images/COX6A1_barplot.png' target='_blank'>Link<\/a>","<a href='images/ATP5F1B_barplot.png' target='_blank'>Link<\/a>","<a href='images/RPL7A_barplot.png' target='_blank'>Link<\/a>","<a href='images/RPL18A_barplot.png' target='_blank'>Link<\/a>","<a href='images/RPS18_barplot.png' target='_blank'>Link<\/a>","<a href='images/AGR2_barplot.png' target='_blank'>Link<\/a>","<a href='images/RPL5_barplot.png' target='_blank'>Link<\/a>","<a href='images/RPS12_barplot.png' target='_blank'>Link<\/a>","<a href='images/PPIA_barplot.png' target='_blank'>Link<\/a>","<a href='images/RPS20_barplot.png' target='_blank'>Link<\/a>","<a href='images/RPL12_barplot.png' target='_blank'>Link<\/a>","<a href='images/TPT1_barplot.png' target='_blank'>Link<\/a>","<a href='images/RPS8_barplot.png' target='_blank'>Link<\/a>","<a href='images/RPL6_barplot.png' target='_blank'>Link<\/a>","<a href='images/RPS2_barplot.png' target='_blank'>Link<\/a>","<a href='images/RPL30_barplot.png' target='_blank'>Link<\/a>","<a href='images/EEF1A1_barplot.png' target='_blank'>Link<\/a>","<a href='images/HSPB1_barplot.png' target='_blank'>Link<\/a>","<a href='images/NPM1_barplot.png' target='_blank'>Link<\/a>","<a href='images/RPL32_barplot.png' target='_blank'>Link<\/a>","<a href='images/RPL8_barplot.png' target='_blank'>Link<\/a>","<a href='images/RPS3_barplot.png' target='_blank'>Link<\/a>","<a href='images/RPLP2_barplot.png' target='_blank'>Link<\/a>","<a href='images/RPLP0_barplot.png' target='_blank'>Link<\/a>","<a href='images/RPL27A_barplot.png' target='_blank'>Link<\/a>","<a href='images/RPS6_barplot.png' target='_blank'>Link<\/a>","<a href='images/RPL13_barplot.png' target='_blank'>Link<\/a>","<a href='images/RPL23_barplot.png' target='_blank'>Link<\/a>","<a href='images/RPS11_barplot.png' target='_blank'>Link<\/a>"],["<a href='images/PSMB4_pheatmap.png' target='_blank'>Link<\/a>","<a href='images/RPL13A_pheatmap.png' target='_blank'>Link<\/a>","<a href='images/RPLP1_pheatmap.png' target='_blank'>Link<\/a>","<a href='images/COX6A1_pheatmap.png' target='_blank'>Link<\/a>","<a href='images/ATP5F1B_pheatmap.png' target='_blank'>Link<\/a>","<a href='images/RPL7A_pheatmap.png' target='_blank'>Link<\/a>","<a href='images/RPL18A_pheatmap.png' target='_blank'>Link<\/a>","<a href='images/RPS18_pheatmap.png' target='_blank'>Link<\/a>","<a href='images/AGR2_pheatmap.png' target='_blank'>Link<\/a>","<a href='images/RPL5_pheatmap.png' target='_blank'>Link<\/a>","<a href='images/RPS12_pheatmap.png' target='_blank'>Link<\/a>","<a href='images/PPIA_pheatmap.png' target='_blank'>Link<\/a>","<a href='images/RPS20_pheatmap.png' target='_blank'>Link<\/a>","<a href='images/RPL12_pheatmap.png' target='_blank'>Link<\/a>","<a href='images/TPT1_pheatmap.png' target='_blank'>Link<\/a>","<a href='images/RPS8_pheatmap.png' target='_blank'>Link<\/a>","<a href='images/RPL6_pheatmap.png' target='_blank'>Link<\/a>","<a href='images/RPS2_pheatmap.png' target='_blank'>Link<\/a>","<a href='images/RPL30_pheatmap.png' target='_blank'>Link<\/a>","<a href='images/EEF1A1_pheatmap.png' target='_blank'>Link<\/a>","<a href='images/HSPB1_pheatmap.png' target='_blank'>Link<\/a>","<a href='images/NPM1_pheatmap.png' target='_blank'>Link<\/a>","<a href='images/RPL32_pheatmap.png' target='_blank'>Link<\/a>","<a href='images/RPL8_pheatmap.png' target='_blank'>Link<\/a>","<a href='images/RPS3_pheatmap.png' target='_blank'>Link<\/a>","<a href='images/RPLP2_pheatmap.png' target='_blank'>Link<\/a>","<a href='images/RPLP0_pheatmap.png' target='_blank'>Link<\/a>","<a href='images/RPL27A_pheatmap.png' target='_blank'>Link<\/a>","<a href='images/RPS6_pheatmap.png' target='_blank'>Link<\/a>","<a href='images/RPL13_pheatmap.png' target='_blank'>Link<\/a>","<a href='images/RPL23_pheatmap.png' target='_blank'>Link<\/a>","<a href='images/RPS11_pheatmap.png' target='_blank'>Link<\/a>"],["<a href='images/PSMB4_transcripts.png' target='_blank'>Link<\/a>","<a href='images/RPL13A_transcripts.png' target='_blank'>Link<\/a>","<a href='images/RPLP1_transcripts.png' target='_blank'>Link<\/a>","<a href='images/COX6A1_transcripts.png' target='_blank'>Link<\/a>","<a href='images/ATP5F1B_transcripts.png' target='_blank'>Link<\/a>","<a href='images/RPL7A_transcripts.png' target='_blank'>Link<\/a>","<a href='images/RPL18A_transcripts.png' target='_blank'>Link<\/a>","<a href='images/RPS18_transcripts.png' target='_blank'>Link<\/a>","<a href='images/AGR2_transcripts.png' target='_blank'>Link<\/a>","<a href='images/RPL5_transcripts.png' target='_blank'>Link<\/a>","<a href='images/RPS12_transcripts.png' target='_blank'>Link<\/a>","<a href='images/PPIA_transcripts.png' target='_blank'>Link<\/a>","<a href='images/RPS20_transcripts.png' target='_blank'>Link<\/a>","<a href='images/RPL12_transcripts.png' target='_blank'>Link<\/a>","<a href='images/TPT1_transcripts.png' target='_blank'>Link<\/a>","<a href='images/RPS8_transcripts.png' target='_blank'>Link<\/a>","<a href='images/RPL6_transcripts.png' target='_blank'>Link<\/a>","<a href='images/RPS2_transcripts.png' target='_blank'>Link<\/a>","<a href='images/RPL30_transcripts.png' target='_blank'>Link<\/a>","<a href='images/EEF1A1_transcripts.png' target='_blank'>Link<\/a>","<a href='images/HSPB1_transcripts.png' target='_blank'>Link<\/a>","<a href='images/NPM1_transcripts.png' target='_blank'>Link<\/a>","<a href='images/RPL32_transcripts.png' target='_blank'>Link<\/a>","<a href='images/RPL8_transcripts.png' target='_blank'>Link<\/a>","<a href='images/RPS3_transcripts.png' target='_blank'>Link<\/a>","<a href='images/RPLP2_transcripts.png' target='_blank'>Link<\/a>","<a href='images/RPLP0_transcripts.png' target='_blank'>Link<\/a>","<a href='images/RPL27A_transcripts.png' target='_blank'>Link<\/a>","<a href='images/RPS6_transcripts.png' target='_blank'>Link<\/a>","<a href='images/RPL13_transcripts.png' target='_blank'>Link<\/a>","<a href='images/RPL23_transcripts.png' target='_blank'>Link<\/a>","<a href='images/RPS11_transcripts.png' target='_blank'>Link<\/a>"]],"container":"<div class='header'><img src=\"data:image/svg+xml;base64,<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 226.304 43.209" height="163.31" width="855.323">
  <path d="M144.078 9.583a.121.121 0 00-.066.027c-1.145.955-2.17 2.049-3.137 3.174a.121.121 0 00-.002.002c-.86 1.079-1.837 2.245-2.035 3.701-.072.469.11.882.41 1.178.3.296.712.483 1.125.537.174.029.32.004.428-.066a.37.37 0 00.17-.305c0-.228-.21-.435-.538-.455-.566-.165-.895-.801-.712-1.361.496-1.575 1.636-2.857 2.761-4.067l.004-.006c.501-.481.946-1.047 1.477-1.412 1.017.478 1.955 1.478 1.896 2.639-.163 1.325-1.262 2.302-2.285 3.125-.185.113-.422.33-.638.5a1.91 1.91 0 01-.237.162c.216-.483.309-.92.33-1.354.025-.49-.009-.987.045-1.593a.122.122 0 00-.242-.024c-.273 2.419-1.67 4.592-3.5 6.164-1.928 1.278-.984.632-2.885 1.97a.121.121 0 00.13.204c.388-.214.595-.308.694-.34.049-.015.068-.013.065-.013h.002c.018.013.057.04.105.043.054.002.1-.011.164-.037.128-.052.33-.163.713-.397a.121.121 0 00.008-.006c1.717-1.24 4.07-1.472 6.026-.683a.121.121 0 00.105-.215c-.483-.283-.645-.48-.853-.647-.157-.126-.383-.218-.705-.293.785-.437 1.646-.787 2.533-.754.784.02 1.434.527 1.754 1.235.131.286.372.624.511.941.07.159.111.309.11.436a.414.414 0 01-.135.312c-1.577.502-3.227.96-4.865.733a.121.121 0 00-.004-.002 1.214 1.214 0 01-.89-.534 1.826 1.826 0 01-.304-1.043.121.121 0 00-.025-.074.533.533 0 00-.205-.174.307.307 0 00-.234-.007c-.143.057-.222.191-.28.341-.057.15-.088.326-.09.494 0 .169.025.332.112.46l-.02-.043c.087.412.341.722.658.937.317.215.698.338 1.065.39a.121.121 0 00.004.003c1.441.153 2.886-.157 4.248-.55a.121.121 0 00.031-.013c.223-.139.694-.203 1.102-.318.204-.058.397-.131.543-.258a.615.615 0 00.203-.553.121.121 0 000-.008c-.25-1.264-.8-2.745-2.143-3.265a.121.121 0 00-.006-.002c-1.554-.506-3.14.245-4.402 1.045a.121.121 0 00-.014.008c-.222.184-.486.248-.781.289-.276.037-.572.057-.863.142l.064-.058a.121.121 0 00.002-.002c.622-.58 1.068-1.297 1.639-1.86 1.247-.845 2.572-1.664 3.523-2.902l.002-.002c.36-.448.693-1.001.86-1.59.165-.586.16-1.214-.166-1.783-.454-.996-1.392-1.62-2.29-2.108a.121.121 0 00-.07-.015z" paint-order="markers stroke fill"/>
  <text style="line-height:1.25;-inkscape-font-specification:'Lobster Two'" x="44.377" y="70.413" font-weight="400" font-size="10.583" font-family="Lobster Two" letter-spacing="0" word-spacing="0" stroke-width=".265" transform="translate(19.81 -93.777)"/>
  <text style="line-height:1.25;-inkscape-font-specification:'Lobster Two'" x="-20.572" y="133.697" font-weight="400" font-size="10.583" font-family="Lobster Two" letter-spacing="0" word-spacing="0" stroke-width=".07" stroke-dasharray=".07 .07" transform="translate(19.81 -93.777)">
    <tspan x="-20.572" y="133.697" style="-inkscape-font-specification:'Lobster Two'" font-size="42.333">DTUrt<tspan letter-spacing=".265">l</tspan>e</tspan>
  </text>
  <path d="M214.62 43.19c-1.387-.216-2.7-1.044-4.015-2.533-.691-.783-2.09-2.58-3.446-4.429l-1.032-1.406-.357.047c-.197.026-1.34.222-2.54.437-3.342.597-4.628.696-9.843.765-4.213.055-5.97-.022-8.312-.36l-1.525-.22-.295.481c-2.814 4.603-3.4 5.297-5.21 6.162-4.77 2.28-10.85-1.26-11.146-6.491-.083-1.474.2-2.6 1.142-4.543.563-1.162.587-1.232.449-1.34-.36-.278-2.468-1.226-3.44-1.547-1.944-.64-3.873-1.018-7.155-1.4-3.197-.372-4.576-.696-5.942-1.398-2.113-1.086-3.685-3.447-4.092-6.145-.403-2.673.408-5.425 2.176-7.391 1.364-1.515 3.813-2.609 5.846-2.61 1.186 0 2.795.395 4.506 1.107.403.168 1.732.786 2.955 1.374 2.723 1.309 3.279 1.558 4.27 1.917l.78.281.122-.296c.329-.796.998-2.117 1.402-2.769 3.19-5.142 9.603-9.022 17.159-10.38 9.017-1.621 18.785.74 25.638 6.199 1.085.864 2.914 2.678 3.708 3.678a20.76 20.76 0 013.195 5.711c.198.544.453 1.129.566 1.3.202.305.212.31.575.258 1.371-.196 1.72-.215 2.341-.133.888.119 1.5.43 2.082 1.059 1.113 1.201 1.432 2.865.806 4.202-.377.805-2.899 3.36-4.86 4.924-.436.348-.805.665-.82.704-.014.039.293.509.683 1.044 1.997 2.741 2.599 3.845 2.843 5.217.399 2.233-.344 4.454-2.027 6.062-1.699 1.623-3.58 2.404-5.957 2.472-.545.015-1.099.011-1.23-.01zm2.12-4.722c.957-.203 1.985-.916 2.326-1.612.185-.377.21-.514.18-.938-.045-.593-.263-1.186-.712-1.93-.535-.89-2.13-3.05-2.25-3.05-.167 0-.773.25-2.775 1.147-1.004.45-2.05.898-2.323.997-.336.121-.474.208-.428.268l1.263 1.683c1.402 1.872 2.32 2.992 2.607 3.18.477.312 1.347.418 2.111.255zM175.991 38c.516-.223 1.008-.754 1.794-1.943.67-1.013.883-1.372.833-1.413a56.63 56.63 0 00-1.623-.65 171.471 171.471 0 01-4.29-1.756c-.046-.02-.27.35-.495.82-1.018 2.12-.85 3.248.664 4.477.708.574.849.644 1.423.7.627.062 1.183-.016 1.694-.235zm19.58-6.43c5.737-.371 13.425-1.708 18.012-3.133 5.937-1.843 9.237-4.004 10.645-6.967.472-.994.324-1.282-.545-1.068-.84.208-1.97.678-6.96 2.898-6.394 2.844-6.439 2.86-9.208 3.423-3.417.693-6.397 1.106-9.287 1.286-1.878.117-6.1.047-7.461-.123-2.045-.256-4.946-.864-7.62-1.596-2.768-.758-3.977-1.233-7.077-2.778-2.356-1.173-3.874-2.043-5.645-3.231-1.475-.991-2.614-1.605-3.86-2.08-1.136-.434-2.657-1.098-4.382-1.915-3.435-1.626-3.94-1.854-4.532-2.052-1.498-.502-3.179-.272-4.105.56-.796.716-1.151 1.853-1.102 3.526.026.888.052 1.042.243 1.429a3.709 3.709 0 002.073 1.86c.678.256 1.517.402 3.463.6 2.642.271 5.274.757 7.511 1.388 1.705.48 2.436.82 5.546 2.578 1.718.97 2.539 1.39 3.88 1.982 4.241 1.87 9.194 3.05 14.297 3.406 1.398.097 4.656.101 6.113.007zm-40.225-13.318c-.12-.014-.219-.062-.219-.108 0-.045-.053-.062-.117-.037-.07.027-.153-.02-.203-.113-.046-.087-.118-.137-.16-.111-.041.025-.075-.01-.075-.08a.2.2 0 00-.12-.172c-.065-.025-.118-.13-.118-.235 0-.105-.033-.19-.074-.19-.098 0-.036-.795.07-.901a.307.307 0 00.083-.186c0-.135.249-.421.366-.421.051 0 .114-.054.14-.12.025-.066.073-.112.107-.102.034.01.133-.02.22-.067.2-.106.802-.075.952.05.062.051.15.07.196.042.046-.029.105.004.131.072.027.069.098.124.159.124.06 0 .11.055.11.122s.056.143.124.169c.068.026.106.094.084.15-.022.057.006.141.061.187.122.101.098.983-.027 1.025-.046.015-.083.092-.083.171 0 .08-.054.164-.12.19a.196.196 0 00-.118.167c0 .07-.07.121-.164.121-.09 0-.14.036-.114.08.028.045-.039.08-.154.08-.112 0-.203.028-.203.063 0 .063-.327.077-.734.03zm43.51 5.347c1.11-.027 2.038-.062 2.064-.077.025-.015.368-.051.76-.08.852-.065 2.065-.19 2.342-.243.11-.02.306-.054.437-.075.131-.02.792-.127 1.468-.237.677-.11 1.374-.221 1.548-.248 2.249-.342 4.914-1.274 6.231-2.18.19-.13.587-.35 1.905-1.053.328-.175.604-.324.615-.333.047-.036-1.24-2.533-1.698-3.296-1.294-2.153-1.524-2.484-2.324-3.35-.455-.494-1.116-1.211-1.468-1.596-.468-.51-.824-.808-1.316-1.104a41.234 41.234 0 01-1.35-.855c-.932-.621-5.463-2.89-6.111-3.06a81.428 81.428 0 01-1.151-.351c-1.179-.365-1.603-.448-2.66-.523-.414-.029-.771-.063-.793-.075-.022-.012-.415-.048-.873-.08-.459-.03-.851-.067-.873-.08-.05-.03-2.868-.149-3.572-.152-.673-.003-2.898.109-3.255.163-.257.039-1.983.506-2.897.784-3.946 1.2-4.528 1.41-5.506 1.976-.497.289-.913.524-.924.524-.01 0-.433.241-.939.536-.797.465-1.152.75-2.673 2.155-.964.89-1.853 1.765-1.975 1.945-.445.655-.865 1.724-1.081 2.753-.175.832-.184.809.458 1.187.316.187.79.471 1.052.632.261.162 1.1.661 1.865 1.11.764.45 1.522.902 1.685 1.005.163.103.574.297.913.432.34.135.652.272.696.305.099.074 1.793.747 2.222.882.175.055.46.147.635.205.85.28 3.343.992 3.973 1.135a261.248 261.248 0 012.298.532c.087.02.552.09 1.032.153a254.32 254.32 0 011.706.233c.197.027.447.063.556.08.485.075 2.223.234 3.572.325.415.028.765.06.778.072.013.012.156.01.318-.003a101.12 101.12 0 012.31-.073zm6.912-2.619c-.847-.283-1.082-.395-1.842-.885-1.077-.693-2.235-1.995-2.833-3.187-.093-.185-.213-.337-.267-.336-.054 0-.49.358-.97.796-1.847 1.682-3.364 2.454-5.082 2.587-1.17.09-2.482-.258-3.53-.938-.593-.384-1.715-1.48-2.464-2.407-.298-.368-.636-.694-.803-.776-.272-.133-.31-.134-.615-.011-.18.071-.778.48-1.33.909-1.596 1.24-2.972 1.823-4.302 1.823-.958 0-1.55-.399-1.36-.916.1-.268.264-.35 1.04-.517 1.16-.25 1.25-.354 1.547-1.812.255-1.257.42-2.73.48-4.3l.051-1.316-.653-.27c-1.07-.441-1.468-.926-1.23-1.495.167-.395.53-.513 1.117-.363.609.155 1.817.738 2.346 1.132.472.351 1.249 1.15 2.111 2.173.319.377.705.782.86.9l.279.213.274-.195c.15-.107.636-.526 1.078-.931 1.662-1.52 3.175-2.22 4.816-2.223 1.686-.004 3.094.83 5.287 3.13.69.722 1.294 1.313 1.344 1.313.05 0 .52-.324 1.042-.72 2.126-1.61 3.342-2.133 4.974-2.136.791-.002 1.035.069 1.22.353.114.172.118.238.033.444-.121.294-.38.435-1.086.594-.846.19-.965.256-1.05.582-.213.806-.612 4.049-.684 5.566-.06 1.257-.025 1.323.978 1.856.396.21.798.466.894.568.234.25.16.528-.198.75-.507.313-.657.318-1.472.045zM194.692 18c.2-.2.472-1.91.757-4.76.179-1.79.202-2.634.079-2.865-.073-.136-.173-.153-.88-.153-.769 0-.805.008-1.021.224a1.47 1.47 0 00-.314.536c-.252.882-.655 4.07-.771 6.106-.052.906-.051.909.139.988.304.128 1.867.068 2.011-.076zm9.137-.83c.127-.305.389-2.244.474-3.515.06-.881.048-.905-.35-.7-.354.184-1.377 1.324-1.622 1.808-.206.409-.216.47-.13.788.189.701.692 1.47 1.143 1.745.326.199.352.192.485-.127zm-5.938-.167c.789-.392 1.764-1.662 1.766-2.299.002-.48-1.07-2.37-1.665-2.934-.234-.223-.258-.23-.354-.098-.236.32-.662 2.923-.734 4.483l-.048 1.052h.314c.177 0 .49-.088.721-.204zm-6.895-.912c.08-.228.265-2.07.36-3.567.057-.887.002-.97-.443-.677-.564.37-1.3 1.172-1.459 1.589-.124.324-.131.437-.053.811.114.54.57 1.37.979 1.778.372.372.504.386.616.066zm-5.196-1.227c.56-.534 1-1.165 1-1.437 0-.39-1.263-2.454-1.56-2.548-.048-.016-.112.093-.14.243-.089.473-.283 3.078-.284 3.807l-.001.7.24-.142c.13-.078.466-.358.745-.623z"/>
</svg>
\" /><div class='txt'><h2 align='center' data-toc-skip>Dex2hr vs. EtOH (399 vs.\n                      397)<\/h2><h4 align='center' data-toc-skip>Significant differential genes: 32<br>Significant differential transcripts: 37<\/h4><\/div><\/div><table>\n  <thead>\n    <tr>\n      <th>geneID<\/th>\n      <th>gene_qval<\/th>\n      <th>min_tx_qval<\/th>\n      <th>n_tx<\/th>\n      <th>n_sig_tx<\/th>\n      <th>max(Dex2hr-EtOH)<\/th>\n      <th>chromosome<\/th>\n      <th>tx_expr_in_max<\/th>\n      <th>barplot<\/th>\n      <th>pheatmap<\/th>\n      <th>transcript_view<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"dom":"lBfrtip","orderClasses":true,"buttons":[{"extend":"collection","buttons":["csv","excel"],"text":"Download"}],"autoWidth":true,"pageLength":10,"lengthMenu":[10,25,32],"columnDefs":[{"targets":"_all","render":"function(data, type, row, meta) {\n                              if (type == 'display') {\n                                return data;\n                              } else if (type == 'filter' || type == 'type'){\n                                var new_dat = String(data).replace(/<.*?>/g, '').replace(/%/g, '')\n                                if(new_dat !== '' && !isNaN(Number(new_dat))){\n                                  console.log(data);\n                                  console.log(type);\n                                  console.log(new_dat);\n                                  return Number(new_dat);\n                                }else{\n                                  return new_dat;\n                                }\n                              } else if (type == 'sort'){\n                                var new_dat = String(data).replace(/<.*?>/g, '').replace(/%/g, '')\n                                if(new_dat !== '' && !isNaN(Number(new_dat))){\n                                  return Math.abs(Number(new_dat));\n                                }else{\n                                return new_dat;\n                                }\n                              }\n                            }"}],"initComplete":"function(settings, json) {\nfunction addStyleString(str) {\nvar node = document.createElement('style');\nnode.innerHTML = str\ndocument.body.appendChild(node);\n}\naddStyleString('td { padding: 5px 7px !important; max-width: 200px; text-align: center;}');\naddStyleString('.datatables { margin-left: auto; margin-right: auto;}');\naddStyleString('.dt-buttons { margin-left: 50px;}');\naddStyleString('.header img { float: left; max-width: 25%; min-width: 150px;}');\naddStyleString('.header .txt { display: flow-root; padding-right: 15%;}');\naddStyleString('.header h2 { font-weight: 900; margin: auto; padding: initial;}');\naddStyleString('.header h4 { margin: auto; padding: initial;}');\naddStyleString('#DTUrtle_table-ir5PnZVFFVLkrev { font-size: 120%;}');\n}","order":[],"orderCellsTop":true}},"evals":["options.columnDefs.0.render","options.initComplete"],"jsHooks":[]}</script>
</div>
<hr>
</div>
</div>
<div id="workflow-with-seurat-object" class="section level2">
<h2 class="hasAnchor">
<a href="#workflow-with-seurat-object" class="anchor"></a>Workflow with Seurat object</h2>
<p>Some users might already have analyzed their (gene-level) single-cell data and performed clustering, dimensional reduction etc. For these users, <em>DTUrtle</em> offers direct support for <em>Seurat</em> objects (Seurat V3 or newer).</p>
<div id="seurat-object-creation" class="section level3">
<h3 class="hasAnchor">
<a href="#seurat-object-creation" class="anchor"></a>Seurat object creation</h3>
<p>For this tutorial, we will create a <em>Seurat</em> object on the fly - users with an already existing object might want to directly jump to the second step.</p>
<p>We create a <em>Seurat</em> object on the gene-level counts, using the <em>DTUrtle</em> functionality <code><a href="../reference/summarize_to_gene.html">summarize_to_gene()</a></code>:</p>
<div class="sourceCode" id="cb35"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">Seurat</span>)

<span class="co">#summarize counts to gene-level</span>
<span class="no">cts_gene</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/summarize_to_gene.html">summarize_to_gene</a></span>(<span class="no">cts</span>, <span class="kw">tx2gene</span> <span class="kw">=</span> <span class="no">tx2gene</span>, <span class="kw">fun</span> <span class="kw">=</span> <span class="st">"sum"</span>)

<span class="co">##compare dimensions</span>
<span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span>(<span class="no">cts</span>)
<span class="co">#&gt; [1] 77347   796</span>
<span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span>(<span class="no">cts_gene</span>)
<span class="co">#&gt; [1] 18412   796</span></pre></body></html></div>
<p>We use these gene level counts for a standard <em>Seurat</em> analysis (Normalization + Scaling, PCA, Clustering, UMAP).</p>
<div class="sourceCode" id="cb36"><html><body><pre class="r"><span class="no">seur_obj</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/CreateSeuratObject.html">CreateSeuratObject</a></span>(<span class="kw">counts</span> <span class="kw">=</span> <span class="no">cts_gene</span>, <span class="kw">meta.data</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(<span class="kw">row.names</span><span class="kw">=</span><span class="no">pd</span>$<span class="no">id</span>, <span class="st">"group"</span><span class="kw">=</span><span class="no">pd</span>$<span class="no">group</span>, <span class="kw">stringsAsFactors</span> <span class="kw">=</span> <span class="no">F</span>))
<span class="co">#&gt; Warning: Feature names cannot have underscores ('_'), replacing with dashes</span>
<span class="co">#&gt; ('-')</span>

<span class="co">#check if names are kept</span>
<span class="fu"><a href="https://rdrr.io/r/base/all.html">all</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/Cells.html">Cells</a></span>(<span class="no">seur_obj</span>) <span class="kw">%in%</span> <span class="no">pd</span>$<span class="no">id</span>)
<span class="co">#&gt; [1] TRUE</span>

<span class="co">#run basic pipeline</span>
<span class="no">seur_obj</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/NormalizeData.html">NormalizeData</a></span>(<span class="no">seur_obj</span>)
<span class="no">seur_obj</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/FindVariableFeatures.html">FindVariableFeatures</a></span>(<span class="no">seur_obj</span>, <span class="kw">selection.method</span> <span class="kw">=</span> <span class="st">"vst"</span>, <span class="kw">nfeatures</span> <span class="kw">=</span> <span class="fl">5000</span>)
<span class="no">seur_obj</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/ScaleData.html">ScaleData</a></span>(<span class="no">seur_obj</span>, <span class="kw">model.use</span> <span class="kw">=</span> <span class="st">"negbinom"</span>)
<span class="co">#&gt; Centering and scaling data matrix</span>
<span class="no">seur_obj</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/RunPCA.html">RunPCA</a></span>(<span class="no">seur_obj</span>, <span class="kw">seed.use</span> <span class="kw">=</span> <span class="fl">42</span>, <span class="kw">features</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/VariableFeatures.html">VariableFeatures</a></span>(<span class="kw">object</span> <span class="kw">=</span> <span class="no">seur_obj</span>))
<span class="co">#&gt; PC_ 1 </span>
<span class="co">#&gt; Positive:  RPL7A, RPS3A, PPIA, RPSA, RPL13, RPS10, RPL6, RPS17, KRT19, TUBA1B </span>
<span class="co">#&gt;     KRT18, EEF1G, AGR2, C1QBP, UBB, GAGE12F, NME1-NME2, PRDX1, SNRPD3, PPIB </span>
<span class="co">#&gt;     SERPINA6, ARL6IP1, GSTM3, NQO1, TP53I3, PIP, C15orf65, GNG11, LRRC59, UBE2C </span>
<span class="co">#&gt; Negative:  RPL6P27, AC091153.3, PNMT, BBS9, DHX58, PCLO, ZNF629, STIM1, AL031058.1, SAMD4A </span>
<span class="co">#&gt;     RAB28, SVIL, LINC02163, FOXP1, PRELID2, RBM38, ARHGEF16, AF235103.3, XYLB, KIFC3 </span>
<span class="co">#&gt;     ZNF248, TMEM229B, FUT1, MAPK15, AC131009.3, KDM3A, ARNTL, TMEM17, CHPF2, CIART </span>
<span class="co">#&gt; PC_ 2 </span>
<span class="co">#&gt; Positive:  RBAK, AL049840.6, HMGCS2, JUNB, RAD51B, INO80D, MAN1B1-DT, CAMKK1, ELF2, CSRNP2 </span>
<span class="co">#&gt;     F3, ELMOD3, AC011445.2, AC090809.1, AC012073.1, DNAJC27-AS1, PPP2CB, PARD3, LRFN1, ASXL2 </span>
<span class="co">#&gt;     ZNF345, LINC01205, AL845472.2, BX470102.2, RALY-AS1, CARD10, LINC00664, AC022784.1, HEATR5B, ID1 </span>
<span class="co">#&gt; Negative:  DNMT3B, SBF2, SRP14-AS1, TNFAIP2, AC104667.2, AC022784.5, TFCP2, RPL13P12, C17orf75, KPNA3 </span>
<span class="co">#&gt;     UBFD1, PGR, CC2D1A, ZNF131, RBM47, FOXRED2, ZBTB5, RPRD2, ZNF239, FDXACB1 </span>
<span class="co">#&gt;     LTV1, ALDH1A3, MYO1F, CEP55, ADAT1, GLUL, UBE3C, CCDC43, IGSF5, VWA5A </span>
<span class="co">#&gt; PC_ 3 </span>
<span class="co">#&gt; Positive:  NME1-NME2, PSMD7, PIP, AC116036.2, EEF1G, ZNF197, C19orf53, PRDX1, KRT18, KRT19 </span>
<span class="co">#&gt;     ZCRB1, RPS10-NUDT3, LENG1, NADSYN1, PLOD3, GAGE12J, RPSA, LINC01089, PRDM11, AL358472.2 </span>
<span class="co">#&gt;     SETD4, TAC3, CCNB1, PPM1J, BMP1, MCM7, FASTKD3, RPL13, MFHAS1, DAGLB </span>
<span class="co">#&gt; Negative:  H1-0, DNAJB4, SLC16A9, LINC01446, ZDHHC9, ZFYVE26, FAM210B, DDX21, MIS18A, ZNF131 </span>
<span class="co">#&gt;     KLHL18, ADAT1, UBE3C, GORAB, RNF103, SLC26A11, DDI2, BTN2A2, TMEM87A, TAF1C </span>
<span class="co">#&gt;     ARHGAP11A, MRPS35, EYA3, LINC01311, ZNF492, MT2A, LPAR1, DHFR2, UBFD1, FOXRED2 </span>
<span class="co">#&gt; PC_ 4 </span>
<span class="co">#&gt; Positive:  C15orf65, SRP14-AS1, TNFAIP2, BX890604.2, DNMT3B, EPHX1, AC022784.5, C17orf75, ALKBH2, DCAF12 </span>
<span class="co">#&gt;     EML4, AC104667.2, TSPAN12, BICD2, MTATP6P1, C9orf85, CHST15, ASMTL, TBILA, ARVCF </span>
<span class="co">#&gt;     MREG, RGS12, PIP5K1C, ABCC1, ZNF702P, PLAAT2, SRGN, LRRC59, BMP1, RIC8B </span>
<span class="co">#&gt; Negative:  PANK2-AS1, RAB20, MT1X, AKR1A1, ZNF37BP, MX1, NOL6, LIPT2, TJP1, CLIC3 </span>
<span class="co">#&gt;     EXOC1, SLC15A3, COMMD2, ELMO3, PNPLA3, FAM122A, RIPK2, TRMT2A, AC005332.4, CYP2U1 </span>
<span class="co">#&gt;     OSBPL3, KIF16B, AC132872.1, KIAA1211L, ARSD, TBC1D10A, AC245033.1, PRDX1, SPATA13, AL356277.3 </span>
<span class="co">#&gt; PC_ 5 </span>
<span class="co">#&gt; Positive:  DCAF7, PPP4R4, IGFBP5, GOPC, RAB7A, PTPN14, ANLN, CLPTM1L, CYP2B7P, YAP1 </span>
<span class="co">#&gt;     FAM120A, ACIN1, ZIC2, GRHL2, ALOXE3, SOAT1, TPH1, SLC2A6, PDCD4, HLCS </span>
<span class="co">#&gt;     XXYLT1, NIPSNAP2, ARHGAP11A, OTUB2, TXNIP, RACGAP1, ATR, ERO1B, KIF14, SLC5A3 </span>
<span class="co">#&gt; Negative:  KRT18, AC027682.6, AC129926.1, KRT19, PPIA, RPL17-C18orf32, METTL4, NUDT14, CCDC97, TYMP </span>
<span class="co">#&gt;     CDIP1, FBXO2, RAB19, CCNJ, HSD17B8, GRAMD2B, NME1-NME2, PMS2P4, C1GALT1C1, RASSF8-AS1 </span>
<span class="co">#&gt;     AC010913.1, GLIS2, SPR, IFNGR1, FASTKD2, ENPP4, ZNF696, PANK2-AS1, NME4, EFCAB2</span>

<span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/ElbowPlot.html">ElbowPlot</a></span>(<span class="no">seur_obj</span>)</pre></body></html></div>
<p><img src="Hoffman_human_single-cell_analysis_files/figure-html/unnamed-chunk-36-1.png" width="80%" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb37"><html><body><pre class="r">
<span class="co">#7 PCs seem like a reasonable cut-off</span>
<span class="no">dims</span> <span class="kw">&lt;-</span> <span class="fl">1</span>:<span class="fl">7</span>

<span class="no">seur_obj</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/FindNeighbors.html">FindNeighbors</a></span>(<span class="no">seur_obj</span>, <span class="kw">dims</span> <span class="kw">=</span> <span class="no">dims</span>, <span class="kw">k.param</span><span class="kw">=</span><span class="fl">10</span>)
<span class="co">#&gt; Computing nearest neighbor graph</span>
<span class="co">#&gt; Computing SNN</span>
<span class="no">seur_obj</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/FindClusters.html">FindClusters</a></span>(<span class="no">seur_obj</span>, <span class="kw">resolution</span> <span class="kw">=</span> <span class="fl">0.7</span>, <span class="kw">n.start</span><span class="kw">=</span><span class="fl">100</span>, <span class="kw">n.iter</span><span class="kw">=</span><span class="fl">100</span>, <span class="kw">random.seed</span> <span class="kw">=</span> <span class="fl">42</span>)
<span class="co">#&gt; Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Number of nodes: 796</span>
<span class="co">#&gt; Number of edges: 11143</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Running Louvain algorithm...</span>
<span class="co">#&gt; Maximum modularity in 100 random starts: 0.7644</span>
<span class="co">#&gt; Number of communities: 6</span>
<span class="co">#&gt; Elapsed time: 0 seconds</span>
<span class="no">seur_obj</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/RunUMAP.html">RunUMAP</a></span>(<span class="no">seur_obj</span>, <span class="kw">dims</span> <span class="kw">=</span> <span class="no">dims</span>, <span class="kw">n.epochs</span><span class="kw">=</span><span class="fl">1000</span>, <span class="kw">seed.use</span><span class="kw">=</span><span class="fl">42</span>, <span class="kw">umap.method</span> <span class="kw">=</span> <span class="st">"umap-learn"</span>)</pre></body></html></div>
<p>The computed UMAP coordinates look like this:</p>
<div class="sourceCode" id="cb38"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/DimPlot.html">DimPlot</a></span>(<span class="no">seur_obj</span>, <span class="kw">reduction</span> <span class="kw">=</span> <span class="st">"umap"</span>, <span class="kw">group.by</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"seurat_clusters"</span>, <span class="st">"group"</span>))</pre></body></html></div>
<p><img src="Hoffman_human_single-cell_analysis_files/figure-html/unnamed-chunk-37-1.png" width="80%" style="display: block; margin: auto;"></p>
<hr>
</div>
<div id="seurat-usage-with-dturtle" class="section level3">
<h3 class="hasAnchor">
<a href="#seurat-usage-with-dturtle" class="anchor"></a>Seurat usage with DTUrtle</h3>
<p><em>Seurat</em> objects can be utilized in two parts of the DTUrtle workflow: <code><a href="../reference/combine_to_matrix.html">combine_to_matrix()</a></code> and <code>run_umap()</code>.</p>
<p>To better highlight the functionality, lets assume we only want to perform a DTU analysis in one of the identified cell clusters - Cluster 4.</p>
<p>We subset the <em>Seurat</em> object to this cluster only:</p>
<div class="sourceCode" id="cb39"><html><body><pre class="r"><span class="no">seur_obj_sub</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/subset.Seurat.html">subset</a></span>(<span class="no">seur_obj</span>, <span class="kw">subset</span> <span class="kw">=</span> <span class="no">seurat_clusters</span><span class="kw">==</span><span class="fl">4</span>)
<span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span>(<span class="no">seur_obj_sub</span>)
<span class="co">#&gt; [1] 18412   103</span></pre></body></html></div>
<p>We can provide this subsetted <em>Seurat</em> object as a parameter in <code><a href="../reference/combine_to_matrix.html">combine_to_matrix()</a></code>. <em>DTUrtle</em> will try to auto-detect cell name extensions and will subset the count matrix to the cells from the <em>Seurat</em> object. Additionally we can provide a <code>tx2gene</code> data frame, which will be added as feature-level metadata to the <em>Seurat</em> assay.</p>
<div class="sourceCode" id="cb40"><html><body><pre class="r"><span class="no">seur_obj_sub</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/combine_to_matrix.html">combine_to_matrix</a></span>(<span class="kw">tx_list</span> <span class="kw">=</span> <span class="no">cts_list</span>, <span class="kw">seurat_obj</span> <span class="kw">=</span> <span class="no">seur_obj_sub</span>, <span class="kw">tx2gene</span> <span class="kw">=</span> <span class="no">tx2gene</span>)
<span class="co">#&gt; Merging matrices</span>
<span class="co">#&gt; Of 796 cells, 103 (13%) could be found in the provided seurat object.</span>
<span class="co">#&gt; 693 (87%) are unique to the transcriptomic files.</span>
<span class="co">#&gt; The seurat object contains 0 additional cells.</span>
<span class="co">#&gt; Subsetting!</span>
<span class="co">#&gt; Excluding 172918 overall not expressed features.</span>
<span class="co">#&gt; 54283 features left.</span>
<span class="co">#&gt; Adding assay 'dtutx'</span></pre></body></html></div>
<p>The transcript level expression of the 103 cells have been carried over to the <em>Seurat</em> object in an own assay.</p>
<div class="sourceCode" id="cb41"><html><body><pre class="r"><span class="no">seur_obj_sub</span>@<span class="kw">assays</span>$<span class="no">dtutx</span>@<span class="kw">data</span>[<span class="fl">1</span>:<span class="fl">4</span>, <span class="fl">1</span>:<span class="fl">4</span>]
<span class="co">#&gt; 4 x 4 sparse Matrix of class "dgCMatrix"</span>
<span class="co">#&gt;                TGAATTCAGACTAGTCTG CGGCGTCTTACGTTCTTG TCGGGACCTCTAGAGTGA</span>
<span class="co">#&gt; AL627309.1-201                  .                  .                  .</span>
<span class="co">#&gt; CICP27-201                      .                  .                  .</span>
<span class="co">#&gt; AL627309.6-201                  .                  .                  .</span>
<span class="co">#&gt; AL627309.5-206                  .                  .                  .</span>
<span class="co">#&gt;                AACGTGTCCAAGTTGCTC</span>
<span class="co">#&gt; AL627309.1-201                  .</span>
<span class="co">#&gt; CICP27-201                      .</span>
<span class="co">#&gt; AL627309.6-201                  .</span>
<span class="co">#&gt; AL627309.5-206                  .</span>
<span class="no">seur_obj_sub</span>@<span class="kw">assays</span>$<span class="no">dtutx</span>@<span class="kw">meta.features</span>[<span class="fl">1</span>:<span class="fl">4</span>,]
<span class="co">#&gt;                transcript_name  gene_name seqnames  start    end width strand</span>
<span class="co">#&gt; AL627309.1-201  AL627309.1-201 AL627309.1     chr1  89295 120932 31638      -</span>
<span class="co">#&gt; CICP27-201          CICP27-201     CICP27     chr1 131025 134836  3812      +</span>
<span class="co">#&gt; AL627309.6-201  AL627309.6-201 AL627309.6     chr1 135141 135895   755      -</span>
<span class="co">#&gt; AL627309.5-206  AL627309.5-206 AL627309.5     chr1 167129 169240  2112      -</span>
<span class="co">#&gt;                source       type score phase           gene_id</span>
<span class="co">#&gt; AL627309.1-201 HAVANA transcript    NA    NA ENSG00000238009.6</span>
<span class="co">#&gt; CICP27-201     HAVANA transcript    NA    NA ENSG00000233750.3</span>
<span class="co">#&gt; AL627309.6-201 HAVANA transcript    NA    NA ENSG00000268903.1</span>
<span class="co">#&gt; AL627309.5-206 HAVANA transcript    NA    NA ENSG00000241860.7</span>
<span class="co">#&gt;                           gene_type level    hgnc_id          havana_gene</span>
<span class="co">#&gt; AL627309.1-201               lncRNA     2       &lt;NA&gt; OTTHUMG00000001096.1</span>
<span class="co">#&gt; CICP27-201     processed_pseudogene     1 HGNC:48835 OTTHUMG00000001257.1</span>
<span class="co">#&gt; AL627309.6-201 processed_pseudogene     2       &lt;NA&gt; OTTHUMG00000182518.1</span>
<span class="co">#&gt; AL627309.5-206               lncRNA     2       &lt;NA&gt; OTTHUMG00000002480.1</span>
<span class="co">#&gt;                    transcript_id      transcript_type transcript_support_level</span>
<span class="co">#&gt; AL627309.1-201 ENST00000466430.5               lncRNA                        5</span>
<span class="co">#&gt; CICP27-201     ENST00000442987.3 processed_pseudogene                       NA</span>
<span class="co">#&gt; AL627309.6-201 ENST00000494149.2 processed_pseudogene                       NA</span>
<span class="co">#&gt; AL627309.5-206 ENST00000655252.1               lncRNA                     &lt;NA&gt;</span>
<span class="co">#&gt;                   tag    havana_transcript exon_number exon_id         ont</span>
<span class="co">#&gt; AL627309.1-201  basic OTTHUMT00000003225.1        &lt;NA&gt;    &lt;NA&gt;        &lt;NA&gt;</span>
<span class="co">#&gt; CICP27-201      basic OTTHUMT00000003691.1        &lt;NA&gt;    &lt;NA&gt; PGO:0000004</span>
<span class="co">#&gt; AL627309.6-201  basic OTTHUMT00000461982.1        &lt;NA&gt;    &lt;NA&gt; PGO:0000004</span>
<span class="co">#&gt; AL627309.5-206 TAGENE OTTHUMT00000528579.1        &lt;NA&gt;    &lt;NA&gt;        &lt;NA&gt;</span>
<span class="co">#&gt;                protein_id ccdsid</span>
<span class="co">#&gt; AL627309.1-201       &lt;NA&gt;   &lt;NA&gt;</span>
<span class="co">#&gt; CICP27-201           &lt;NA&gt;   &lt;NA&gt;</span>
<span class="co">#&gt; AL627309.6-201       &lt;NA&gt;   &lt;NA&gt;</span>
<span class="co">#&gt; AL627309.5-206       &lt;NA&gt;   &lt;NA&gt;</span></pre></body></html></div>
<p><em>Seurat</em> functions can be used for visualizing expression of single transcripts:</p>
<div class="sourceCode" id="cb42"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/FeaturePlot.html">FeaturePlot</a></span>(<span class="no">seur_obj_sub</span>, <span class="kw">features</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"PPIA-204"</span>, <span class="st">"PPIA-209"</span>))</pre></body></html></div>
<p><img src="Hoffman_human_single-cell_analysis_files/figure-html/unnamed-chunk-41-1.png" width="80%" style="display: block; margin: auto;"></p>
<hr>
<p>The enhanced <em>Seurat</em> object can be used in the actual DTU analysis, if we provide it to the <em>DTUrtle</em> function <code><a href="../reference/run_drimseq.html">run_drimseq()</a></code>. As the <code>tx2gene</code> information was already added in the <code><a href="../reference/combine_to_matrix.html">combine_to_matrix()</a></code> step, we only need to provide the column names of the transcript and gene identifier columns of the feature level metadata. Everything we need for the analysis is present in the <em>Seurat</em> object:</p>
<div class="sourceCode" id="cb43"><html><body><pre class="r">dturtle_sub &lt;- run_drimseq(counts = seur_obj_sub, pd = seur_obj_sub@meta.data,
                           tx2gene = c("transcript_name", "gene_name"),
                           cond_col="group", cond_levels = c("Dex2hr", "EtOH"), 
                           filtering_strategy = "sc", BPPARAM = biocpar)
#&gt; Using tx2gene columns:
#&gt;  transcript_name ---&gt; 'feature_id'
#&gt;  gene_name ---&gt; 'gene_id'
#&gt; 
#&gt; Comparing Dex2hr vs EtOH
#&gt; 
#&gt; Proceed with cells/samples: 
#&gt; Dex2hr   EtOH 
#&gt;     68     35
#&gt; 
#&gt; Filtering...
#&gt; 
  |                                                                            
  |                                                                      |   0%
  |                                                                            
  |=======                                                               |  10%
  |                                                                            
  |==============                                                        |  20%
  |                                                                            
  |=====================                                                 |  30%
  |                                                                            
  |============================                                          |  40%
  |                                                                            
  |===================================                                   |  50%
  |                                                                            
  |==========================================                            |  60%
  |                                                                            
  |=================================================                     |  70%
  |                                                                            
  |========================================================              |  80%
  |                                                                            
  |===============================================================       |  90%
  |                                                                            
  |======================================================================| 100%
#&gt; Retain 35301 of 54283 features.
#&gt; Removed 18982 features.
#&gt; 
#&gt; Performing statistical tests...
#&gt; * Calculating mean gene expression..
#&gt; Took 0.7662 seconds.
#&gt; * Estimating common precision..
#&gt; Took 1.942 seconds.
#&gt; ! Using common_precision = 778.5341 as prec_init !
#&gt; * Estimating genewise precision..
#&gt; ! Using loess fit as a shrinkage factor !
#&gt; Took 5.623 seconds.
#&gt; * Fitting the DM model..
#&gt;    Using the one way approach.
#&gt; Took 7.5172 seconds.
#&gt; * Fitting the BB model..
#&gt;    Using the one way approach.
#&gt; Took 3.7895 seconds.
#&gt; * Fitting the DM model..
#&gt;    Using the one way approach.
#&gt; Took 6.2214 seconds.
#&gt; * Calculating likelihood ratio statistics..
#&gt; Took 0.003 seconds.
#&gt; * Fitting the BB model..
#&gt;    Using the one way approach.
#&gt; Took 3.4503 seconds.
#&gt; * Calculating likelihood ratio statistics..
#&gt; Took 0.0126 seconds.</pre></body></html></div>
<p>With this <code>DTUrtle</code> object we can proceed in the standard way, i.e. with <code><a href="../reference/posthoc_and_stager.html">posthoc_and_stager()</a></code> and following visualization.</p>
<div class="sourceCode" id="cb44"><html><body><pre class="r"><span class="no">dturtle_sub</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/posthoc_and_stager.html">posthoc_and_stager</a></span>(<span class="kw">dturtle</span> <span class="kw">=</span> <span class="no">dturtle_sub</span>, <span class="kw">ofdr</span> <span class="kw">=</span> <span class="fl">0.05</span>, <span class="kw">posthoc</span> <span class="kw">=</span> <span class="fl">0.1</span>)
<span class="co">#&gt; Posthoc filtered 33200 features.</span>
<span class="co">#&gt; The returned adjusted p-values are based on a stage-wise testing approach and are only valid for the provided target OFDR level of 5%. If a different target OFDR level is of interest,the entire adjustment should be re-run.</span>
<span class="co">#&gt; Found 4 significant genes with 2 significant transcripts (OFDR: 0.05)</span></pre></body></html></div>
<hr>
</div>
</div>
<div id="session-info" class="section level2">
<h2 class="hasAnchor">
<a href="#session-info" class="anchor"></a>Session info</h2>
<pre><code>#&gt; R version 3.6.2 (2019-12-12)
#&gt; Platform: x86_64-pc-linux-gnu (64-bit)
#&gt; Running under: Debian GNU/Linux 10 (buster)
#&gt; 
#&gt; Matrix products: default
#&gt; BLAS/LAPACK: /usr/lib/x86_64-linux-gnu/libopenblasp-r0.3.5.so
#&gt; 
#&gt; locale:
#&gt;  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
#&gt;  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
#&gt;  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=C             
#&gt;  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
#&gt;  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
#&gt; [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
#&gt; 
#&gt; attached base packages:
#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     
#&gt; 
#&gt; other attached packages:
#&gt; [1] Seurat_3.1.4  DTUrtle_0.2.6
#&gt; 
#&gt; loaded via a namespace (and not attached):
#&gt;   [1] reticulate_1.15             tidyselect_1.0.0           
#&gt;   [3] RSQLite_2.2.0               AnnotationDbi_1.48.0       
#&gt;   [5] htmlwidgets_1.5.1           grid_3.6.2                 
#&gt;   [7] BiocParallel_1.20.1         Rtsne_0.15                 
#&gt;   [9] munsell_0.5.0               mutoss_0.1-12              
#&gt;  [11] codetools_0.2-16            ica_1.0-2                  
#&gt;  [13] DT_0.13                     future_1.16.0              
#&gt;  [15] colorspace_1.4-1            Biobase_2.46.0             
#&gt;  [17] sparseDRIMSeq_0.1.2         knitr_1.28                 
#&gt;  [19] rstudioapi_0.11             stats4_3.6.2               
#&gt;  [21] ROCR_1.0-7                  gbRd_0.4-11                
#&gt;  [23] listenv_0.8.0               Rdpack_0.11-1              
#&gt;  [25] labeling_0.3                tximport_1.14.2            
#&gt;  [27] GenomeInfoDbData_1.2.2      mnormt_1.5-6               
#&gt;  [29] bit64_0.9-7                 farver_2.0.3               
#&gt;  [31] pheatmap_1.0.12             rprojroot_1.3-2            
#&gt;  [33] TH.data_1.0-10              Matrix.utils_0.9.8         
#&gt;  [35] vctrs_0.2.4                 xfun_0.12                  
#&gt;  [37] biovizBase_1.34.1           BiocFileCache_1.10.2       
#&gt;  [39] fishpond_1.2.0              R6_2.4.1                   
#&gt;  [41] markdown_1.1                GenomeInfoDb_1.22.1        
#&gt;  [43] rsvd_1.0.3                  locfit_1.5-9.4             
#&gt;  [45] AnnotationFilter_1.10.0     bitops_1.0-6               
#&gt;  [47] DelayedArray_0.12.2         assertthat_0.2.1           
#&gt;  [49] scales_1.1.0                multcomp_1.4-12            
#&gt;  [51] nnet_7.3-12                 gtable_0.3.0               
#&gt;  [53] npsurv_0.4-0                globals_0.12.5             
#&gt;  [55] sandwich_2.5-1              ensembldb_2.10.2           
#&gt;  [57] rlang_0.4.5                 splines_3.6.2              
#&gt;  [59] rtracklayer_1.46.0          lazyeval_0.2.2             
#&gt;  [61] acepack_1.4.1               dichromat_2.0-0            
#&gt;  [63] checkmate_2.0.0             formattable_0.2.0.1        
#&gt;  [65] yaml_2.2.1                  reshape2_1.4.3             
#&gt;  [67] GenomicFeatures_1.38.2      crosstalk_1.1.0.1          
#&gt;  [69] backports_1.1.5             Hmisc_4.4-0                
#&gt;  [71] tools_3.6.2                 ggplot2_3.3.0              
#&gt;  [73] ellipsis_0.3.0              gplots_3.0.3               
#&gt;  [75] RColorBrewer_1.1-2          BiocGenerics_0.32.0        
#&gt;  [77] ggridges_0.5.2              TFisher_0.2.0              
#&gt;  [79] Rcpp_1.0.3                  plyr_1.8.6                 
#&gt;  [81] base64enc_0.1-3             progress_1.2.2             
#&gt;  [83] zlibbioc_1.32.0             purrr_0.3.3                
#&gt;  [85] RCurl_1.98-1.1              prettyunits_1.1.1          
#&gt;  [87] rpart_4.1-15                openssl_1.4.1              
#&gt;  [89] pbapply_1.4-2               cowplot_1.0.0              
#&gt;  [91] S4Vectors_0.24.3            zoo_1.8-7                  
#&gt;  [93] SummarizedExperiment_1.16.1 grr_0.9.5                  
#&gt;  [95] ggrepel_0.8.2               cluster_2.1.0              
#&gt;  [97] fs_1.4.1                    magrittr_1.5               
#&gt;  [99] data.table_1.12.8           lmtest_0.9-37              
#&gt; [101] RANN_2.6.1                  mvtnorm_1.1-0              
#&gt; [103] ProtGenerics_1.18.0         fitdistrplus_1.0-14        
#&gt; [105] matrixStats_0.56.0          patchwork_1.0.0            
#&gt; [107] hms_0.5.3                   lsei_1.2-0                 
#&gt; [109] mime_0.9                    evaluate_0.14              
#&gt; [111] XML_3.99-0.3                jpeg_0.1-8.1               
#&gt; [113] IRanges_2.20.2              gridExtra_2.3              
#&gt; [115] compiler_3.6.2              biomaRt_2.42.1             
#&gt; [117] tibble_3.0.0                KernSmooth_2.23-16         
#&gt; [119] crayon_1.3.4                htmltools_0.5.0            
#&gt; [121] Formula_1.2-3               tidyr_1.0.2                
#&gt; [123] DBI_1.1.0                   dbplyr_1.4.2               
#&gt; [125] MASS_7.3-51.4               rappdirs_0.3.1             
#&gt; [127] Matrix_1.2-18               cli_2.0.2                  
#&gt; [129] gdata_2.18.0                parallel_3.6.2             
#&gt; [131] Gviz_1.30.3                 metap_1.3                  
#&gt; [133] igraph_1.2.5                GenomicRanges_1.38.0       
#&gt; [135] pkgconfig_2.0.3             pkgdown_1.5.1              
#&gt; [137] sn_1.6-1                    GenomicAlignments_1.22.1   
#&gt; [139] numDeriv_2016.8-1.1         foreign_0.8-72             
#&gt; [141] plotly_4.9.2.1              multtest_2.42.0            
#&gt; [143] XVector_0.26.0              bibtex_0.4.2.2             
#&gt; [145] stringr_1.4.0               VariantAnnotation_1.32.0   
#&gt; [147] digest_0.6.23               tsne_0.1-3                 
#&gt; [149] sctransform_0.2.1           RcppAnnoy_0.0.16           
#&gt; [151] Biostrings_2.54.0           rmarkdown_2.1              
#&gt; [153] leiden_0.3.3                htmlTable_1.13.3           
#&gt; [155] uwot_0.1.8                  edgeR_3.28.1               
#&gt; [157] curl_4.3                    Rsamtools_2.2.3            
#&gt; [159] gtools_3.8.2                lifecycle_0.2.0            
#&gt; [161] nlme_3.1-142                jsonlite_1.6.1             
#&gt; [163] stageR_1.8.0                viridisLite_0.3.0          
#&gt; [165] desc_1.2.0                  askpass_1.1                
#&gt; [167] limma_3.42.2                BSgenome_1.54.0            
#&gt; [169] fansi_0.4.1                 pillar_1.4.3               
#&gt; [171] lattice_0.20-38             plotrix_3.7-7              
#&gt; [173] httr_1.4.1                  survival_3.1-8             
#&gt; [175] glue_1.4.1                  png_0.1-7                  
#&gt; [177] bit_1.1-15.2                stringi_1.4.5              
#&gt; [179] blob_1.2.1                  latticeExtra_0.6-29        
#&gt; [181] caTools_1.18.0              memoise_1.1.0              
#&gt; [183] dplyr_0.8.5                 irlba_2.3.3                
#&gt; [185] future.apply_1.4.0          ape_5.3</code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Tobias Tekath.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
