<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Pre-processing of the Hoffman et al. human single-cell RNA-seq data • DTUrtle</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><!-- docsearch --><script src="../docsearch.js"></script><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.3/docsearch.min.css" integrity="sha256-QOSRU/ra9ActyXkIBbiIB144aDBdtvXBcNc3OTNuX/Q=" crossorigin="anonymous">
<link href="../docsearch.css" rel="stylesheet">
<script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script><meta property="og:title" content="Pre-processing of the Hoffman et al. human single-cell RNA-seq data">
<meta property="og:description" content="Exemplified pre-processing of human single-cell RNA-seq data (Illumina SureCell).">
<meta property="og:image" content="https://tobitekath.github.io/DTUrtle/logo.svg">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">DTUrtle</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.8.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Functions and data</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/index.html">Vignette overview</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Single-cell</li>
    <li>
      <a></a>
    </li>
    <li class="dropdown-header">Illumina SureCell</li>
    <li>
      <a href="../articles/Hoffman_human_single-cell_preprocess.html">Human single-cell RNA-seq pre-processing</a>
    </li>
    <li>
      <a href="../articles/Hoffman_human_single-cell_analysis.html">Human single-cell RNA-seq analysis</a>
    </li>
    <li>
      <a></a>
    </li>
    <li class="dropdown-header">10X Chromium</li>
    <li>
      <a href="../articles/Tabular_Muris_mouse_single-cell_preprocess.html">Mouse Tabular Muris single-cell RNA-seq pre-processing</a>
    </li>
    <li>
      <a href="../articles/Tabular_Muris_mouse_single-cell_analysis.html">Mouse Tabular Muris single-cell RNA-seq analysis</a>
    </li>
    <li>
      <a></a>
    </li>
    <li class="dropdown-header">Smart-seq2</li>
    <li>
      <a href="../articles/Wuidart_mouse_single-cell_preprocess.html">Mouse single-cell RNA-seq pre-processing</a>
    </li>
    <li>
      <a href="../articles/Wuidart_mouse_single-cell_analysis.html">Mouse single-cell RNA-seq analysis</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Bulk</li>
    <li>
      <a href="../articles/Hoffman_human_bulk_preprocess.html">Human bulk RNA-seq pre-processing</a>
    </li>
    <li>
      <a href="../articles/Hoffman_human_bulk_analysis.html">Human bulk RNA-seq analysis</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/TobiTekath/DTUrtle/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
<form class="navbar-form navbar-right hidden-xs hidden-sm" role="search">
        <div class="form-group">
          <input type="search" class="form-control" name="search-input" id="search-input" placeholder="Search..." aria-label="Search for..." autocomplete="off">
</div>
      </form>
      
    </div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><link href="Hoffman_human_single-cell_preprocess_files/anchor-sections-1.0/anchor-sections.css" rel="stylesheet">
<script src="Hoffman_human_single-cell_preprocess_files/anchor-sections-1.0/anchor-sections.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Pre-processing of the Hoffman et al. human single-cell RNA-seq data</h1>
                        <h4 class="author">Tobias Tekath</h4>
            
            <h4 class="date">2021-07-09</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/TobiTekath/DTUrtle/blob/master/vignettes/Hoffman_human_single-cell_preprocess.Rmd"><code>vignettes/Hoffman_human_single-cell_preprocess.Rmd</code></a></small>
      <div class="hidden name"><code>Hoffman_human_single-cell_preprocess.Rmd</code></div>

    </div>

    
    
<p>This vignette exemplifies the pre-processing of <strong>single-cell RNA-seq data for analysis with DTUrtle</strong>. The data used in this vignette is publicly available as <em>Bioproject PRJNA594939</em> and the used <em>FASTQ</em>-files can be downloaded from <a href="https://www.ebi.ac.uk/ena/browser/view/PRJNA594939">here</a>. The corresponding publication from Hoffman et al. based on this data set can be found <a href="https://doi.org/10.1038/s42003-020-0837-0">here</a>.</p>
<p>For this vignette we focus on two of the available single-cell RNA-seq samples:</p>
<ul>
<li>
<a href="https://www.ebi.ac.uk/ena/browser/view/SAMN13541133">SAMN13541133</a> which represents a control sample (18 hour treatment with ethanol (EtOH))</li>
<li>
<a href="https://www.ebi.ac.uk/ena/browser/view/SAMN13541131">SAMN13541131</a> which represents a sample after 2 hours of Dexamethasone treatment (Dex2hr)</li>
</ul>
<p>The <em>FASTQ</em>-files can be directly obtained from ENA, alternatively they are also available as <em>SRA</em>-files from <a href="https://www.ncbi.nlm.nih.gov/bioproject/PRJNA594939">GEO</a>, which can be converted to <em>FASTQ</em>-format.</p>
<p>After downloading a <strong>MD5-check</strong> is strongly encouraged.</p>
<div id="preparing-fastq-files" class="section level2">
<h2 class="hasAnchor">
<a href="#preparing-fastq-files" class="anchor"></a>Preparing FASTQ-files</h2>
<p>For this vignette, it is assumed that the <em>FASTQ</em>-files mentioned above have been downloaded to a directory called <code>samples</code>.</p>
<p>After downloading the <em>FASTQ</em>-files, the first step is to rename the files and split them into two files each. This data set was produced using a Illumina SureCell 3’ kit, which produces paired end reads (as all current single-cell protocols). For unknown reasons the original two <em>FASTQ</em>-files per sample have been appended for / during upload. More information about the SureCell format can be found <a href="https://teichlab.github.io/scg_lib_structs/methods_html/SureCell.html">here</a>.</p>
<p>The read files can be split with the following <em>bash</em> commands:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash"><code class="sourceCode bash"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="bu">cd</span> <span class="va">$PATH_TO_ANAYLSIS_ROOT_DIRECTORY</span>$/samples</a>
<a class="sourceLine" id="cb1-2" data-line-number="2"></a>
<a class="sourceLine" id="cb1-3" data-line-number="3"><span class="co">#rename</span></a>
<a class="sourceLine" id="cb1-4" data-line-number="4"><span class="fu">mv</span> SRR10669459.fastq.gz sc_EtOH.fastq.gz</a>
<a class="sourceLine" id="cb1-5" data-line-number="5"><span class="fu">mv</span> SRR10669461.fastq.gz sc_Dex2hr.fastq.gz</a>
<a class="sourceLine" id="cb1-6" data-line-number="6"></a>
<a class="sourceLine" id="cb1-7" data-line-number="7"><span class="co">#unpack the files</span></a>
<a class="sourceLine" id="cb1-8" data-line-number="8"><span class="fu">gunzip</span> sc_*.fastq.gz </a>
<a class="sourceLine" id="cb1-9" data-line-number="9"></a>
<a class="sourceLine" id="cb1-10" data-line-number="10"><span class="co">#split</span></a>
<a class="sourceLine" id="cb1-11" data-line-number="11"><span class="co"># the line numbers can be computed grepping the first linker sequence </span></a>
<a class="sourceLine" id="cb1-12" data-line-number="12"><span class="co"># or counting the number of '1'  at the end of the read name.</span></a>
<a class="sourceLine" id="cb1-13" data-line-number="13"><span class="fu">split</span> -l 1215291200 --numeric-suffixes=1 --additional-suffix=.fastq.gz sc_EtOH.fastq sc_EtOH_</a>
<a class="sourceLine" id="cb1-14" data-line-number="14"><span class="fu">split</span> -l 1394785580 --numeric-suffixes=1 --additional-suffix=.fastq.gz sc_Dex2hr.fastq sc_Dex2hr_</a></code></pre></div>
<p>Unfortunately, the read names do not directly match between the samples. The downstream tools would refuse to analyze samples without matching read names. We can overcome this with the following <em>bash</em> commands:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash"><code class="sourceCode bash"><a class="sourceLine" id="cb2-1" data-line-number="1"><span class="co">#create files with matching names</span></a>
<a class="sourceLine" id="cb2-2" data-line-number="2"><span class="fu">cat</span> sc_EtOH_2.fastq <span class="kw">|</span> <span class="fu">sed</span> -e <span class="st">'s/^\(@SRR10669459.\)\(.*\) /echo \1$((\2-303822800)) /e'</span> <span class="op">&gt;</span> sc_EtOH_2_matched.fastq</a>
<a class="sourceLine" id="cb2-3" data-line-number="3"><span class="fu">cat</span> sc_Dex2hr_2.fastq <span class="kw">|</span> <span class="fu">sed</span> -e <span class="st">'s/^\(@SRR10669461.\)\(.*\) /echo \1$((\2-348696395)) /e'</span> <span class="op">&gt;</span> sc_Dex2hr_2_matched.fastq</a>
<a class="sourceLine" id="cb2-4" data-line-number="4"></a>
<a class="sourceLine" id="cb2-5" data-line-number="5"><span class="co">#pack again</span></a>
<a class="sourceLine" id="cb2-6" data-line-number="6"><span class="fu">gzip</span> sc_*_1.fastq</a>
<a class="sourceLine" id="cb2-7" data-line-number="7"><span class="fu">gzip</span> sc_*_matched.fastq</a>
<a class="sourceLine" id="cb2-8" data-line-number="8"></a>
<a class="sourceLine" id="cb2-9" data-line-number="9"><span class="co">#optionally: remove temporary files</span></a>
<a class="sourceLine" id="cb2-10" data-line-number="10"><span class="fu">rm</span> *.fastq</a>
<a class="sourceLine" id="cb2-11" data-line-number="11"></a>
<a class="sourceLine" id="cb2-12" data-line-number="12"><span class="co">#optionally: remove concatenated files</span></a>
<a class="sourceLine" id="cb2-13" data-line-number="13"><span class="fu">rm</span> sc_EtOH.fastq.gz </a>
<a class="sourceLine" id="cb2-14" data-line-number="14"><span class="fu">rm</span> sc_Dex2hr.fastq.gz</a></code></pre></div>
<p>There should be four packed <em>FASTQ</em>-files present in the folder - two per sample.</p>
</div>
<div id="determining-valid-cellular-barcodes" class="section level2">
<h2 class="hasAnchor">
<a href="#determining-valid-cellular-barcodes" class="anchor"></a>Determining valid cellular barcodes</h2>
<p>As with most single-cell data, the first analysis step is the determination of valid cellular barcodes. As the Illumina SureCell 3’ protocol was used, the R1 reads only contain barcode and UMI information.</p>
<p>To error-correct the cell barcodes we use <code>umi_tools</code> (Version 1.0.1). The read data was created using the Illumina SureCell protocol, which uses a quite complex barcode / UMI pattern. Luckily, <code>umi_tools</code> can process arbitrary complex barcode / UMI structures via regular expressions. We allow correction of barcodes with up to one error and allow for an edit distance of 1 in each linker segment with the following <em>bash</em> code:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash"><code class="sourceCode bash"><a class="sourceLine" id="cb3-1" data-line-number="1"><span class="co">#installation in conda evironment</span></a>
<a class="sourceLine" id="cb3-2" data-line-number="2"><span class="ex">conda</span> install -c bioconda umi_tools </a>
<a class="sourceLine" id="cb3-3" data-line-number="3"></a>
<a class="sourceLine" id="cb3-4" data-line-number="4"><span class="co">#create folder and change path</span></a>
<a class="sourceLine" id="cb3-5" data-line-number="5"><span class="fu">mkdir</span> ../umi_tools</a>
<a class="sourceLine" id="cb3-6" data-line-number="6"><span class="bu">cd</span> ../umi_tools</a>
<a class="sourceLine" id="cb3-7" data-line-number="7"></a>
<a class="sourceLine" id="cb3-8" data-line-number="8"><span class="co">#collect and error correct barcode / UMI - `subset-reads` and `set-cell-number` have been chosen high to not exclude any relevant information</span></a>
<a class="sourceLine" id="cb3-9" data-line-number="9"><span class="ex">umi_tools</span> whitelist --ed-above-threshold=<span class="st">'correct'</span> --subset-reads=1000000000 --error-correct-threshold=1 --set-cell-number=10000 --extract-method=<span class="st">'regex'</span> -p <span class="st">'(?P&lt;discard_1&gt;.{0,20})(?P&lt;cell_1&gt;.{6})(?P&lt;discard_2&gt;TAGCCATCGCATTGC){e&lt;=1}(?P&lt;cell_2&gt;.{6})(?P&lt;discard_3&gt;TACCTCTGAGCTGAA){e&lt;=1}(?P&lt;cell_3&gt;.{6})(?P&lt;discard_4&gt;ACG)(?P&lt;umi_1&gt;.{8})(?P&lt;discard_5&gt;GAC).*'</span> -I ../samples/sc_EtOH_1.fastq.gz -S etoh_whitelist.tsv</a>
<a class="sourceLine" id="cb3-10" data-line-number="10"></a>
<a class="sourceLine" id="cb3-11" data-line-number="11"><span class="ex">umi_tools</span> whitelist --ed-above-threshold=<span class="st">'correct'</span> --subset-reads=1000000000 --error-correct-threshold=1 --set-cell-number=10000 --extract-method=<span class="st">'regex'</span> -p <span class="st">'(?P&lt;discard_1&gt;.{0,20})(?P&lt;cell_1&gt;.{6})(?P&lt;discard_2&gt;TAGCCATCGCATTGC){e&lt;=1}(?P&lt;cell_2&gt;.{6})(?P&lt;discard_3&gt;TACCTCTGAGCTGAA){e&lt;=1}(?P&lt;cell_3&gt;.{6})(?P&lt;discard_4&gt;ACG)(?P&lt;umi_1&gt;.{8})(?P&lt;discard_5&gt;GAC).*'</span> -I ../samples/sc_Dex2hr_1.fastq.gz -S dex2hr_whitelist.tsv</a></code></pre></div>
<p>The <em>TSV</em> files contain all found barcode sequences in the samples, together with possible erroneous sequenced barcodes (less than 2 edit distance away) which will be corrected.</p>
<p>In the publication, only 400 cells of each sample were used in the analysis. For simplicity, we stick to this process. We can get the 400 valid cell barcodes for each sample from the supplementary count files from <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE141834">GEO</a>: <code>GSE141834_scRNAseq_rawCounts.txt.gz</code></p>
<p>Assume you have downloaded and unpacked the mentioned file. Lists of the valid cell barcodes can be created with this <em>R</em> code:</p>
<div class="sourceCode" id="cb4"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/base/getwd.html">setwd</a></span><span class="op">(</span><span class="st">"$PATH_TO_ANAYLSIS_ROOT_DIRECTORY$/umi_tools"</span><span class="op">)</span>

<span class="co">#get valid barcodes from raw count table</span>
<span class="va">raw_expr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.table</a></span><span class="op">(</span><span class="st">"GSE141834_scRNAseq_rawCounts.txt"</span>, header<span class="op">=</span><span class="cn">TRUE</span>, sep<span class="op">=</span><span class="st">"\t"</span><span class="op">)</span>
<span class="va">etoh_selected_cells</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span><span class="op">(</span><span class="st">".*_"</span>,<span class="st">""</span>,<span class="fu"><a href="https://rdrr.io/r/base/grep.html">grep</a></span><span class="op">(</span><span class="st">"Dex.00.2_"</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">raw_expr</span><span class="op">)</span>, value <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>
<span class="va">dex2hr_selected_cells</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span><span class="op">(</span><span class="st">".*_"</span>,<span class="st">""</span>,<span class="fu"><a href="https://rdrr.io/r/base/grep.html">grep</a></span><span class="op">(</span><span class="st">"Dex.02.2_"</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">raw_expr</span><span class="op">)</span>, value <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>

<span class="co">#read-in created umi_tools whitelists</span>
<span class="va">etoh_white</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.table</a></span><span class="op">(</span><span class="st">"etoh_whitelist.tsv"</span>, stringsAsFactors <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="va">dex2hr_white</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.table</a></span><span class="op">(</span><span class="st">"dex2hr_whitelist.tsv"</span>, stringsAsFactors <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>

<span class="co">#write subset to file</span>
<span class="va">etoh_white_sub</span> <span class="op">&lt;-</span> <span class="va">etoh_white</span><span class="op">[</span><span class="va">etoh_white</span><span class="op">$</span><span class="va">V1</span> <span class="op">%in%</span> <span class="va">etoh_selected_cells</span>,<span class="op">]</span>
<span class="fu"><a href="https://rdrr.io/r/utils/write.table.html">write.table</a></span><span class="op">(</span><span class="va">etoh_white_sub</span>, <span class="st">"etoh_whitelist_sub.tsv"</span>, sep <span class="op">=</span> <span class="st">"\t"</span>, quote <span class="op">=</span> <span class="cn">FALSE</span>, row.names <span class="op">=</span> <span class="cn">FALSE</span>, col.names <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>

<span class="va">dex2hr_white_sub</span> <span class="op">&lt;-</span> <span class="va">dex2hr_white</span><span class="op">[</span><span class="va">dex2hr_white</span><span class="op">$</span><span class="va">V1</span> <span class="op">%in%</span> <span class="va">dex2hr_selected_cells</span>,<span class="op">]</span>
<span class="fu"><a href="https://rdrr.io/r/utils/write.table.html">write.table</a></span><span class="op">(</span><span class="va">dex2hr_white_sub</span>, <span class="st">"dex2hr_whitelist_sub.tsv"</span>, sep <span class="op">=</span> <span class="st">"\t"</span>, quote <span class="op">=</span> <span class="cn">FALSE</span>, row.names <span class="op">=</span> <span class="cn">FALSE</span>, col.names <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></pre></div>
<p>The subsetted <em>TSV</em> files contain only the 400 valid barcodes per cell, together with the barcodes that shall be corrected.</p>
</div>
<div id="extract-and-correct-barcodes-umis" class="section level2">
<h2 class="hasAnchor">
<a href="#extract-and-correct-barcodes-umis" class="anchor"></a>Extract and correct barcodes &amp; UMIs</h2>
<p>The actual barcode extraction and correction is again performed with <code>umi_tools</code>. In this process, only the reads of the R2 files are kept, that had a matching barcode in the R1 file. The barcode and UMI sequence are added to the sequence identifier line of each read in the new files. The not matching parts of the R1 reads are kept in the new R1 file. The <code>umi_tools</code> <em>bash</em> command looks like this:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode bash"><code class="sourceCode bash"><a class="sourceLine" id="cb5-1" data-line-number="1"><span class="ex">umi_tools</span> extract --filter-cell-barcode --error-correct-cell --whitelist=etoh_whitelist_sub.tsv --extract-method=regex -p <span class="st">'(?P&lt;discard_1&gt;.{0,20})(?P&lt;cell_1&gt;.{6})(?P&lt;discard_2&gt;TAGCCATCGCATTGC){e&lt;=1}(?P&lt;cell_2&gt;.{6})(?P&lt;discard_3&gt;TACCTCTGAGCTGAA){e&lt;=1}(?P&lt;cell_3&gt;.{6})(?P&lt;discard_4&gt;ACG)(?P&lt;umi_1&gt;.{8})(?P&lt;discard_5&gt;GAC).*'</span> -I ../samples/sc_EtOH_1.fastq.gz -S sc_EtOH_1_extracted.fastq.gz --read2-in ../samples/sc_EtOH_2_matched.fastq.gz --read2-out sc_EtOH_2_extracted.fastq.gz</a>
<a class="sourceLine" id="cb5-2" data-line-number="2"></a>
<a class="sourceLine" id="cb5-3" data-line-number="3"></a>
<a class="sourceLine" id="cb5-4" data-line-number="4"><span class="ex">umi_tools</span> extract --filter-cell-barcode --error-correct-cell --whitelist=dex2hr_whitelist_sub.tsv --extract-method=regex -p <span class="st">'(?P&lt;discard_1&gt;.{0,20})(?P&lt;cell_1&gt;.{6})(?P&lt;discard_2&gt;TAGCCATCGCATTGC){e&lt;=1}(?P&lt;cell_2&gt;.{6})(?P&lt;discard_3&gt;TACCTCTGAGCTGAA){e&lt;=1}(?P&lt;cell_3&gt;.{6})(?P&lt;discard_4&gt;ACG)(?P&lt;umi_1&gt;.{8})(?P&lt;discard_5&gt;GAC).*'</span> -I ../samples/sc_Dex2hr_1.fastq.gz -S sc_Dex2hr_1_extracted.fastq.gz --read2-in ../samples/sc_Dex2hr_2_matched.fastq.gz --read2-out sc_Dex2hr_2_extracted.fastq.gz</a></code></pre></div>
<p>The <code>umi_tools</code> folder should now contain four gzipped <em>FASTQ</em> files, two for each sample.</p>
<hr>
<p>We plan to use <code>Alevin</code> for transcript level quantification, which does not support the Illumina SureCell protocol by default. We can overcome this limitation by slightly editing the new R1 files with some <em>bash</em> code:</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode bash"><code class="sourceCode bash"><a class="sourceLine" id="cb6-1" data-line-number="1"><span class="fu">zgrep</span> <span class="st">"^@"</span> sc_EtOH_1_extracted.fastq.gz <span class="kw">|</span>  <span class="fu">awk</span> -F <span class="st">'_'</span> -v OFS=<span class="st">'\n'</span> <span class="st">'{split($3,a," ");print $0,$2a[1],"+","KKKKKKKKKKKKKKKKKKKKKKKKKK"}'</span> <span class="kw">|</span> <span class="fu">gzip</span> <span class="op">&gt;</span> sc_EtOH_1_altered.fastq.gz</a>
<a class="sourceLine" id="cb6-2" data-line-number="2"></a>
<a class="sourceLine" id="cb6-3" data-line-number="3"></a>
<a class="sourceLine" id="cb6-4" data-line-number="4"><span class="fu">zgrep</span> <span class="st">"^@"</span> sc_Dex2hr_1_extracted.fastq.gz <span class="kw">|</span>  <span class="fu">awk</span> -F <span class="st">'_'</span> -v OFS=<span class="st">'\n'</span> <span class="st">'{split($3,a," ");print $0,$2a[1],"+","KKKKKKKKKKKKKKKKKKKKKKKKKK"}'</span> <span class="kw">|</span> <span class="fu">gzip</span> <span class="op">&gt;</span> sc_Dex2hr_1_altered.fastq.gz</a>
<a class="sourceLine" id="cb6-5" data-line-number="5"></a>
<a class="sourceLine" id="cb6-6" data-line-number="6"><span class="co">#optional: remove `extracted` R1 files</span></a>
<a class="sourceLine" id="cb6-7" data-line-number="7"><span class="fu">rm</span> sc_*_1_extracted.fastq.gz</a></code></pre></div>
<p>With this command, we copied the barcode and UMI sequence of each reads sequence identifier to the actual raw sequence line. We also added some quality values for the bases (the actual value is not relevant).</p>
<p>There should now be an <code>altered</code> R1 <em>FASTQ</em> file per sample, together with an <code>extracted</code> R2 <em>FASTQ</em> file.</p>
</div>
<div id="quantification-with-alevin" class="section level2">
<h2 class="hasAnchor">
<a href="#quantification-with-alevin" class="anchor"></a>Quantification with Alevin</h2>
<p><code>Alevin</code> is the single-cell counterpart of <code>Salmon</code>, one of the standard tools for bulk RNA-seq quantification. It is also integrated into the <code>Salmon</code> (Version 1.1.0) package. As most of the available transcript quantifiers, <code>Alevin/Salmon</code> do not perform a standard genomic alignment, but perform a quasi-mapping directly to the transcriptome. Reads, that could be originating from multiple transcript isoforms, are assigned to equivalence classes, where actual counts are later derived by expectation maximization algorithms.</p>
<p>To run the <code>Alevin</code> quantification, we first need a transcriptomic index. For this example, we will use the <a href="ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_34/gencode.v34.transcripts.fa.gz">transcript sequences</a> and <a href="ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_34/gencode.v34.annotation.gtf.gz">comprehensive gene annotation</a> from <a href="https://www.gencodegenes.org/">Gencode</a> <a href="https://www.gencodegenes.org/human/release_34.html">Version 34</a>.</p>
<p>Assume you have downloaded and unpacked the above mentioned files in the analysis root directory. To build the actual index we run this <em>bash</em> code:</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode bash"><code class="sourceCode bash"><a class="sourceLine" id="cb7-1" data-line-number="1"><span class="co">#installation in conda evironment</span></a>
<a class="sourceLine" id="cb7-2" data-line-number="2"><span class="ex">conda</span> install salmon</a>
<a class="sourceLine" id="cb7-3" data-line-number="3"></a>
<a class="sourceLine" id="cb7-4" data-line-number="4"><span class="co">#create folder and change directory</span></a>
<a class="sourceLine" id="cb7-5" data-line-number="5"><span class="fu">mkdir</span> ../alevin</a>
<a class="sourceLine" id="cb7-6" data-line-number="6"><span class="bu">cd</span> ../alevin</a>
<a class="sourceLine" id="cb7-7" data-line-number="7"></a>
<a class="sourceLine" id="cb7-8" data-line-number="8"><span class="co">#build index - `p` specifies the number of threads to use</span></a>
<a class="sourceLine" id="cb7-9" data-line-number="9"><span class="ex">salmon</span> index -t ../gencode.v34.transcripts.fa --gencode -p 4 -i index</a></code></pre></div>
<p>To perform the actual quantification, we need one last file specifying the transcript to gene mapping. <code>Alevin</code>, unlike <code>Salmon</code>, aggregates the quantification counts to gene level by default. For our DTU analysis we require transcript level counts, so we will only provide a <code>transcript_id</code> to <code>transcript_name</code> mapping. This could also be a <code>transcript_id</code> to <code>transcript_id</code> mapping file (so each id mapping to itself).</p>
<p>We can create such a mapping file with the help of the <code>DTUrtle</code> package in <em>R</em>:</p>
<div class="sourceCode" id="cb8"><pre class="downlit">
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/TobiTekath/DTUrtle">DTUrtle</a></span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/getwd.html">setwd</a></span><span class="op">(</span><span class="st">"$PATH_TO_ANAYLSIS_ROOT_DIRECTORY$/alevin"</span><span class="op">)</span>

<span class="co">#import information from the GTF file</span>
<span class="va">tx2gene</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/import_gtf.html">import_gtf</a></span><span class="op">(</span>gtf_file <span class="op">=</span> <span class="st">"../gencode.v34.annotation.gtf"</span><span class="op">)</span>

  <span class="co">##optional for later name consistency:</span>
  <span class="va">tx2gene</span><span class="op">$</span><span class="va">gene_name</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/one_to_one_mapping.html">one_to_one_mapping</a></span><span class="op">(</span>name <span class="op">=</span> <span class="va">tx2gene</span><span class="op">$</span><span class="va">gene_name</span>, 
                                          id <span class="op">=</span> <span class="va">tx2gene</span><span class="op">$</span><span class="va">gene_id</span><span class="op">)</span>
  <span class="va">tx2gene</span><span class="op">$</span><span class="va">transcript_name</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/one_to_one_mapping.html">one_to_one_mapping</a></span><span class="op">(</span>name <span class="op">=</span> <span class="va">tx2gene</span><span class="op">$</span><span class="va">transcript_name</span>, 
                                                id <span class="op">=</span> <span class="va">tx2gene</span><span class="op">$</span><span class="va">transcript_id</span><span class="op">)</span>
  

<span class="fu"><a href="https://rdrr.io/r/utils/write.table.html">write.table</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">tx2gene</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"transcript_id"</span>, <span class="st">"transcript_name"</span><span class="op">)</span><span class="op">]</span>, 
            file <span class="op">=</span> <span class="st">"txmap.tsv"</span>, sep <span class="op">=</span> <span class="st">"\t"</span>, quote <span class="op">=</span> <span class="cn">FALSE</span>, 
            row.names <span class="op">=</span> <span class="cn">FALSE</span>, col.names <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></pre></div>
<p>The actual quantification can be performed in <em>bash</em> with:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode bash"><code class="sourceCode bash"><a class="sourceLine" id="cb9-1" data-line-number="1"><span class="ex">salmon</span> alevin -l A -i index/ -1 ../umi_tools/sc_EtOH_1_altered.fastq.gz -2 ../umi_tools/sc_EtOH_2_extracted.fastq.gz --keepCBFraction 1 --end 5 --umiLength 8 --barcodeLength 18 --dumpFeatures --tgMap txmap.tsv -o sc_EtOH</a>
<a class="sourceLine" id="cb9-2" data-line-number="2"></a>
<a class="sourceLine" id="cb9-3" data-line-number="3"><span class="ex">salmon</span> alevin -l A -i index/ -1 ../umi_tools/sc_Dex2hr_1_altered.fastq.gz -2 ../umi_tools/sc_Dex2hr_2_extracted.fastq.gz --keepCBFraction 1 --end 5 --umiLength 8 --barcodeLength 18 --dumpFeatures --tgMap txmap.tsv -o sc_Dex2hr</a></code></pre></div>
<hr>
<p><strong>This concludes the preprocessing of the data. Please see the corresponding analysis vignette for an usage example of DTUrtle.</strong></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Tobias Tekath.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  
<script src="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.js" integrity="sha256-GKvGqXDznoRYHCwKXGnuchvKSwmx9SRMrZOTh2g4Sb0=" crossorigin="anonymous"></script><script>
  docsearch({
    
    
    apiKey: 'a6c9f43b010fdd4b2e64792389fd67d2',
    indexName: 'dturtle',
    inputSelector: 'input#search-input.form-control',
    transformData: function(hits) {
      return hits.map(function (hit) {
        hit.url = updateHitURL(hit);
        return hit;
      });
    }
  });
</script>
</body>
</html>
