<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Pre-processing of Tabular Muris mouse single-cell RNA-seq data â€¢ DTUrtle</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><!-- docsearch --><script src="../docsearch.js"></script><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.3/docsearch.min.css" integrity="sha256-QOSRU/ra9ActyXkIBbiIB144aDBdtvXBcNc3OTNuX/Q=" crossorigin="anonymous">
<link href="../docsearch.css" rel="stylesheet">
<script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script><meta property="og:title" content="Pre-processing of Tabular Muris mouse single-cell RNA-seq data">
<meta property="og:description" content="Exemplified pre-processing of mouse single-cell RNA-seq data (10X Chromium).">
<meta property="og:image" content="https://tobitekath.github.io/DTUrtle/logo.svg">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">DTUrtle</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Functions and data</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/index.html">Vignette overview</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Single-cell</li>
    <li>
      <a></a>
    </li>
    <li class="dropdown-header">Illumina SureCell</li>
    <li>
      <a href="../articles/Hoffman_human_single-cell_preprocess.html">Human single-cell RNA-seq pre-processing</a>
    </li>
    <li>
      <a href="../articles/Hoffman_human_single-cell_analysis.html">Human single-cell RNA-seq analysis</a>
    </li>
    <li>
      <a></a>
    </li>
    <li class="dropdown-header">10X Chromium</li>
    <li>
      <a href="../articles/Tabular_Muris_mouse_single-cell_preprocess.html">Mouse Tabular Muris single-cell RNA-seq pre-processing</a>
    </li>
    <li>
      <a href="../articles/Tabular_Muris_mouse_single-cell_analysis.html">Mouse Tabular Muris single-cell RNA-seq analysis</a>
    </li>
    <li>
      <a></a>
    </li>
    <li class="dropdown-header">Smart-seq2</li>
    <li>
      <a href="../articles/Wuidart_mouse_single-cell_preprocess.html">Mouse single-cell RNA-seq pre-processing</a>
    </li>
    <li>
      <a href="../articles/Wuidart_mouse_single-cell_analysis.html">Mouse single-cell RNA-seq analysis</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Bulk</li>
    <li>
      <a href="../articles/Hoffman_human_bulk_preprocess.html">Human bulk RNA-seq pre-processing</a>
    </li>
    <li>
      <a href="../articles/Hoffman_human_bulk_analysis.html">Human bulk RNA-seq analysis</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/TobiTekath/DTUrtle/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
<form class="navbar-form navbar-right hidden-xs hidden-sm" role="search">
        <div class="form-group">
          <input type="search" class="form-control" name="search-input" id="search-input" placeholder="Search..." aria-label="Search for..." autocomplete="off">
</div>
      </form>
      
    </div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><link href="Tabular_Muris_mouse_single-cell_preprocess_files/anchor-sections-1.0/anchor-sections.css" rel="stylesheet">
<script src="Tabular_Muris_mouse_single-cell_preprocess_files/anchor-sections-1.0/anchor-sections.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Pre-processing of Tabular Muris mouse single-cell RNA-seq data</h1>
                        <h4 class="author">Tobias Tekath</h4>
            
            <h4 class="date">2021-11-18</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/TobiTekath/DTUrtle/blob/master/vignettes/Tabular_Muris_mouse_single-cell_preprocess.Rmd"><code>vignettes/Tabular_Muris_mouse_single-cell_preprocess.Rmd</code></a></small>
      <div class="hidden name"><code>Tabular_Muris_mouse_single-cell_preprocess.Rmd</code></div>

    </div>

    
    
<p>This vignette exemplifies the pre-processing of <strong>single-cell RNA-seq data for analysis with DTUrtle</strong>. The data used in this vignette is from the Tabular Muris consortium and publicly available as <em>Bioproject PRJNA432002</em>. The used files can be downloaded from <a href="https://www.ebi.ac.uk/ena/browser/view/PRJNA432002">here</a>. The corresponding publication for the data set can be found <a href="https://doi.org/10.1038/s41586-018-0590-4">here</a>.</p>
<p>For this vignette we focus on two of the available single-cell RNA-seq samples:</p>
<ul>
<li>
<a href="https://www.ebi.ac.uk/ena/browser/view/SAMN08712875">SAMN08712875</a> which is a 10x Chromium v2 sample from the mammary gland (Mammary_Gland-10X_P7_13)</li>
<li>
<a href="https://www.ebi.ac.uk/ena/browser/view/SAMN08712876">SAMN08712876</a> which is a 10x Chromium v2 sample from the mammary gland (Mammary_Gland-10X_P7_12)</li>
</ul>
<p>There are <em>BAM</em>-files directly available from ENA, which will be converted back to <em>FASTQ</em>-files in this vignette.</p>
<p>After downloading a <strong>MD5-check</strong> is strongly encouraged.</p>
<div id="preparing-bam-files" class="section level2">
<h2 class="hasAnchor">
<a href="#preparing-bam-files" class="anchor"></a>Preparing BAM-files</h2>
<p>For this vignette, it is assumed that the <em>BAM</em>-files mentioned above have been downloaded to a directory called <code>samples</code>.</p>
<p>After downloading the <em>BAM</em>-files, which originate from an analysis with the 10x CellRanger pipeline, can be transferred back to the <em>FASTQ</em>-format without data loss. For this task we will use the 10x toolkit <code>bamtofastq</code> (Version 1.3.1), which can be obtained from <a href="https://github.com/10XGenomics/bamtofastq">GitHub</a></p>
<p>Download the <code>bamtofastq_linux</code> or <code>bamtofastq_macos</code> binaries from the releases page into the <code>samples</code> folder and continue with the following <em>bash</em> commands:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash"><code class="sourceCode bash"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="bu">cd</span> <span class="va">$PATH_TO_ANAYLSIS_ROOT_DIRECTORY</span>$/samples</a>
<a class="sourceLine" id="cb1-2" data-line-number="2"></a>
<a class="sourceLine" id="cb1-3" data-line-number="3"><span class="co">#make downloaded binary executable</span></a>
<a class="sourceLine" id="cb1-4" data-line-number="4"><span class="fu">chmod</span> +x bamtofastq_linux</a>
<a class="sourceLine" id="cb1-5" data-line-number="5"></a>
<a class="sourceLine" id="cb1-6" data-line-number="6"><span class="co">#convert BAMs back to FASTQ - increased read limit to not inflate the number of files.</span></a>
<a class="sourceLine" id="cb1-7" data-line-number="7"><span class="ex">./bamtofastq_linux</span> --reads-per-fastq=500000000 10X_P7_12.bam 10X_P7_12</a>
<a class="sourceLine" id="cb1-8" data-line-number="8"><span class="ex">./bamtofastq_linux</span> --reads-per-fastq=500000000 10X_P7_13.bam 10X_P7_13</a>
<a class="sourceLine" id="cb1-9" data-line-number="9"></a>
<a class="sourceLine" id="cb1-10" data-line-number="10"></a>
<a class="sourceLine" id="cb1-11" data-line-number="11"><span class="co">#optionally: remove BAM files</span></a>
<a class="sourceLine" id="cb1-12" data-line-number="12"><span class="fu">rm</span> 10X_P7_12.bam </a>
<a class="sourceLine" id="cb1-13" data-line-number="13"><span class="fu">rm</span> 10X_P7_13.bam</a></code></pre></div>
<p>There are now two folders with the <em>FASTQ</em>-files present in the <code>samples</code> folder. Apparently the reads originate from two sequencing lanes each - this was probably done to increase the total sequencing depth and coverage.</p>
</div>
<div id="quantification-with-alevin" class="section level2">
<h2 class="hasAnchor">
<a href="#quantification-with-alevin" class="anchor"></a>Quantification with Alevin</h2>
<p><code>Alevin</code> is the single-cell counterpart of <code>Salmon</code>, one of the standard tools for bulk RNA-seq quantification. It is also integrated into the <code>Salmon</code> (Version 1.1.0) package. As most of the available transcript quantifiers, <code>Alevin/Salmon</code> do not perform a standard genomic alignment, but perform a quasi-mapping directly to the transcriptome. Reads, that could be originating from multiple transcript isoforms, are assigned to equivalence classes, where actual counts are later derived by expectation maximization algorithms.</p>
<p>To run the <code>Alevin</code> quantification, we first need a transcriptomic index. For this example, we will use the <a href="ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M24/gencode.vM24.transcripts.fa.gz">transcript sequences</a> and <a href="ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M24/gencode.vM24.annotation.gtf.gz">comprehensive gene annotation</a> from <a href="https://www.gencodegenes.org/">Gencode</a> <a href="https://www.gencodegenes.org/mouse/release_M24.html">Version M24</a>.</p>
<p>Assume you have downloaded and unpacked the above mentioned files in the analysis root directory. To build the actual index we run this <em>bash</em> code:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash"><code class="sourceCode bash"><a class="sourceLine" id="cb2-1" data-line-number="1"><span class="co">#installation in conda evironment</span></a>
<a class="sourceLine" id="cb2-2" data-line-number="2"><span class="ex">conda</span> install salmon</a>
<a class="sourceLine" id="cb2-3" data-line-number="3"></a>
<a class="sourceLine" id="cb2-4" data-line-number="4"><span class="co">#create folder and change directory</span></a>
<a class="sourceLine" id="cb2-5" data-line-number="5"><span class="fu">mkdir</span> ../alevin</a>
<a class="sourceLine" id="cb2-6" data-line-number="6"><span class="bu">cd</span> ../alevin</a>
<a class="sourceLine" id="cb2-7" data-line-number="7"></a>
<a class="sourceLine" id="cb2-8" data-line-number="8"><span class="co">#build index - `p` specifies the number of threads to use</span></a>
<a class="sourceLine" id="cb2-9" data-line-number="9"><span class="ex">salmon</span> index -t ../gencode.vM24.transcripts.fa --gencode -p 4 -i index</a></code></pre></div>
<p>To perform the actual quantification, we need one last file specifying the transcript to gene mapping. <code>Alevin</code>, unlike <code>Salmon</code>, aggregates the quantification counts to gene level by default. For our DTU analysis we require transcript level counts, so we will only provide a <code>transcript_id</code> to <code>transcript_name</code> mapping. This could also be a <code>transcript_id</code> to <code>transcript_id</code> mapping file (so each id mapping to itself).</p>
<p>We can create such a mapping file with the help of the <code>DTUrtle</code> package in <em>R</em>:</p>
<div class="sourceCode" id="cb3"><pre class="downlit">
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/TobiTekath/DTUrtle">DTUrtle</a></span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/getwd.html">setwd</a></span><span class="op">(</span><span class="st">"$PATH_TO_ANAYLSIS_ROOT_DIRECTORY$/alevin"</span><span class="op">)</span>

<span class="co">#import information from the GTF file</span>
<span class="va">tx2gene</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/import_gtf.html">import_gtf</a></span><span class="op">(</span>gtf_file <span class="op">=</span> <span class="st">"../gencode.vM24.annotation.gtf"</span><span class="op">)</span>

  <span class="co">##optional for later name consistency:</span>
  <span class="va">tx2gene</span><span class="op">$</span><span class="va">gene_name</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/one_to_one_mapping.html">one_to_one_mapping</a></span><span class="op">(</span>name <span class="op">=</span> <span class="va">tx2gene</span><span class="op">$</span><span class="va">gene_name</span>, 
                                          id <span class="op">=</span> <span class="va">tx2gene</span><span class="op">$</span><span class="va">gene_id</span><span class="op">)</span>
  <span class="va">tx2gene</span><span class="op">$</span><span class="va">transcript_name</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/one_to_one_mapping.html">one_to_one_mapping</a></span><span class="op">(</span>name <span class="op">=</span> <span class="va">tx2gene</span><span class="op">$</span><span class="va">transcript_name</span>, 
                                                id <span class="op">=</span> <span class="va">tx2gene</span><span class="op">$</span><span class="va">transcript_id</span><span class="op">)</span>
  

<span class="fu"><a href="https://rdrr.io/r/utils/write.table.html">write.table</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">tx2gene</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"transcript_id"</span>, <span class="st">"transcript_name"</span><span class="op">)</span><span class="op">]</span>, 
            file <span class="op">=</span> <span class="st">"txmap.tsv"</span>, sep <span class="op">=</span> <span class="st">"\t"</span>, quote <span class="op">=</span> <span class="cn">FALSE</span>, 
            row.names <span class="op">=</span> <span class="cn">FALSE</span>, col.names <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></pre></div>
<p>The actual quantification can be performed in <em>bash</em> with:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode bash"><code class="sourceCode bash"><a class="sourceLine" id="cb4-1" data-line-number="1"><span class="co">#change back to sample directory for shorter command</span></a>
<a class="sourceLine" id="cb4-2" data-line-number="2"><span class="bu">cd</span> ../samples/10X_P7_12/10X_P7_12_MissingLibrary_1_H2H5YDMXX</a>
<a class="sourceLine" id="cb4-3" data-line-number="3"><span class="ex">salmon</span> alevin -l ISR -i ../../../alevin/index/ -1 bamtofastq_S1_L001_R1_001.fastq.gz bamtofastq_S1_L002_R1_001.fastq.gz -2 bamtofastq_S1_L001_R2_001.fastq.gz bamtofastq_S1_L002_R2_001.fastq.gz --chromium -p 20 --tgMap ../../../alevin/txmap.tsv --dumpFeatures -o ../../../alevin/10X_P7_12</a>
<a class="sourceLine" id="cb4-4" data-line-number="4"></a>
<a class="sourceLine" id="cb4-5" data-line-number="5"><span class="bu">cd</span> ../../10X_P7_13/10X_P7_13_MissingLibrary_1_H2H5YDMXX</a>
<a class="sourceLine" id="cb4-6" data-line-number="6"><span class="ex">salmon</span> alevin -l ISR -i ../../../alevin/index/ -1 bamtofastq_S1_L001_R1_001.fastq.gz bamtofastq_S1_L002_R1_001.fastq.gz -2 bamtofastq_S1_L001_R2_001.fastq.gz bamtofastq_S1_L002_R2_001.fastq.gz --chromium -p 20 --tgMap ../../../alevin/txmap.tsv --dumpFeatures -o ../../../alevin/10X_P7_13</a></code></pre></div>
<p>For this analysis we used the <code>Alevin</code> algorithms for cell detection and barcode sequence correction, resulting in transcript level quantification data for 3992 (10X_P7_12) and 4326 cells (10X_P7_13).</p>
<hr>
<p><strong>This concludes the preprocessing of the data. Please see the corresponding analysis vignette for an usage example of DTUrtle.</strong></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Tobias Tekath.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  
<script src="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.js" integrity="sha256-GKvGqXDznoRYHCwKXGnuchvKSwmx9SRMrZOTh2g4Sb0=" crossorigin="anonymous"></script><script>
  docsearch({
    
    
    apiKey: 'a6c9f43b010fdd4b2e64792389fd67d2',
    indexName: 'dturtle',
    inputSelector: 'input#search-input.form-control',
    transformData: function(hits) {
      return hits.map(function (hit) {
        hit.url = updateHitURL(hit);
        return hit;
      });
    }
  });
</script>
</body>
</html>
