<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Pre-processing of the Wuidart et al. mouse single-cell RNA-seq data • DTUrtle</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><!-- docsearch --><script src="../docsearch.js"></script><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.3/docsearch.min.css" integrity="sha256-QOSRU/ra9ActyXkIBbiIB144aDBdtvXBcNc3OTNuX/Q=" crossorigin="anonymous">
<link href="../docsearch.css" rel="stylesheet">
<script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script><meta property="og:title" content="Pre-processing of the Wuidart et al. mouse single-cell RNA-seq data">
<meta property="og:description" content="Exemplified pre-processing of mouse single-cell RNA-seq data (Smart-seq2).">
<meta property="og:image" content="https://tobitekath.github.io/DTUrtle/logo.svg">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">DTUrtle</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Functions and data</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/index.html">Vignette overview</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Single-cell</li>
    <li>
      <a></a>
    </li>
    <li class="dropdown-header">Illumina SureCell</li>
    <li>
      <a href="../articles/Hoffman_human_single-cell_preprocess.html">Human single-cell RNA-seq pre-processing</a>
    </li>
    <li>
      <a href="../articles/Hoffman_human_single-cell_analysis.html">Human single-cell RNA-seq analysis</a>
    </li>
    <li>
      <a></a>
    </li>
    <li class="dropdown-header">10X Chromium</li>
    <li>
      <a href="../articles/Tabular_Muris_mouse_single-cell_preprocess.html">Mouse Tabular Muris single-cell RNA-seq pre-processing</a>
    </li>
    <li>
      <a href="../articles/Tabular_Muris_mouse_single-cell_analysis.html">Mouse Tabular Muris single-cell RNA-seq analysis</a>
    </li>
    <li>
      <a></a>
    </li>
    <li class="dropdown-header">Smart-seq2</li>
    <li>
      <a href="../articles/Wuidart_mouse_single-cell_preprocess.html">Mouse single-cell RNA-seq pre-processing</a>
    </li>
    <li>
      <a href="../articles/Wuidart_mouse_single-cell_analysis.html">Mouse single-cell RNA-seq analysis</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Bulk</li>
    <li>
      <a href="../articles/Hoffman_human_bulk_preprocess.html">Human bulk RNA-seq pre-processing</a>
    </li>
    <li>
      <a href="../articles/Hoffman_human_bulk_analysis.html">Human bulk RNA-seq analysis</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/TobiTekath/DTUrtle/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
<form class="navbar-form navbar-right hidden-xs hidden-sm" role="search">
        <div class="form-group">
          <input type="search" class="form-control" name="search-input" id="search-input" placeholder="Search..." aria-label="Search for..." autocomplete="off">
</div>
      </form>
      
    </div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><link href="Wuidart_mouse_single-cell_preprocess_files/anchor-sections-1.0/anchor-sections.css" rel="stylesheet">
<script src="Wuidart_mouse_single-cell_preprocess_files/anchor-sections-1.0/anchor-sections.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Pre-processing of the Wuidart et al. mouse single-cell RNA-seq data</h1>
                        <h4 class="author">Tobias Tekath</h4>
            
            <h4 class="date">2021-10-08</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/TobiTekath/DTUrtle/blob/master/vignettes/Wuidart_mouse_single-cell_preprocess.Rmd"><code>vignettes/Wuidart_mouse_single-cell_preprocess.Rmd</code></a></small>
      <div class="hidden name"><code>Wuidart_mouse_single-cell_preprocess.Rmd</code></div>

    </div>

    
    
<p>This vignette exemplifies the analysis of <strong>single-cell RNA-seq data with DTUrtle</strong>. The data used in this vignette is publicly available as <em>Bioproject PRJNA433520</em> and the used <em>FASTQ</em>-files can be downloaded from <a href="https://www.ebi.ac.uk/ena/browser/view/PRJNA433520">here</a>. The corresponding publication from Wuidart et al. can be found <a href="https://doi.org/10.1038/s41556-018-0095-2">here</a>.</p>
<p>The <em>FASTQ</em>-files can be directly obtained from ENA, alternatively they are also available as <em>SRA</em>-files from <a href="https://www.ncbi.nlm.nih.gov/bioproject/PRJNA433520">GEO</a>, which can be converted to <em>FASTQ</em>-format.</p>
<p>As this vignette deals with single-cell RNA-seq data generated via the Smart-seq2 protocol, a (standard) sample of this Project represents reads from a single cell. We initially download all 384 available samples/cells - this can be done directly from ENA.</p>
<p>Suppose you have downloaded the Bioproject report table like <a href="https://www.ebi.ac.uk/ena/portal/api/filereport?accession=PRJNA433520&amp;result=read_run&amp;fields=study_accession,sample_accession,scientific_name,fastq_md5,fastq_ftp,sample_title&amp;format=tsv&amp;download=true">this</a> to your <code>YOUR_PATH</code> directory - selecting the columns <code>fastq_ftp</code>, <code>fastq_md5</code>, <code>sample_accession</code>, <code>sample_title</code>, <code>scientific_name</code> and <code>study_accession</code>.</p>
<p>This table contains all needed information to download and <strong>MD5-check</strong> the files - as well as to create a basic sample meta information table.</p>
<p>Downloading can be done like this:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash"><code class="sourceCode bash"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="co">#YOUR_PATH contains the downloaded Bioproject report table `filereport_read_run_PRJNA433520_tsv.txt`</span></a>
<a class="sourceLine" id="cb1-2" data-line-number="2"><span class="bu">cd</span> <span class="st">'YOUR_PATH'</span></a>
<a class="sourceLine" id="cb1-3" data-line-number="3"><span class="fu">mkdir</span> samples</a>
<a class="sourceLine" id="cb1-4" data-line-number="4"></a>
<a class="sourceLine" id="cb1-5" data-line-number="5"><span class="co">#create a file of only the ftp-links for download:</span></a>
<a class="sourceLine" id="cb1-6" data-line-number="6"><span class="fu">sed</span> 1d filereport_read_run_PRJNA433520_tsv.txt <span class="kw">|</span> <span class="fu">cut</span> -f 5 <span class="kw">|</span> <span class="fu">sed</span> -e <span class="st">'s/^/ftp:\/\//'</span> <span class="op">&gt;</span> samples/wget_links.txt</a>
<a class="sourceLine" id="cb1-7" data-line-number="7"></a>
<a class="sourceLine" id="cb1-8" data-line-number="8"><span class="co">#download with wget</span></a>
<a class="sourceLine" id="cb1-9" data-line-number="9"><span class="bu">cd</span> samples</a>
<a class="sourceLine" id="cb1-10" data-line-number="10"><span class="fu">wget</span> -i wget_links.txt</a></code></pre></div>
<p>After downloading a <strong>MD5-check</strong> is strongly encouraged.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash"><code class="sourceCode bash"><a class="sourceLine" id="cb2-1" data-line-number="1"><span class="co">#create a reference table with filenames and expected MD5-sum</span></a>
<a class="sourceLine" id="cb2-2" data-line-number="2"><span class="fu">sed</span> 1d ../filereport_read_run_PRJNA433520_tsv.txt <span class="kw">|</span> <span class="fu">cut</span> -f 4,5 <span class="kw">|</span> <span class="fu">awk</span> <span class="st">'sub(".*/",$1" ")'</span> <span class="op">&gt;</span> md5_check.txt</a>
<a class="sourceLine" id="cb2-3" data-line-number="3"></a>
<a class="sourceLine" id="cb2-4" data-line-number="4"><span class="co">#perform the check with md5sum</span></a>
<a class="sourceLine" id="cb2-5" data-line-number="5"><span class="ex">md5sum</span> -c md5_check.txt </a></code></pre></div>
<p>You should only proceed, if all MD5-sums match!</p>
<div id="preparing-fastq-files" class="section level2">
<h2 class="hasAnchor">
<a href="#preparing-fastq-files" class="anchor"></a>Preparing FASTQ-files</h2>
<p>As described in Wuidart et al., the sequences likely contain left-over adapter sequences and an adapter trimming is recommended.</p>
<p>We will trim the reads to remove adapters and perform a very relaxed quality filtering. We will use the tool <code>trim-galore</code> (Version 0.6.4_dev) for this task.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash"><code class="sourceCode bash"><a class="sourceLine" id="cb3-1" data-line-number="1"><span class="co">#create needed folders</span></a>
<a class="sourceLine" id="cb3-2" data-line-number="2"><span class="fu">mkdir</span> -p ../trim/fastqc/multiqc</a>
<a class="sourceLine" id="cb3-3" data-line-number="3"></a>
<a class="sourceLine" id="cb3-4" data-line-number="4"><span class="ex">trim_galore</span> -fastqc_args <span class="st">"--outdir 'YOUR_PATH'/trim/fastqc/ -t 4"</span> --stringency 3 --trim-n --cores 4 -o ../trim/ *.fastq.gz</a></code></pre></div>
<p>The trimmed files are the newly created folder <code>/trim</code>. By looking at the generated <code>FastQC</code> reports, we can check how many sequences have been removed by the trimming, and if other QC paramters indicate problems. Optionally we can summarize the hundreds of reports to a single report with <code>MultiQC</code>:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode bash"><code class="sourceCode bash"><a class="sourceLine" id="cb4-1" data-line-number="1"><span class="bu">cd</span> ../trim</a>
<a class="sourceLine" id="cb4-2" data-line-number="2"><span class="ex">multiqc</span> -o fastqc/multiqc -ip .</a></code></pre></div>
</div>
<div id="transcript-level-quantification" class="section level2">
<h2 class="hasAnchor">
<a href="#transcript-level-quantification" class="anchor"></a>Transcript level quantification</h2>
<p>The files are now ready for transcript level quantification. We will use the tool <code>Salmon</code> (Version 1.1.0) for this. Although we have single-cell RNA-seq data, samples/cells from the Smart-seq2 protocol can be treated as standard bulk samples (you have quite a lot of them, though). Therefore we can simply use the <code>Salmon</code> for quantification and do not have to use specialised single-cell variants like <code>Alevin</code>.</p>
<p><code>Salmon</code> does not perform a standard genomic alignment, but performs a quasi-mapping directly to the transcriptome. Reads, that could be originating from multiple transcript isoforms, are assigned to equivalence classes, where actual counts are derived by expectation maximization algorithms.</p>
<p>To run the <code>Salmon</code> quantification, we first need a transcriptomic index. For this example, we will use the <a href="ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_34/gencode.v34.transcripts.fa.gz">transcript sequences</a> and <a href="ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_34/gencode.v34.annotation.gtf.gz">comprehensive gene annotation</a> from <a href="https://www.gencodegenes.org/">Gencode</a> <a href="https://www.gencodegenes.org/human/release_34.html">Version 34</a>.</p>
<p>Additionally, in this experiment ERCC Spike-In sequences have been used in the sequencing. To also be able to quantify those not naturally occuring sequences, we should add those manually to the transcriptome and the annotation.</p>
<p>We can get the ERCC <code>FASTA</code>-sequences and a corresponing <code>GTF</code>-file directly from <a href="https://assets.thermofisher.com/TFS-Assets/LSG/manuals/ERCC92.zip">ThermoFisher</a> and combine those files with the Gencode files:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode bash"><code class="sourceCode bash"><a class="sourceLine" id="cb5-1" data-line-number="1"><span class="co">#create folder and change directory</span></a>
<a class="sourceLine" id="cb5-2" data-line-number="2"><span class="fu">mkdir</span> ../salmon</a>
<a class="sourceLine" id="cb5-3" data-line-number="3"><span class="bu">cd</span> ../salmon</a>
<a class="sourceLine" id="cb5-4" data-line-number="4"></a>
<a class="sourceLine" id="cb5-5" data-line-number="5"><span class="co">#download gencode</span></a>
<a class="sourceLine" id="cb5-6" data-line-number="6"><span class="ex">curl</span> -LO ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_34/gencode.v34.transcripts.fa.gz</a>
<a class="sourceLine" id="cb5-7" data-line-number="7"><span class="fu">gunzip</span> gencode.v34.transcripts.fa.gz</a>
<a class="sourceLine" id="cb5-8" data-line-number="8"><span class="ex">curl</span> -LO ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_34/gencode.v34.annotation.gtf.gz</a>
<a class="sourceLine" id="cb5-9" data-line-number="9"><span class="fu">gunzip</span> gencode.v34.annotation.gtf.gz</a>
<a class="sourceLine" id="cb5-10" data-line-number="10"></a>
<a class="sourceLine" id="cb5-11" data-line-number="11"><span class="co">#download ERCC</span></a>
<a class="sourceLine" id="cb5-12" data-line-number="12"><span class="ex">curl</span> -LO https://assets.thermofisher.com/TFS-Assets/LSG/manuals/ERCC92.zip</a>
<a class="sourceLine" id="cb5-13" data-line-number="13"><span class="fu">unzip</span> ERCC92.zip</a>
<a class="sourceLine" id="cb5-14" data-line-number="14"></a>
<a class="sourceLine" id="cb5-15" data-line-number="15"><span class="co">#combine references</span></a>
<a class="sourceLine" id="cb5-16" data-line-number="16"><span class="fu">cat</span> ../tabula_muris/gencode.vM24.transcripts.fa ERCC92.fa <span class="op">&gt;</span> gencode.vM24.transcripts_spike.fa</a>
<a class="sourceLine" id="cb5-17" data-line-number="17"><span class="fu">cat</span> ../tabula_muris/gencode.vM24.annotation.gtf ERCC92.gtf <span class="op">&gt;</span> gencode.vM24.annotation_spike.gtf</a>
<a class="sourceLine" id="cb5-18" data-line-number="18"></a>
<a class="sourceLine" id="cb5-19" data-line-number="19"><span class="co">#optionally remove original gencode files</span></a>
<a class="sourceLine" id="cb5-20" data-line-number="20"><span class="fu">rm</span> gencode.v34.transcripts.fa.gz</a>
<a class="sourceLine" id="cb5-21" data-line-number="21"><span class="fu">rm</span> gencode.v34.annotation.gtf.gz</a></code></pre></div>
<p>This reference <code>FASTA</code> including the ERCC-sequences can then be used to build the <code>Salmon</code> index:</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode bash"><code class="sourceCode bash"><a class="sourceLine" id="cb6-1" data-line-number="1"><span class="co">#installation in conda evironment</span></a>
<a class="sourceLine" id="cb6-2" data-line-number="2"><span class="ex">conda</span> install salmon</a>
<a class="sourceLine" id="cb6-3" data-line-number="3"></a>
<a class="sourceLine" id="cb6-4" data-line-number="4"><span class="co">#build index - `p` specifies the number of threads to use</span></a>
<a class="sourceLine" id="cb6-5" data-line-number="5"><span class="ex">salmon</span> index -t ../gencode.vM24.transcripts_spike.fa --gencode -p 4 -i index</a></code></pre></div>
<p>The actual quantification can be performed in <em>bash</em> with:</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode bash"><code class="sourceCode bash"><a class="sourceLine" id="cb7-1" data-line-number="1"><span class="bu">cd</span> ../trim</a>
<a class="sourceLine" id="cb7-2" data-line-number="2"></a>
<a class="sourceLine" id="cb7-3" data-line-number="3"><span class="co">#loop over files</span></a>
<a class="sourceLine" id="cb7-4" data-line-number="4"><span class="kw">for</span> <span class="ex">i</span> in *.fq.gz<span class="kw">;</span> <span class="kw">do</span> <span class="ex">salmon</span> quant -l A -i ../salmon/index/ -r <span class="va">$i</span> --seqBias --gcBias --validateMappings --rangeFactorizationBins 4 --threads 4 -o ../salmon/<span class="va">${i/</span>_trimmed.fq.gz<span class="va">/}</span><span class="kw">;</span> <span class="kw">done</span></a></code></pre></div>
<hr>
<p><strong>This concludes the preprocessing of the data. Please see the corresponding analysis vignette for an usage example of DTUrtle.</strong></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Tobias Tekath.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  
<script src="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.js" integrity="sha256-GKvGqXDznoRYHCwKXGnuchvKSwmx9SRMrZOTh2g4Sb0=" crossorigin="anonymous"></script><script>
  docsearch({
    
    
    apiKey: 'a6c9f43b010fdd4b2e64792389fd67d2',
    indexName: 'dturtle',
    inputSelector: 'input#search-input.form-control',
    transformData: function(hits) {
      return hits.map(function (hit) {
        hit.url = updateHitURL(hit);
        return hit;
      });
    }
  });
</script>
</body>
</html>
