<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Visualize as barplot — plot_proportion_barplot • DTUrtle</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Visualize as barplot — plot_proportion_barplot" />
<meta property="og:description" content="Visualize genes and it's transcript proportions in a barplot." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">DTUrtle</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../reference/index.html">Functions and data</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/mkin.html">Vignette human bulk RNA-seq</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Visualize as barplot</h1>
    
    <div class="hidden name"><code>plot_proportion_barplot.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Visualize genes and it's transcript proportions in a barplot.</p>
    </div>

    <pre class="usage"><span class='fu'>plot_proportion_barplot</span>(
  <span class='no'>dturtle</span>,
  <span class='kw'>genes</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>meta_gene_id</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>group_colors</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>fit_line_color</span> <span class='kw'>=</span> <span class='st'>"red"</span>,
  <span class='kw'>savepath</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>filename_ext</span> <span class='kw'>=</span> <span class='st'>"_barplot.png"</span>,
  <span class='kw'>BPPARAM</span> <span class='kw'>=</span> <span class='kw pkg'>BiocParallel</span><span class='kw ns'>::</span><span class='fu'><a href='https://rdrr.io/pkg/BiocParallel/man/SerialParam-class.html'>SerialParam</a></span>(),
  <span class='no'>...</span>
)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>dturtle</th>
      <td><p><code>dturtle</code> result object of <code><a href='posthoc_and_stager.html'>posthoc_and_stager()</a></code>.</p></td>
    </tr>
    <tr>
      <th>genes</th>
      <td><p>Character vector of genes to plot. If <code>NULL</code>, defaults to all found significant genes (<code>sig_genes</code>).</p></td>
    </tr>
    <tr>
      <th>meta_gene_id</th>
      <td><p>Optionally specify the column name in <code>meta_table_gene</code>, which contains real gene identifiers or gene names.</p></td>
    </tr>
    <tr>
      <th>group_colors</th>
      <td><p>Optionally specify the colours for the two sample groups in the plot. Must be a named vector, with the group names as names.</p></td>
    </tr>
    <tr>
      <th>fit_line_color</th>
      <td><p>Optionally specify the colour to use for the mean fit line.</p></td>
    </tr>
    <tr>
      <th>savepath</th>
      <td><p>If you want your files to be saved to disk, specify a save path here. The directories will be created if necessary.</p></td>
    </tr>
    <tr>
      <th>filename_ext</th>
      <td><p>Optionally specify a file name extension here, which also defines the save image format. The file name will be 'gene_name+extension'.</p></td>
    </tr>
    <tr>
      <th>BPPARAM</th>
      <td><p>If multicore processing should be used, specify a <code>BiocParallelParam</code> object here. Among others, can be <code>SerialParam()</code> (default) for non-multicore processing or <code>MulticoreParam('number_cores')</code> for multicore processing. See <code><a href='https://rdrr.io/pkg/BiocParallel/man/BiocParallel-package.html'>BiocParallel</a></code> for more information.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>Arguments passed on to <code><a href='https://ggplot2.tidyverse.org/reference/ggsave.html'>ggplot2::ggsave</a></code></p><dl'>
    <dt><code>filename</code></dt><dd><p>File name to create on disk.</p></dd>
    <dt><code>plot</code></dt><dd><p>Plot to save, defaults to last plot displayed.</p></dd>
    <dt><code>device</code></dt><dd><p>Device to use. Can either be a device function
(e.g. <code><a href='https://rdrr.io/r/grDevices/png.html'>png()</a></code>), or one of "eps", "ps", "tex" (pictex),
"pdf", "jpeg", "tiff", "png", "bmp", "svg" or "wmf" (windows only).</p></dd>
    <dt><code>path</code></dt><dd><p>Path of the directory to save plot to: <code>path</code> and <code>filename</code>
are combined to create the fully qualified file name. Defaults to the
working directory.</p></dd>
    <dt><code>scale</code></dt><dd><p>Multiplicative scaling factor.</p></dd>
    <dt><code>width</code></dt><dd><p>Plot size in <code>units</code> ("in", "cm", or "mm").
If not supplied, uses the size of current graphics device.</p></dd>
    <dt><code>height</code></dt><dd><p>Plot size in <code>units</code> ("in", "cm", or "mm").
If not supplied, uses the size of current graphics device.</p></dd>
    <dt><code>units</code></dt><dd><p>Plot size in <code>units</code> ("in", "cm", or "mm").
If not supplied, uses the size of current graphics device.</p></dd>
    <dt><code>dpi</code></dt><dd><p>Plot resolution. Also accepts a string input: "retina" (320),
"print" (300), or "screen" (72). Applies only to raster output types.</p></dd>
    <dt><code>limitsize</code></dt><dd><p>When <code>TRUE</code> (the default), <code>ggsave</code> will not
save images larger than 50x50 inches, to prevent the common error of
specifying dimensions in pixels.</p></dd>
  
</dl></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>If no <code>savepath</code> is provided, returns a list of the created plots for further processing.</p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>Shows the transcripts proportional change per analysis group, together with the mean fit value via a horizontal line. Significant transcript's names are marked in red.</p>
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code><a href='run_drimseq.html'>run_drimseq()</a></code> and <code><a href='posthoc_and_stager.html'>posthoc_and_stager()</a></code> for DTU object creation. <code><a href='create_dtu_table.html'>create_dtu_table()</a></code> and <code><a href='plot_dtu_table.html'>plot_dtu_table()</a></code> for table visualization.</p>
<p>Other DTUrtle visualization: 
<code><a href='create_dtu_table.html'>create_dtu_table</a>()</code>,
<code><a href='plot_proportion_pheatmap.html'>plot_proportion_pheatmap</a>()</code>,
<code><a href='plot_transcripts_view.html'>plot_transcripts_view</a>()</code></p></div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by <a href='https://hesselberthlab.org'>Tobias Tekath</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


