<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Visualize transcripts on genomic scale — plot_transcripts_view • DTUrtle</title>

<!-- favicons -->
<link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png" />
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png" />
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png" />
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png" />

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>


<!-- docsearch -->
<script src="../docsearch.js"></script>
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.3/docsearch.min.css" integrity="sha256-QOSRU/ra9ActyXkIBbiIB144aDBdtvXBcNc3OTNuX/Q=" crossorigin="anonymous" />
<link href="../docsearch.css" rel="stylesheet">
<script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script>



<meta property="og:title" content="Visualize transcripts on genomic scale — plot_transcripts_view" />
<meta property="og:description" content="Visualize exon-intron structure of transcripts on genomic scale." />
<meta property="og:image" content="https://tobitekath.github.io/DTUrtle/logo.svg" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">DTUrtle</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.7.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../reference/index.html">Functions and data</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/index.html">Vignette overview</a>
    </li>
    <li class="divider"></li>
    <li class="dropdown-header">Single-cell</li>
    <li>
      <a></a>
    </li>
    <li class="dropdown-header">Illumina SureCell</li>
    <li>
      <a href="../articles/Hoffman_human_single-cell_preprocess.html">Human single-cell RNA-seq pre-processing</a>
    </li>
    <li>
      <a href="../articles/Hoffman_human_single-cell_analysis.html">Human single-cell RNA-seq analysis</a>
    </li>
    <li>
      <a></a>
    </li>
    <li class="dropdown-header">10X Chromium</li>
    <li>
      <a href="../articles/Tabular_Muris_mouse_single-cell_preprocess.html">Mouse Tabular Muris single-cell RNA-seq pre-processing</a>
    </li>
    <li>
      <a href="../articles/Tabular_Muris_mouse_single-cell_analysis.html">Mouse Tabular Muris single-cell RNA-seq analysis</a>
    </li>
    <li class="divider"></li>
    <li class="dropdown-header">Bulk</li>
    <li>
      <a href="../articles/Hoffman_human_bulk_preprocess.html">Human bulk RNA-seq pre-processing</a>
    </li>
    <li>
      <a href="../articles/Hoffman_human_bulk_analysis.html">Human bulk RNA-seq analysis</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/TobiTekath/DTUrtle/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
      <form class="navbar-form navbar-right hidden-xs hidden-sm" role="search">
        <div class="form-group">
          <input type="search" class="form-control" name="search-input" id="search-input" placeholder="Search..." aria-label="Search for..." autocomplete="off">
        </div>
      </form>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Visualize transcripts on genomic scale</h1>
    <small class="dont-index">Source: <a href='https://github.com/TobiTekath/DTUrtle/blob/master/R/visualization.R'><code>R/visualization.R</code></a></small>
    <div class="hidden name"><code>plot_transcripts_view.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Visualize exon-intron structure of transcripts on genomic scale.</p>
    </div>

    <pre class="usage"><span class='fu'>plot_transcripts_view</span><span class='op'>(</span>
  <span class='va'>dturtle</span>,
  genes <span class='op'>=</span> <span class='cn'>NULL</span>,
  <span class='va'>gtf</span>,
  <span class='va'>genome</span>,
  one_to_one <span class='op'>=</span> <span class='cn'>NULL</span>,
  reduce_introns <span class='op'>=</span> <span class='cn'>T</span>,
  reduce_introns_fill <span class='op'>=</span> <span class='st'>"grey95"</span>,
  reduce_introns_min_size <span class='op'>=</span> <span class='fl'>50</span>,
  fontsize_vec <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>10</span>, <span class='fl'>1.1</span>, <span class='fl'>0.6</span><span class='op'>)</span>,
  arrow_colors <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"#7CAE00"</span>, <span class='st'>"#00BFC4"</span><span class='op'>)</span>,
  arrow_start <span class='op'>=</span> <span class='fl'>0.88</span>,
  extension_factors <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>0.025</span>, <span class='fl'>0.22</span><span class='op'>)</span>,
  savepath <span class='op'>=</span> <span class='cn'>NULL</span>,
  filename_ext <span class='op'>=</span> <span class='st'>"_transcripts.png"</span>,
  add_to_table <span class='op'>=</span> <span class='cn'>F</span>,
  BPPARAM <span class='op'>=</span> <span class='fu'>BiocParallel</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/BiocParallel/man/SerialParam-class.html'>SerialParam</a></span><span class='op'>(</span><span class='op'>)</span>,
  <span class='va'>...</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>dturtle</th>
      <td><p><code>dturtle</code> result object of <code><a href='posthoc_and_stager.html'>posthoc_and_stager()</a></code>.</p></td>
    </tr>
    <tr>
      <th>genes</th>
      <td><p>Character vector of genes to plot. If <code>NULL</code>, defaults to all found significant genes (<code>sig_genes</code>).</p></td>
    </tr>
    <tr>
      <th>gtf</th>
      <td><p>Either path to a <code>gtf/gff</code> file which will be read or a <code>granges</code> object of a already read in <code>gtf/gff</code> file. Such an object can be created with <code><a href='import_gtf.html'>import_gtf("GTF_PATH", feature_type=NULL, out_df=FALSE)</a></code>. See <code><a href='import_gtf.html'>import_gtf()</a></code> for more information.</p></td>
    </tr>
    <tr>
      <th>genome</th>
      <td><p>The genome on which to create the ideogram tracks. This has to be a valid <code>UCSC genome identifier</code> (e.g. 'hg38', 'mm10', 'danRer11', etc.). Can also be NULL to skip ideogram track generation.</p></td>
    </tr>
    <tr>
      <th>one_to_one</th>
      <td><p>Specify <code>TRUE</code>, if one_to_one mapping of gene/transcript identifiers with their respective names was enforced before (with <code><a href='one_to_one_mapping.html'>one_to_one_mapping()</a></code>). If a non default extension character (<code>ext</code>) has been used, please specify the used extension character.</p></td>
    </tr>
    <tr>
      <th>reduce_introns</th>
      <td><p>Logical if intron ranges shall be shrunken down, highlighting the exonic structure.</p></td>
    </tr>
    <tr>
      <th>reduce_introns_fill</th>
      <td><p>Optionally specify the background color of ranges where introns have been reduced.</p></td>
    </tr>
    <tr>
      <th>reduce_introns_min_size</th>
      <td><p>Specify the minimal size introns are reduced to (in bp).</p></td>
    </tr>
    <tr>
      <th>fontsize_vec</th>
      <td><p>Vector of fontsizes to use. The first value is the side annotation text fontsize (in pt), the second value the cex factor for the title, the third value the cex factor for the feature &amp; chromosome names.</p></td>
    </tr>
    <tr>
      <th>arrow_colors</th>
      <td><p>Specify the colors of the arrows indicating the direction of proportional changes. The first color string is for a positive change, the second for a negative one.</p></td>
    </tr>
    <tr>
      <th>arrow_start</th>
      <td><p>Advanced: Set the x coordinate of the arrow annotations (in NPC)</p></td>
    </tr>
    <tr>
      <th>extension_factors</th>
      <td><p>Advanced: Specify the extension factors to extend the plotted genomic range. The first value if for the extension of the front (left) side, the second for the back (right).</p></td>
    </tr>
    <tr>
      <th>savepath</th>
      <td><p>If you want your files to be saved to disk, specify a save path here. The directories will be created if necessary.</p></td>
    </tr>
    <tr>
      <th>filename_ext</th>
      <td><p>Optionally specify a file name extension here, which also defines the save image format. The file name will be 'gene_name+extension'.</p></td>
    </tr>
    <tr>
      <th>add_to_table</th>
      <td><p>If a <code>savepath</code> is provided, add the filepaths of the created plots to the corresponding entries in <code>dtu_table</code>. The name of the column that shall be created can be specified here.</p></td>
    </tr>
    <tr>
      <th>BPPARAM</th>
      <td><p>If multicore processing should be used, specify a <code>BiocParallelParam</code> object here. Among others, can be <code>SerialParam()</code> (default) for non-multicore processing or <code>MulticoreParam('number_cores')</code> for multicore processing. See <code><a href='https://rdrr.io/pkg/BiocParallel/man/BiocParallel-package.html'>BiocParallel</a></code> for more information.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>Arguments passed down to <code><a href='https://rdrr.io/r/grDevices/png.html'>png</a></code> or <code><a href='https://rdrr.io/r/grDevices/cairo.html'>cairo_pdf</a></code> or <code><a href='https://rdrr.io/r/grDevices/pdf.html'>pdf</a></code> or <code><a href='https://rdrr.io/r/grDevices/png.html'>jpeg</a></code>, depending on <code>filename_ext</code> ending and capabilities (cairo_pdf or pdf).</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>Returns list of saved plots, for adding to the DTU table. If no <code>savepath</code> is provided, returns a list of the created plots for further processing.</p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>Reduced intron length is computed by taking the square root, but is not less than the specified <code>reduce_introns_min_size</code> length.
If less <code>GRanges</code> are found than expected, try setting <code>one_to_one</code> to <code>TRUE</code> or the used extension character.
When calling this function many times, try importing the GTF-File with <code>gtf &lt;- rtracklayer::import("PATH_TO_GTF")</code> once and pass it to the <code>gtf</code> parameter to improve performance.</p>
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code><a href='run_drimseq.html'>run_drimseq()</a></code> and <code><a href='posthoc_and_stager.html'>posthoc_and_stager()</a></code> for DTU object creation. <code><a href='create_dtu_table.html'>create_dtu_table()</a></code> and <code><a href='plot_dtu_table.html'>plot_dtu_table()</a></code> for table visualization.</p>
<p>Other DTUrtle visualization: 
<code><a href='create_dtu_table.html'>create_dtu_table</a>()</code>,
<code><a href='plot_dimensional_reduction.html'>plot_dimensional_reduction</a>()</code>,
<code><a href='plot_proportion_barplot.html'>plot_proportion_barplot</a>()</code>,
<code><a href='plot_proportion_pheatmap.html'>plot_proportion_pheatmap</a>()</code></p></div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Tobias Tekath.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  
<script src="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.js" integrity="sha256-GKvGqXDznoRYHCwKXGnuchvKSwmx9SRMrZOTh2g4Sb0=" crossorigin="anonymous"></script>
<script>
  docsearch({
    
    
    apiKey: 'a6c9f43b010fdd4b2e64792389fd67d2',
    indexName: 'dturtle',
    inputSelector: 'input#search-input.form-control',
    transformData: function(hits) {
      return hits.map(function (hit) {
        hit.url = updateHitURL(hit);
        return hit;
      });
    }
  });
</script>



  </body>
</html>


