<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Main DRIMSeq results — run_drimseq • DTUrtle</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Main DRIMSeq results — run_drimseq" />
<meta property="og:description" content="Compute the main DRIMSeq results." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">DTUrtle</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../reference/index.html">Functions and data</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/mkin.html">Vignette human bulk RNA-seq</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Main DRIMSeq results</h1>
    
    <div class="hidden name"><code>run_drimseq.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Compute the main DRIMSeq results.</p>
    </div>

    <pre class="usage"><span class='fu'>run_drimseq</span>(
  <span class='no'>counts</span>,
  <span class='no'>tx2gene</span>,
  <span class='no'>pd</span>,
  <span class='kw'>id_col</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='no'>cond_col</span>,
  <span class='kw'>cond_levels</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>filtering_strategy</span> <span class='kw'>=</span> <span class='st'>"bulk"</span>,
  <span class='kw'>BPPARAM</span> <span class='kw'>=</span> <span class='kw pkg'>BiocParallel</span><span class='kw ns'>::</span><span class='fu'><a href='https://rdrr.io/pkg/BiocParallel/man/SerialParam-class.html'>SerialParam</a></span>(),
  <span class='kw'>force_dense</span> <span class='kw'>=</span> <span class='no'>T</span>,
  <span class='kw'>carry_over_metadata</span> <span class='kw'>=</span> <span class='no'>T</span>,
  <span class='no'>...</span>
)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>counts</th>
      <td><p>Can be either:</p><ol>
<li><p>(sparse) matrix with feature counts, where the rows correspond to features (e.g. transcripts). One column per sample/cell with the count data for the specified features must be present (The names of these columns must match with the identifiers in <code>id_col</code>).</p></li>
<li><p>Seurat object with a transcription level assay as <code>active.assay</code> (most likely result object from <code><a href='combine_to_matrix.html'>combine_to_matrix()</a></code>)</p></li>
</ol></td>
    </tr>
    <tr>
      <th>tx2gene</th>
      <td><p>Data frame, where the first column consists of feature identifiers and the second column consists of corresponding gene identifiers. Feature identifiers must match with the rownames of the counts object. If a Seurat object is provided in <code>counts</code> and <code>tx2gene</code> was provided in <code><a href='combine_to_matrix.html'>combine_to_matrix()</a></code>, a vector of the colnames of the specific feature and gene identifierss is sufficient.</p></td>
    </tr>
    <tr>
      <th>pd</th>
      <td><p>Data frame with at least a column of sample/cell identifiers (rownames or specified in <code>id_col</code>) and the comparison group definition (specified in <code>cond_col</code>).</p></td>
    </tr>
    <tr>
      <th>id_col</th>
      <td><p>Name of the column in <code>pd</code>, where unique sample/cell identifiers are stored. If <code>NULL</code> (default), use rownames of <code>pd</code>.</p></td>
    </tr>
    <tr>
      <th>cond_col</th>
      <td><p>Name of the column in <code>pd</code>, where the comparison groups/conditions are defined. If more than 2 levels/groups are present, the groups that should be used must be specified in <code>cond_levels</code>.</p></td>
    </tr>
    <tr>
      <th>cond_levels</th>
      <td><p>Define two levels/groups of <code>cond_col</code>, that should be compared. The order of the levels states the comparison formula (i.e. <code>cond_col[1]-cond_col[2]</code>).</p></td>
    </tr>
    <tr>
      <th>filtering_strategy</th>
      <td><p>Define the filtering strategy to reduce and noise and increase statistical power.</p><ul>
<li><p><code>'bulk'</code>: Predefined strategy for bulk RNAseq data. (default)</p></li>
<li><p><code>'sc'</code>: Predefined strategy for single-cell RNAseq data.</p></li>
<li><p><code>'own'</code>: Can be used to specify a user-defined strategy via the <code>...</code> argument (using the parameters of <code>dmFilter()</code>).</p></li>
</ul></td>
    </tr>
    <tr>
      <th>BPPARAM</th>
      <td><p>If multicore processing should be used, specify a <code>BiocParallelParam</code> object here. Among others, can be <code>SerialParam()</code> (default) for non-multicore processing or <code>MulticoreParam('number_cores')</code> for multicore processing. See <code><a href='https://rdrr.io/pkg/BiocParallel/man/BiocParallel-package.html'>BiocParallel</a></code> for more information.</p></td>
    </tr>
    <tr>
      <th>force_dense</th>
      <td><p>If you do not want to use a sparse Matrix for DRIMSeq calculations, you can force a dense conversion by specifying <code>TRUE</code>. Increases memory usage, but also reduces runtime drastically (currently).</p></td>
    </tr>
    <tr>
      <th>carry_over_metadata</th>
      <td><p>Specify if compatible additional columns of <code>tx2gene</code> shall be carried over to the gene and transcript level <code>meta_table</code> in the results. Columns with <code>NA</code> values are not carried over.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>If <code>filtering_strategy='own'</code> specify the wished parameters of <code><a href='https://rdrr.io/pkg/sparseDRIMSeq/man/dmFilter.html'>dmFilter</a></code> here.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p><code>dturtle</code> object with the key results, that can be used in the DTUrtle steps hereafter. The object is just a easily accesible list with the following items:</p><ul>
<li><p><code>meta_table_gene</code>: Data frame of the expressed-in ratio of all genes. Expressed-in is defined as expression &gt; 0. Can be used to add gene level meta-information for plotting.</p></li>
<li><p><code>meta_table_tx</code>: Data frame of the expressed-in ratio of all transcripts. Expressed-in is defined as expression &gt; 0. Can be used to add transcript level meta-information for plotting.</p></li>
<li><p><code>meta_table_sample</code>: Data frame of the provided sample level information (<code>pd</code>). Can be used to add sample level meta-information for plotting.</p></li>
<li><p><code>drim</code>: The results of the DRIMSeq statistical computations (<code>dmTest()</code>).</p></li>
<li><p><code>design</code>: The design matrix generated from the specified <code>pd</code> columns.</p></li>
<li><p><code>group</code>: Vector which sample/cell belongs to which comparison group.</p></li>
<li><p><code>used_filtering_options</code>: Vector of the used filtering options.</p></li>
</ul>

    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>Run the main DRIMSeq pipeline, including generation of a design matrix, gene/feature filtering and running the statistical computations of DRIMSeq (<code>dmPrecision()</code>, <code>dmFit()</code> and <code>dmTest()</code>)</p>
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p>Other DTUrtle: 
<code><a href='combine_to_matrix.html'>combine_to_matrix</a>()</code>,
<code><a href='import_counts.html'>import_counts</a>()</code>,
<code><a href='posthoc_and_stager.html'>posthoc_and_stager</a>()</code></p></div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by <a href='https://hesselberthlab.org'>Tobias Tekath</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


